---
title: 'Blood biomarkers representing maternal-fetal interface tissues used to 
        predict early-and late-onset preeclampsia but not COVID-19 infection'
author:
  - name: Herdiantri Sufriyana
    affiliation:
    - &gibi Graduate Institute of Biomedical Informatics, College of Medical
      Science and Technology, Taipei Medical University, Taipei, Taiwan
    - Department of Medical Physiology, Faculty of Medicine, Universitas 
      Nahdlatul Ulama Surabaya, Surabaya, Indonesia
  - name: Hotimah Masdan Salim
    affiliation:
    - Department of Molecular Biology, Faculty of Medicine, Universitas 
      Nahdlatul Ulama Surabaya, Surabaya, Indonesia
  - name: Akbar Reza Muhammad
    affiliation:
    - Faculty of Medicine, Universitas Nahdlatul Ulama Surabaya, Surabaya, 
      Indonesia
  - name: Yu-Wei Wu
    affiliation:
    - *gibi
    - &tmuh Clinical Big Data Research Center, Taipei Medical University
      Hospital, Taipei, Taiwan
  - name: Emily Chia-Yu Su
    affiliation:
    - *gibi
    - *tmuh
    - Research Center for Artificial Intelligence in Medicine, Taipei Medical
      University, Taipei, Taiwan
    email: emilysu@tmu.edu.tw
output: 
  html_document
  # pdf_document
  # word_document:
  #   reference_docx: styles_pest.docx
always_allow_html: yes
---

# Set up programming environment

```{r Stop codes if R version is false}
r_version='4.0.2'
if(paste0(version$major,'.',version$minor)!=r_version){
  stop(paste0('This code uses R version of ',r_version))
}
```

```{r Set up reproducible environment, include=FALSE}
renv::init(restart=F)
renv::activate()
renv::restore(prompt=T)
```

```{r Set sample kind, include=FALSE}
# sample.kind=NULL # if using R 3.5 or earlier
sample.kind='Rounding' # if using R 3.6 or later
```

```{r Set to run or not run very heavy computations, include=FALSE}
# Many computations were very heavy;
# thus, we provided the RDS files as substitutes and load only ones
# that can be ran in most computers.
# Set to TRUE if you want to run the very heavy computations.
run_heavy_computation=F
```

```{r Recommended memory limit, include=FALSE}
# Minimum 1 GB free memory should be allocated
memory.limit(size=1000000000)

# Minimum connection size
Sys.setenv('VROOM_CONNECTION_SIZE'=128000*1024)
```

```{r Load packages, include=FALSE}
library(tidyverse)
library(dslabs)
library(GEOquery)
library(Biobase)
library(biomaRt)
library(oligo)
library(RSQLite)
library(DBI)
library(Rcpp)
library(pd.hugene.1.0.st.v1)
library(pd.hta.2.0)
library(pd.clariom.s.human)
library(limma)
library(illuminaio)
library(parallelDist)
library(WGCNA)
library(sva)
library(pbapply)
library(lubridate)
library(rsdr)
library(caret)
library(glmnet)
library(Rborist)
library(gbm)
library(gam)
library(MLeval)
library(doParallel)
library(preprocessCore)
library(matrixStats)
library(Rtsne)
library(clixo)
library(divnn)
library(igraph)
library(reticulate)
library(ggnetwork)
library(rpart.plot)
library(knitr)
library(kableExtra)
  
options(dplyr.summarise.inform=F)
dslabs::ds_theme_set()
select=dplyr::select
summarize=dplyr::summarize
rename=dplyr::rename
slice=dplyr::slice
reticulate::use_condaenv('./renv/python/condaenvs/renv-python',required=T)
renv::use_python(name='./renv/python/condaenvs/renv-python')
```

```{r}
timestamper=c()
timestamper[1]=as.character(now())
```

# Data load

```{r Create empty lists for DEA-related variables, include=FALSE}
pdata=list()
edata=list()
fdata=list()
rdata=list()
cdata=list()
odata=list()
ldata=list()
gdata=list()
```

## Phenotype data

```{r Load GSE108497 phenotype, include=FALSE}
# Expression set with the processed transcript expression
pdata$GSE108497$eset=
  suppressWarnings(suppressMessages(
    getGEO(
      filename='data/GSE108497/GSE108497_series_matrix.txt.gz'
      ,getGPL=F
      ,parseCharacteristics=F
    )
  ))

# Metadata of phenotype variables
pdata$GSE108497$phenolabel=
  suppressWarnings(suppressMessages(
    read_csv('data/GSE108497/GSE108497_varMetadata.csv') %>%
      mutate(code=str_to_lower(code))
  ))

# Phenotype data
pdata$GSE108497$phenotype=
  pdata$GSE108497$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(colnames(.) %>% str_remove_all(':ch1'))
```

```{r Describe phenotype metadata of GSE108497, eval=FALSE, include=FALSE}
pdata$GSE108497$phenolabel
```

```{r Explore raw phenotype of GSE108497, eval=FALSE, include=FALSE}
rbind(
  
    # Not SLE, not PDD, 34 weeks' gestation or more
    pdata$GSE108497$phenotype %>%
      filter(
        sle==0
        & pe==0
        & iugr==0
        & sga==0
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)>=34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='not_sle_not_pdd_late')
    
    # Not SLE, not PDD, <34 weeks' gestation
    ,pdata$GSE108497$phenotype %>%
      filter(
        sle==0
        & pe==0
        & iugr==0
        & sga==0
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)<34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='not_sle_not_pdd_early')
    
    # SLE, not PDD, 34 weeks' gestation or more
    ,pdata$GSE108497$phenotype %>%
      filter(
        sle==1
        & pe==0
        & iugr==0
        & sga==0
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)>=34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='sle_not_pdd_late')
    
    # SLE, not PDD, <34 weeks' gestation
    ,pdata$GSE108497$phenotype %>%
      filter(
        sle==1
        & pe==0
        & iugr==0
        & sga==0
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)<34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='sle_not_pdd_early')
    
    # SLE, isolated FGR, 34 weeks' gestation or more
    ,pdata$GSE108497$phenotype %>%
      filter(
        sle==1
        & pe==0
        & sga==1
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)>=34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='sle_isga_late')
    
    # SLE, isolated FGR, <34 weeks' gestation
    ,pdata$GSE108497$phenotype %>%
      filter(
        sle==1
        & pe==0
        & sga==1
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)<34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='sle_isga_early')
    
    # SLE, not isolated PE, 34 weeks' gestation or more
    ,pdata$GSE108497$phenotype %>%
      filter(
        sle==1
        & pe==1
        & sga==0
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)>=34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='sle_not_ipe_late')
    
    # SLE, not isolated PE, <34 weeks' gestation
    ,pdata$GSE108497$phenotype %>%
      filter(
        sle==1
        & pe==1
        & sga==0
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)<34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='sle_not_ipe_early')
    
    # SLE, PE-FGR, 34 weeks' gestation or more
    ,pdata$GSE108497$phenotype %>%
      filter(
        sle==1
        & pe==1
        & sga==1
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)>=34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='sle_pe_sga_late')
    
    # SLE, PE-FGR, <34 weeks' gestation
    ,pdata$GSE108497$phenotype %>%
      filter(
        sle==1
        & pe==1
        & sga==1
        & apl==0
        & lac==0
        & fd==0
        & nnd==0
        & as.numeric(ga_at_end_of_pregnancy)<34
      ) %>%
      pull(time_point) %>%
      table(time_point=.) %>%
      as.data.frame() %>%
      mutate(condition='sle_pe_sga_early')
  
  ) %>%
  spread(time_point,Freq,fill=0)
```

```{r Summarize raw phenotype of GSE108497, eval=FALSE, include=FALSE}
pdata$GSE108497$phenotype %>%
  filter(
    apl==0
    & lac==0
    & fd==0
    & nnd==0
  ) %>%
  mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
  mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
  select(sle,early,pe,sga,time_point) %>%
  table()%>%
  as.data.frame() %>%
  spread(time_point,Freq) %>%
  group_by(sle,early,pe,sga) %>%
  summarize_all(sum) %>%
  mutate(total_preg=`<16 weeks`+`16-23 weeks`+`24-31 weeks`+`32-40 weeks`) %>%
  select(total_preg,everything()) %>%
  ungroup() %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  filter(
    (early==0 & pe==0 & sga==0)
    |
    (early==1 & pe==1 & sga==0)
    |
    (early==0 & pe==1 & sga==0)
    |
    (early==0 & pe==0 & sga==1)
  ) %>%
  arrange(factor(seq,c(1,7,3,2)))
```

```{r Load GSE73685 phenotype, include=FALSE}
# Expression set with the processed transcript expression
pdata$GSE73685$eset=
  suppressWarnings(suppressMessages(
    getGEO(
      filename='data/GSE73685/GSE73685_series_matrix.txt.gz'
      ,getGPL=F
      ,parseCharacteristics=F
    )
  ))
```

```{r Explore raw phenotype of GSE73685, eval=FALSE, include=FALSE}
pdata$GSE73685$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  select(outcome,subject_id) %>%
  filter(!duplicated(.)) %>%
  group_by(outcome) %>%
  summarize(n=n())
```

```{r Summarize raw phenotype of GSE73685, eval=FALSE, include=FALSE}
pdata$GSE73685$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  lapply(
      X=.$tissue_type %>%
        .[!duplicated(.)] %>%
        .[str_to_lower(.)!='maternal blood']
      ,Y=.
      ,function(X,Y){
    Z=group_by(.,subject_id) %>%
      filter(
        str_to_lower(tissue_type)==str_to_lower(X)
        | str_to_lower(tissue_type)=='maternal blood'
      ) %>%
      filter(n()==2) %>%
      ungroup() %>%
      select(subject_id,outcome) %>%
      filter(!duplicated(.)) %>%
      pull(outcome) %>%
      table()
    
    c(tissue=X,Z) %>%
      t() %>%
      as.data.frame() %>%
      gather(outcome,n,-tissue)
  }) %>%
  do.call(rbind,.) %>%
  mutate(sum_n=sum(as.integer(n))) %>%
  group_by(tissue) %>%
  mutate(n_tissue=sum(as.integer(n))) %>%
  spread(outcome,n,fill=0)
```

```{r Load GSE85307 phenotype, include=FALSE}
# Expression set with the processed transcript expression
pdata$GSE85307$eset=
  suppressWarnings(suppressMessages(
    getGEO(
      filename='data/GSE85307/GSE85307-GPL6244_series_matrix.txt.gz'
      ,getGPL=F
      ,parseCharacteristics=F
    )
  ))
```

```{r Summarize raw phenotype of GSE85307, eval=FALSE, include=FALSE}
pdata$GSE85307$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=
      as.numeric(blood_draw_at_gestation_week)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
    ,early=ifelse(gestation_weeks<=34,'Early_Onset','Late_Onset')
    ,trt=
      ifelse(
        `vitamind_25(0h)_ng/ml_whole_blood_baseline`<30
        ,'D<30','D30+'
      )
  ) %>%
  unite(outcome,early,pregnancy_condition,trt) %>%
  filter() %>%
  select(outcome,time_point) %>%
  table() %>%
  as.data.frame() %>%
  spread(time_point,Freq,fill=0) %>%
  mutate(outcome=str_replace_all(outcome,'_',' ')) %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  arrange(factor(seq,c(3,1,5,4,2,6)))
```

```{r Load GSE86200 phenotype, include=FALSE}
# Expression set with the processed transcript expression
pdata$GSE86200$eset=
  suppressWarnings(suppressMessages(
    getGEO(
      filename='data/GSE86200/GSE86200_series_matrix.txt.gz'
      ,getGPL=F
      ,parseCharacteristics=F
    )
  ))
```

```{r Summarize raw phenotype of GSE86200, eval=FALSE, include=FALSE}
pdata$GSE86200$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=
      floor(as.numeric(gestational_day_at_enrollment)/7)
    ,blood_draw_at_gestation_week=
      ifelse(str_detect(title,'32-38'),34,blood_draw_at_gestation_week)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
    ,trt=
      ifelse(
        str_detect(treatment_group,'Intervention')
        ,'Treated','Untreated'
      )
  ) %>%
  unite(outcome,pregnancy_condition,trt) %>%
  select(outcome,time_point) %>%
  table() %>%
  as.data.frame() %>%
  spread(time_point,Freq,fill=0) %>%
  mutate(outcome=str_replace_all(outcome,'_',' ')) %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  arrange(factor(seq,c(2,4,1,3)))
```

```{r Load GSE149437 phenotype, include=FALSE}
# Expression set with the processed transcript expression
pdata$GSE149437$eset=
  suppressWarnings(suppressMessages(
    getGEO(
      filename='data/GSE149437/GSE149437_series_matrix.txt.gz'
      ,getGPL=F
      ,parseCharacteristics=F
    )
  ))
```

```{r Summarize raw phenotype of GSE149437, eval=FALSE, include=FALSE}
pdata$GSE149437$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=as.numeric(gestational_age)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
  ) %>%
  rename(outcome=group) %>%
  select(outcome,time_point) %>%
  table() %>%
  as.data.frame() %>%
  spread(time_point,Freq,fill=0) %>%
  mutate(outcome=str_replace_all(outcome,'_',' ')) %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  arrange(factor(seq,c(1,3,4,2)))
```

```{r Load GSE177477 phenotype, include=FALSE}
# Expression set with the processed transcript expression
pdata$GSE177477$eset=
  suppressWarnings(suppressMessages(
    getGEO(
      filename='data/GSE177477/GSE177477_series_matrix.txt.gz'
      ,getGPL=F
      ,parseCharacteristics=F
    )
  ))

# Add phenotype data from the publication
pdata$GSE177477$eset=
  pdata$GSE177477$eset %>%
  `pData<-`(
    phenoData(.) %>%
      pData() %>%
      mutate(`id:ch1`=sapply(title,function(x)str_split_fixed(x,'\\s+',2)[1])) %>%
      rownames_to_column(var='gsm') %>%
      left_join(
        suppressWarnings(suppressMessages(
          read_csv('data/GSE177477/GSE177477_phenotype.csv')
        ))
        ,by='id:ch1'
      ) %>%
      column_to_rownames(var='gsm')
  ) %>%
  `varMetadata<-`(
    varMetadata(.) %>%
      rownames_to_column(var='variable') %>%
      select(-labelDescription) %>%
      left_join(
        suppressWarnings(suppressMessages(
          read_csv('data/GSE177477/GSE177477_varMetadata.csv')
        ))
        ,by='variable'
      ) %>%
      column_to_rownames(var='variable')
  )
```

```{r Summarize raw phenotype of GSE177477, eval=FALSE, include=FALSE}
pdata$GSE177477$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  separate(
    comorbids
    ,c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
    ,sep='\\|'
  ) %>%
  mutate_at(
    c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
    ,function(x) ifelse(x=='',0,ifelse(is.na(x),0,1))
  ) %>%
  mutate(
    severity=
      factor(
        ifelse(
          is.na(severity)
          ,'uninfected'
          ,severity
        )
        ,c('uninfected','mild','severe')
      )
    ,sex=
      factor(
        ifelse(
          is.na(sex)
          ,'U'
          ,sex
        )
        ,c('U','F','M')
      )
  ) %>%
  rename(outcome=severity) %>%
  select(outcome,symptons,sex) %>%
  table() %>%
  as.data.frame() %>%
  spread(sex,Freq,fill=0) %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  filter(seq%in%c(2,1,6,9)) %>%
  arrange(factor(seq,c(2,1,6,9)))
```

```{r Load GSE192902 phenotype, include=FALSE}
# Expression set with the processed transcript expression
pdata$GSE192902$eset=
  suppressWarnings(suppressMessages(
    getGEO(
      filename='data/GSE192902/GSE192902_series_matrix.txt.gz'
      ,getGPL=F
      ,parseCharacteristics=F
    )
  ))
```

```{r Summarize raw phenotype of GSE192902, eval=FALSE, include=FALSE}
pdata$GSE192902$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    time_point=case_when(
      sampling_time_group=='â‰¤12 weeks gestation'~'12 weeks or less'
      ,sampling_time_group=='13-20 weeks gestation'~'13-20 weeks'
      ,sampling_time_group=='â‰¥23 weeks gestation'~'23 weeks or more'
      ,sampling_time_group=='Post-partum'~'post-partum'
      ,TRUE~''
    )
  ) %>%
  mutate(
    outcome=
      disease %>%
      str_remove_all('[:punct:]')
  ) %>%
  select(outcome,time_point,cohort) %>%
  table() %>%
  as.data.frame() %>%
  spread(time_point,Freq,fill=0) %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,cohort,everything()) %>%
  arrange(factor(seq,c(1,4,7,2,5,8,3,6,9)))
```

```{r Summarize raw phenotype of GSE192902 (no GA), eval=FALSE, include=FALSE}
pdata$GSE192902$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    time_point=case_when(
      sampling_time_group=='â‰¤12 weeks gestation'~'12 weeks or less'
      ,sampling_time_group=='13-20 weeks gestation'~'13-20 weeks'
      ,sampling_time_group=='â‰¥23 weeks gestation'~'23 weeks or more'
      ,sampling_time_group=='Post-partum'~'post-partum'
      ,TRUE~''
    )
  ) %>%
  mutate(
    outcome=
      disease %>%
      str_remove_all('[:punct:]')
  ) %>%
  separate(title,c('subject_id','sampling_time_cat'),sep='_') %>%
  select(cohort,disease,subject_id) %>%
  filter(!duplicated(.)) %>%
  group_by(cohort,disease) %>%
  summarize(n=n()) %>%
  group_by(cohort) %>%
  mutate(total=sum(n)) %>%
  ungroup()
```

```{r Summarize periods of all the datasets, include=FALSE}
data_period=
  rbind(
    
    pdata$GSE108497$eset %>%
      phenoData() %>%
      pData() %>%
      select(submission_date) %>%
      mutate(submission_date=mdy(submission_date)) %>%
      gather() %>%
      mutate(dataset='GSE108497')
    
    ,pdata$GSE73685$eset %>%
      phenoData() %>%
      pData() %>%
      select(submission_date) %>%
      mutate(submission_date=mdy(submission_date)) %>%
      gather() %>%
      mutate(dataset='GSE73685')
    
    ,pdata$GSE85307$eset %>%
      phenoData() %>%
      pData() %>%
      select(submission_date,`microarray_hybridization_date:ch1`) %>%
      mutate(
        submission_date=
          mdy(submission_date)
        ,`microarray_hybridization_date:ch1`=
          mdy(`microarray_hybridization_date:ch1`)
      ) %>%
      gather() %>%
      mutate(dataset='GSE85307')
    
    ,pdata$GSE86200$eset %>%
      phenoData() %>%
      pData() %>%
      select(submission_date,`microarray hybridization date:ch1`) %>%
      mutate(
        submission_date=
          mdy(submission_date)
        ,`microarray hybridization date:ch1`=
          mdy(`microarray hybridization date:ch1`)
      ) %>%
      gather() %>%
      mutate(dataset='GSE86200')
    
    ,pdata$GSE149437$eset %>%
      phenoData() %>%
      pData() %>%
      select(submission_date,`sample_date:ch1`) %>%
      mutate(
        submission_date=
          mdy(submission_date)
        ,`sample_date:ch1`=
          dmy(`sample_date:ch1`)
      ) %>%
      gather() %>%
      mutate(dataset='GSE149437')
    
    ,pdata$GSE177477$eset %>%
      phenoData() %>%
      pData() %>%
      select(submission_date) %>%
      mutate(submission_date=mdy(submission_date)) %>%
      gather() %>%
      mutate(dataset='GSE177477') %>%
      rbind(
        data.frame(
          key='sample_date'
          ,value=ymd(c('2020/04/01','2020/10/31'))
          ,dataset='GSE177477'
        )
      )
    
    ,pdata$GSE192902$eset %>%
      phenoData() %>%
      pData() %>%
      select(submission_date) %>%
      mutate(submission_date=mdy(submission_date)) %>%
      gather() %>%
      mutate(dataset='GSE192902') %>%
      rbind(
        data.frame(
          key='guidelines_date'
          ,value=ymd(c('2020/06/01','2020/06/01'))
          ,dataset='GSE192902'
        )
      )
  ) %>%
  group_by(dataset,key) %>%
  summarize(
    from=min(value)
    ,to=max(value)
  ) %>%
  mutate(
    dataset=
      dataset %>%
      factor(names(pdata) %>% .[c(2,1,3:length(.))] %>% rev())
    ,key=
      key %>%
      str_remove_all('\\:ch1') %>%
      str_replace_all('_',' ') %>%
      str_to_sentence() %>%
      str_remove_all(' date') %>%
      paste0(
        case_when(
          .=='Submission'~'*'
          ,.=='Guidelines'~'†'
          ,TRUE~''
        )
      ) %>%
      factor(.[!duplicated(.)] %>% rev())
    ,dataset_group=
      case_when(
        dataset%in%c('GSE73685')~'1. Derivation dataset'
        ,dataset%in%c('GSE108497')~'2. Development dataset'
        ,dataset%in%c('GSE85307','GSE86200','GSE149437')
         ~'3. Replication datasets'
        ,dataset%in%c('GSE177477')~'4. COVID-19 dataset'
        ,dataset%in%c('GSE192902')
         ~'5. Preeclampsia gene set during the pandemic'
        ,TRUE~''
      )
  )
```

```{r Visualize the periods, echo=FALSE, fig.height=3.46457, fig.width=7.48031}
data_period %>%
  ggplot(aes(dataset,color=key)) +
  geom_errorbar(
    aes(ymin=from,ymax=to)
    ,width=0.5,size=1
    ,na.rm=T
  ) +
  geom_hline(yintercept=ymd('2019-11-17'),lty=2) +
  facet_grid(dataset_group~.,scales='free_y',space='free_y') +
  coord_flip() +
  scale_x_discrete('Dataset') +
  scale_y_date('Date',date_breaks='year') +
  scale_color_manual('',values=c('black','red','brown','purple','blue')) +
  theme(
    axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)
    ,legend.position='top'
    ,strip.text.y=element_text(angle=0,hjust=0,vjust=1)
  )
```

```{r}
timestamper[2]=as.character(now())
```

## Expression data

```{r Load GSE108497 expression, include=FALSE}
# Load the raw data
edata$GSE108497$raw=
  suppressWarnings(suppressMessages(
    read_tsv('data/GSE108497/GSE108497_non-normalized_data.txt')
  )) %>%
  lapply(X=1,Y=.,function(X,Y){
    
    Z=pdata$GSE108497$eset %>%
      phenoData() %>%
      pData() %>%
      select(description) %>%
      rename(key=description) %>%
      rownames_to_column(var='id') %>%
      mutate(id=paste0('sample_',id))
    
    K=Y %>%
      colnames() %>%
      data.frame(key=.) %>%
      left_join(Z,by='key') %>%
      arrange(factor(key,colnames(Y))) %>%
      mutate(key=ifelse(is.na(id),key,id)) %>%
      pull(key)
    
    gather(Y,key,value,-ID_REF) %>%
      left_join(Z,by='key') %>%
      mutate(key=ifelse(is.na(id),key,id)) %>%
      select(-id) %>%
      spread(key,value) %>%
      .[,K]
    
  }) %>%
  .[[1]] %>%
  lapply(X=1,Y=.,function(X,Y){
    Z='data/GSE108497/GSE108497_non-normalized_data_modified.txt'
    write_tsv(Y,Z)
    Z
  }) %>%
  .[[1]] %>%
  read.ilmn(probeid="ID_REF",expr="sample_",verbose=F)

# Background correction without normalization
if(run_heavy_computation){
  edata$GSE108497$unnorm=
    edata$GSE108497$raw %>%
    nec() %>%
    .$E %>%
    ExpressionSet()
  saveRDS(edata$GSE108497$unnorm,'data/GSE108497/GSE108497_unnorm.rds')
}else{
  edata$GSE108497$unnorm=readRDS('data/GSE108497/GSE108497_unnorm.rds')
}

# Background correction with normalization
if(run_heavy_computation){
  edata$GSE108497$norm=
    edata$GSE108497$raw %>%
    neqc() %>%
    .$E %>%
    ExpressionSet()
  saveRDS(edata$GSE108497$norm,'data/GSE108497/GSE108497_norm.rds')
}else{
  edata$GSE108497$norm=readRDS('data/GSE108497/GSE108497_norm.rds')
}
```

```{r Load GSE73685 expression, include=FALSE}
# Load the raw data
edata$GSE73685$raw=
  list.files('data/GSE73685/raw',full.names=T) %>%
  read.celfiles(sampleNames=colnames(pdata$GSE73685$eset),verbose=F)

# Background correction without normalization
if(run_heavy_computation){
  edata$GSE73685$unnorm=
    edata$GSE73685$raw %>%
    rma(background=T,normalize=F)
  saveRDS(edata$GSE73685$unnorm,'data/GSE73685/GSE73685_unnorm.rds')
}else{
  edata$GSE73685$unnorm=readRDS('data/GSE73685/GSE73685_unnorm.rds')
}

# Background correction with normalization
if(run_heavy_computation){
  edata$GSE73685$norm=
    edata$GSE73685$raw %>%
    rma(background=T,normalize=T)
  saveRDS(edata$GSE73685$norm,'data/GSE73685/GSE73685_norm.rds')
}else{
  edata$GSE73685$norm=readRDS('data/GSE73685/GSE73685_norm.rds')
}
```

```{r Load GSE85307 expression, include=FALSE}
# Load the raw data
edata$GSE85307$raw=
  list.files('data/GSE85307/raw',full.names=T) %>%
  read.celfiles(sampleNames=colnames(pdata$GSE85307$eset),verbose=F)

# Background correction without normalization
if(run_heavy_computation){
  edata$GSE85307$unnorm=
    edata$GSE85307$raw %>%
    rma(background=T,normalize=F)
  saveRDS(edata$GSE85307$unnorm,'data/GSE85307/GSE85307_unnorm.rds')
}else{
  edata$GSE85307$unnorm=readRDS('data/GSE85307/GSE85307_unnorm.rds')
}

# Background correction with normalization
if(run_heavy_computation){
  edata$GSE85307$norm=
    edata$GSE85307$raw %>%
    rma(background=T,normalize=T)
  saveRDS(edata$GSE85307$norm,'data/GSE85307/GSE85307_norm.rds')
}else{
  edata$GSE85307$norm=readRDS('data/GSE85307/GSE85307_norm.rds')
}
```

```{r Load GSE86200 expression, include=FALSE}
# Load the raw data
edata$GSE86200$raw=
  list.files('data/GSE86200/raw',full.names=T) %>%
  read.idat(bgxfile='data/GSE86200/HumanHT-12_V4_0_R2_15002873_B.bgx') %>%
  lapply(X=1,Y=.,function(X,Y){
    recol=function(x){
      x %>%
        `colnames<-`(
          colnames(.) %>%
            str_remove_all('data/GSE86200/raw/') %>%
            str_split_fixed('_',2) %>%
            .[,1]
        ) 

    }
    Y$E=recol(Y$E)
    Y$other$NumBeads=recol(Y$other$NumBeads)
    Y$other$STDEV=recol(Y$other$STDEV)
    Y
  }) %>%
  .[[1]]

# Background correction without normalization
if(run_heavy_computation){
  edata$GSE86200$unnorm=
    edata$GSE86200$raw %>%
    nec() %>%
    .$E %>%
    .[rownames(.) %>% .[!duplicated(.)],] %>%
    as.data.frame() %>%
    rownames_to_column(var='Array_Address_Id') %>%
    mutate(seq=seq(nrow(.))) %>%
    left_join(
      edata$GSE86200$raw$genes %>%
        select(Array_Address_Id,Probe_Id) %>%
        filter(!duplicated(.)) %>%
        mutate(Array_Address_Id=as.character(Array_Address_Id))
      ,by='Array_Address_Id'
    ) %>%
    arrange(seq) %>%
    select(-seq,-Array_Address_Id) %>%
    select(Probe_Id,everything()) %>%
    column_to_rownames(var='Probe_Id') %>%
    as.matrix() %>%
    ExpressionSet()
  saveRDS(edata$GSE86200$unnorm,'data/GSE86200/GSE86200_unnorm.rds')
}else{
  edata$GSE86200$unnorm=readRDS('data/GSE86200/GSE86200_unnorm.rds')
}

# Background correction with normalization
if(run_heavy_computation){
  edata$GSE86200$norm=
    edata$GSE86200$raw %>%
    neqc() %>%
    .$E %>%
    .[rownames(.) %>% .[!duplicated(.)],] %>%
    as.data.frame() %>%
    rownames_to_column(var='Array_Address_Id') %>%
    mutate(seq=seq(nrow(.))) %>%
    left_join(
      edata$GSE86200$raw$genes %>%
        select(Array_Address_Id,Probe_Id) %>%
        filter(!duplicated(.)) %>%
        mutate(Array_Address_Id=as.character(Array_Address_Id))
      ,by='Array_Address_Id'
    ) %>%
    arrange(seq) %>%
    select(-seq,-Array_Address_Id) %>%
    select(Probe_Id,everything()) %>%
    column_to_rownames(var='Probe_Id') %>%
    as.matrix() %>%
    ExpressionSet()
  saveRDS(edata$GSE86200$norm,'data/GSE86200/GSE86200_norm.rds')
}else{
  edata$GSE86200$norm=readRDS('data/GSE86200/GSE86200_norm.rds')
}
```

```{r Create calling list to load GSE149437 expression, include=FALSE}
# Load the raw data
suppressWarnings(set.seed(33,sample.kind=sample.kind))
GSE149437_calling_list=
  pdata$GSE149437$eset %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=as.numeric(gestational_age)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
  ) %>%
  rename(outcome=group) %>%
  select(title,outcome,time_point) %>%
  rownames_to_column(var='id') %>%
  group_by(outcome,time_point) %>%
  mutate(subset=sample(rep(1:10,ceiling(n()/10))[1:n()],n(),F)) %>%
  ungroup() %>%
  unite(filename,id,title,sep='_') %>%
  mutate(filename=paste0('data/GSE149437/raw/',filename,'.CEL.gz')) %>%
  lapply(X=1:10,Y=.,function(X,Y){
    Y %>%
      filter(subset==X) %>%
      pull(filename)
  })
```

```{r Load GSE149437 expression, include=FALSE}
# Load the raw data
edata$GSE149437$raw=
  list.files('data/GSE149437/raw',full.names=T) %>%
  read.celfiles(sampleNames=colnames(pdata$GSE149437$eset),verbose=F)

gc()

if(run_heavy_computation){
  # Chunk the raw data
  edata$GSE149437$raw=
    GSE149437_calling_list %>%
    pblapply(X=seq(length(.)),Y=.,Z=edata$GSE149437$raw,function(X,Y,Z){
      K=Z %>%
        .[,Y[[X]] %>%
            str_remove_all('data\\/GSE149437\\/raw/|\\.CEL\\.gz') %>%
            str_split_fixed('_',2) %>%
            .[,1]]
  
      K %>%
        rma(background=T,normalize=F) %>%
        saveRDS(paste0(
          'data/GSE149437/GSE149437_unnorm_'
          ,str_pad(X,2,'left','0')
          ,'.rds'
        ))
  
      K %>%
        rma(background=T,normalize=T) %>%
        saveRDS(paste0(
          'data/GSE149437/GSE149437_norm_'
          ,str_pad(X,2,'left','0')
          ,'.rds'
        ))
  
      if(X==1){
        Z
      }else{
        NULL
      }
    }) %>%
    .[[1]]
}else{
  # Background correction without normalization (per chunk)
  edata$GSE149437$unnorm=
    seq(length(GSE149437_calling_list)) %>%
    pblapply(function(x){
      readRDS(paste0(
        'data/GSE149437/GSE149437_unnorm_'
        ,str_pad(x,2,'left','0')
        ,'.rds'
      ))
    })
  
  # Background correction with normalization (per chunk)
  edata$GSE149437$norm=
    seq(length(GSE149437_calling_list)) %>%
    pblapply(function(x){
      readRDS(paste0(
        'data/GSE149437/GSE149437_norm_'
        ,str_pad(x,2,'left','0')
        ,'.rds'
      ))
    })
  
  # Background correcting
  # Calculating Expression
  # Background correcting
  # Normalizing
  # Calculating Expression
  #   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=57m 33s
  #   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=03s  
  #   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=03s
  
  # Background correcting
  # Calculating Expression
  # Background correcting
  # Normalizing
  # Calculating Expression
  #   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=01h 05m 53s
  #   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=03s  
  #   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=03s
  
  # Background correction without normalization (combining chunks)
  edata$GSE149437$unnorm=
    edata$GSE149437$unnorm %>%
    lapply(X=1,Y=.,function(X,Y){
      for(i in seq(length(Y))){
        if(i==1){
          Z=Y[[i]]
        }else{
          Z=a4Base::combineTwoExpressionSet(Z,Y[[i]])
        }
      }
      Z
    }) %>%
    .[[1]] %>%
    .[,colnames(pdata$GSE149437$eset)]
  
  # Background correction with normalization (combining chunks)
  edata$GSE149437$norm=
    edata$GSE149437$norm %>%
    lapply(X=1,Y=.,function(X,Y){
      for(i in seq(length(Y))){
        if(i==1){
          Z=Y[[i]]
        }else{
          Z=a4Base::combineTwoExpressionSet(Z,Y[[i]])
        }
      }
      Z
    }) %>%
    .[[1]] %>%
    .[,colnames(pdata$GSE149437$eset)]
}
```

```{r Load GSE177477 expression, include=FALSE}
# Load the raw data
edata$GSE177477$raw=
  list.files('data/GSE177477/raw',full.names=T) %>%
  .[!str_detect(.,'chp')] %>%
  read.celfiles(sampleNames=colnames(pdata$GSE177477$eset),verbose=F)

# Background correction without normalization
if(run_heavy_computation){
  edata$GSE177477$unnorm=
    edata$GSE177477$raw %>%
    rma(background=T,normalize=F)
  saveRDS(edata$GSE177477$unnorm,'data/GSE177477/GSE177477_unnorm.rds')
}else{
  edata$GSE177477$unnorm=readRDS('data/GSE177477/GSE177477_unnorm.rds')
}

# Background correction with normalization
if(run_heavy_computation){
  edata$GSE177477$norm=
    edata$GSE177477$raw %>%
    rma(background=T,normalize=T)
  saveRDS(edata$GSE177477$norm,'data/GSE177477/GSE177477_norm.rds')
}else{
  edata$GSE177477$norm=readRDS('data/GSE177477/GSE177477_norm.rds')
}
```

```{r}
timestamper[3]=as.character(now())
```

## Feature data

```{r Download annotation database, eval=FALSE, include=FALSE}
# mart=useMart("ensembl")
# saveRDS(mart,file='data/mart.rds')
# ensembl=useDataset("hsapiens_gene_ensembl",mart)
# saveRDS(ensembl,file='data/ensembl.rds')
```

```{r Load the saved database, include=FALSE}
mart=readRDS(file='data/mart.rds') # Downloaded on 2021-07-27
ensembl=readRDS(file='data/ensembl.rds') # Downloaded on 2021-07-27
listAttributes(ensembl)
```

```{r Download annotation data, include=FALSE}
source('R/download_annotation-function.R')

# GPL10558, downloaded on 2021-07-27
# GPL10558=
#   download_annotation(
#     the_platform='illumina_humanht_12_v4',
#     the_mart=ensembl
#   )
# saveRDS(GPL10558,'data/GPL10558.rds')
GPL10558=readRDS('data/GPL10558.rds')

# GPL6244, downloaded on 2021-07-27
# GPL6244=
#   download_annotation(
#     the_platform='affy_hugene_1_0_st_v1',
#     the_mart=ensembl
#   )
# saveRDS(GPL6244,'data/GPL6244.rds')
GPL6244=readRDS('data/GPL6244.rds')

# GPL28460, download on 2021-09-12
# GPL28460=
#   download_annotation(
#     the_platform='affy_hta_2_0',
#     the_mart=ensembl
#   )
# saveRDS(GPL28460,'data/GPL28460.rds')
GPL28460=
  readRDS('data/GPL28460.rds') %>%
  rownames_to_column(var='id') %>%
  mutate(id=paste0(id,'.1')) %>%
  column_to_rownames(var='id')

# GPL23159, download on 2022-02-11
# GPL23159=list()
# GPL23159$refseq_mrna=
#   download_annotation(
#     the_platform='refseq_mrna',
#     the_mart=ensembl
#   )
# GPL23159$ensembl_transcript_id=
#   download_annotation(
#     the_platform='ensembl_transcript_id',
#     the_mart=ensembl
#   )
# GPL23159$soft_file=
#   suppressWarnings(suppressMessages(
#     read_tsv('data/GSE177477/GSE177477_family.soft.gz',skip=137) %>%
#       setNames(
#         colnames(.) %>%
#           str_remove_all('\\s+|[:punct:]') %>%
#           str_to_lower()
#       )
#   )) %>%
#   filter(
#     probesetid
#     %in% intersect(
#       .$probesetid
#       ,rownames(edata$GSE177477$norm)
#     )
#   ) %>%
#   right_join(
#     lapply(X=1,Y=.,function(X,Y){
#         Y$spotid10 %>%
#           pblapply(function(x)str_split(x,' /// ')[[1]]) %>%
#           pblapply(function(x)x[str_count(x,' // ')==8]) %>%
#           pblapply(sapply,function(x)str_split(x,' // ')[[1]]) %>%
#           pblapply(`dimnames<-`,NULL) %>%
#           pblapply(t) %>%
#           pblapply(as.data.frame) %>%
#           pblapply(X=seq(length(.)),Y=Y$probesetid,Z=.,function(X,Y,Z){
#             Z[[X]] %>%
#               mutate(probesetid=Y[X]) %>%
#               select(probesetid,everything())
#           }) %>%
#           do.call(rbind,.)
#       }) %>%
#       .[[1]]
#     ,by='probesetid'
#   )
# GPL23159=
#   GPL23159$soft_file %>%
#   left_join(
#     GPL23159$refseq_mrna %>%
#       rownames_to_column(var='V1') %>%
#       mutate(V2='RefSeq') %>%
#       rbind(
#         GPL23159$ensembl_transcript_id %>%
#           rownames_to_column(var='V1') %>%
#           mutate(V2='ENSEMBL')
#       )
#     ,by=c('V1','V2')
#   ) %>%
#   select(probesetid,ensembl_gene_id,entrezgene_id,hgnc_symbol) %>%
#   filter(!duplicated(.)) %>%
#   gather(annotation,value,-probesetid) %>%
#   filter(!is.na(value)) %>%
#   group_by(probesetid,annotation) %>%
#   summarise(value=paste0(value,collapse='///')) %>%
#   spread(annotation,value) %>%
#   right_join(
#     edata$GSE177477$norm %>%
#       rownames() %>%
#       data.frame(probesetid=.)
#     ,by='probesetid'
#   ) %>%
#   filter(
#     !(is.na(ensembl_gene_id)) &
#       !(is.na(entrezgene_id)) &
#       !(is.na(hgnc_symbol))
#   ) %>%
#   filter(
#     !(str_detect(ensembl_gene_id,'///')) &
#       !(str_detect(entrezgene_id,'///')) &
#       !(str_detect(hgnc_symbol,'///'))
#   ) %>%
#   column_to_rownames(var='probesetid')
# saveRDS(GPL23159,'data/GPL23159.rds')
GPL23159=readRDS('data/GPL23159.rds')
```

```{r Take common transcripts of the dataset and the annotation, include=FALSE}
fdata$GSE108497=
  GPL10558 %>%
  .[intersect(rownames(edata$GSE108497$norm),rownames(.)),]

fdata$GSE73685=
  GPL6244 %>%
  .[intersect(rownames(edata$GSE73685$norm),rownames(.)),]

fdata$GSE85307=
  GPL6244 %>%
  .[intersect(rownames(edata$GSE85307$norm),rownames(.)),]

fdata$GSE86200=
  GPL10558 %>%
  .[intersect(rownames(edata$GSE86200$norm),rownames(.)),]

fdata$GSE149437=
  GPL28460 %>%
  .[intersect(rownames(edata$GSE149437$norm),rownames(.)),]

fdata$GSE177477=
  GPL23159 %>%
  .[intersect(rownames(edata$GSE177477$norm),rownames(.)),]
```

```{r Check availability of FLT1, eval=FALSE, include=FALSE}
list(GSE108497=fdata$GSE108497$hgnc_symbol
     ,GSE73685=fdata$GSE73685$hgnc_symbol
     ,GSE85307=fdata$GSE85307$hgnc_symbol
     ,GSE86200=fdata$GSE86200$hgnc_symbol
     ,GSE149437=fdata$GSE149437$hgnc_symbol
     ,GSE177477=fdata$GSE177477$hgnc_symbol
     ) %>%
  lapply(
    intersect
    ,c('FLT1')
  )
```

```{r}
timestamper[4]=as.character(now())
```

# Data preprocessing

## Removal of technical outliers

```{r Identify outliers of unnormalized expression by RLE, include=FALSE}
source('R/suspected_outliers-function.R')

rdata$GSE108497=
  edata$GSE108497$unnorm %>%
  .[intersect(rownames(GPL10558),rownames(.)),] %>%
  suspected_outliers()

rdata$GSE73685=
  edata$GSE73685$unnorm %>%
  .[intersect(rownames(GPL6244),rownames(.)),] %>%
  suspected_outliers()

rdata$GSE85307=
  edata$GSE85307$unnorm %>%
  .[intersect(rownames(GPL6244),rownames(.)),] %>%
  suspected_outliers()

rdata$GSE86200=
  edata$GSE86200$unnorm %>%
  .[intersect(rownames(GPL10558),rownames(.)),] %>%
  suspected_outliers()

rdata$GSE149437=
  edata$GSE149437$unnorm %>%
  .[intersect(rownames(GPL28460),rownames(.)),] %>%
  suspected_outliers()

rdata$GSE177477=
  edata$GSE177477$unnorm %>%
  .[intersect(rownames(GPL23159),rownames(.)),] %>%
  suspected_outliers()
```

```{r Identify outliers of normalized expression by HClust, include=FALSE}
cdata$GSE108497=
  edata$GSE108497$norm %>%
  .[intersect(rownames(GPL10558),rownames(.)),] %>%
  exprs() %>%
  t() %>%
  parallelDist(method='manhattan') %>%
  hclust(method='complete')

cdata$GSE73685=
  edata$GSE73685$norm %>%
  .[intersect(rownames(GPL6244),rownames(.)),] %>%
  exprs() %>%
  t() %>%
  parallelDist(method='manhattan') %>%
  hclust(method='complete')

cdata$GSE85307=
  edata$GSE85307$norm %>%
  .[intersect(rownames(GPL6244),rownames(.)),] %>%
  exprs() %>%
  t() %>%
  parallelDist(method='manhattan') %>%
  hclust(method='complete')

cdata$GSE86200=
  edata$GSE86200$norm %>%
  .[intersect(rownames(GPL10558),rownames(.)),] %>%
  exprs() %>%
  t() %>%
  parallelDist(method='manhattan') %>%
  hclust(method='complete')

cdata$GSE149437=
  edata$GSE149437$norm %>%
  .[intersect(rownames(GPL28460),rownames(.)),] %>%
  exprs() %>%
  t() %>%
  parallelDist(method='manhattan') %>%
  hclust(method='complete')

cdata$GSE177477=
  edata$GSE177477$norm %>%
  .[intersect(rownames(GPL23159),rownames(.)),] %>%
  exprs() %>%
  t() %>%
  parallelDist(method='manhattan') %>%
  hclust(method='complete')
```

```{r HClust outliers, eval=FALSE, fig.height=10, fig.width=20, include=FALSE}
plot(cdata$GSE108497,cex=0.6); rect.hclust(cdata$GSE108497,k=2,border=2:3)
cutree(cdata$GSE108497,k=2) %>% table() # 2

plot(cdata$GSE73685,cex=0.6); rect.hclust(cdata$GSE73685,k=3,border=2:4)
cutree(cdata$GSE73685,k=3) %>% table() # 3

plot(cdata$GSE85307,cex=0.6); rect.hclust(cdata$GSE85307,k=3,border=2:5)
cutree(cdata$GSE85307,k=3) %>% table() # 3

plot(cdata$GSE86200,cex=0.6); rect.hclust(cdata$GSE86200,k=4,border=2:5)
cutree(cdata$GSE86200,k=4) %>% table() # 4

plot(cdata$GSE149437,cex=0.6); rect.hclust(cdata$GSE149437,k=3,border=2:4)
cutree(cdata$GSE149437,k=3) %>% table() # 3

plot(cdata$GSE177477,cex=0.6); rect.hclust(cdata$GSE177477,k=2,border=2:4)
cutree(cdata$GSE177477,k=2) %>% table() # 2
```

```{r Identify the outliers by both RLE and HClust, include=FALSE}
odata$GSE108497=
  cutree(cdata$GSE108497,k=2) %>%
  .[. %in% 2] %>%
  names() %>%
  intersect(
    rdata$GSE108497 %>%
      filter(suspect_outlier=='yes') %>%
      .$sample_id %>%
      .[!duplicated(.)]
  )

odata$GSE73685=
  cutree(cdata$GSE73685,k=3) %>%
  .[. %in% 3] %>%
  names() %>%
  intersect(
    rdata$GSE73685 %>%
      filter(suspect_outlier=='yes') %>%
      .$sample_id %>%
      .[!duplicated(.)]
  )

odata$GSE85307=
  cutree(cdata$GSE85307,k=3) %>%
  .[. %in% 3] %>%
  names() %>%
  intersect(
    rdata$GSE85307 %>%
      filter(suspect_outlier=='yes') %>%
      .$sample_id %>%
      .[!duplicated(.)]
  )

odata$GSE86200=
  cutree(cdata$GSE86200,k=4) %>%
  .[. %in% 4] %>%
  names() %>%
  intersect(
    rdata$GSE86200 %>%
      filter(suspect_outlier=='yes') %>%
      .$sample_id %>%
      .[!duplicated(.)]
  )

odata$GSE149437=
  cutree(cdata$GSE149437,k=3) %>%
  .[. %in% 3] %>%
  names() %>%
  intersect(
    rdata$GSE149437 %>%
      filter(suspect_outlier=='yes') %>%
      .$sample_id %>%
      .[!duplicated(.)]
  )

odata$GSE177477=
  cutree(cdata$GSE177477,k=2) %>%
  .[. %in% 2] %>%
  names() %>%
  intersect(
    rdata$GSE177477 %>%
      filter(suspect_outlier=='yes') %>%
      .$sample_id %>%
      .[!duplicated(.)]
  )
```

## Removal of low-expressed probe sets

```{r Determine transcript-wise medians of expression, include=FALSE}
ldata$GSE108497$row_median=
  edata$GSE108497$norm %>%
  .[rownames(.) %>% intersect(rownames(GPL10558))
    ,colnames(.) %>% .[!.%in%odata$GSE108497]] %>%
  exprs() %>%
  rowMedians() %>%
  setNames(intersect(rownames(GPL10558),rownames(edata$GSE108497$norm)))

ldata$GSE73685$row_median=
  edata$GSE73685$norm %>%
  .[rownames(.) %>% intersect(rownames(GPL6244))
    ,colnames(.) %>% .[!.%in%odata$GSE73685]] %>%
  exprs() %>%
  rowMedians() %>%
  setNames(intersect(rownames(GPL6244),rownames(edata$GSE73685$norm)))

ldata$GSE85307$row_median=
  edata$GSE85307$norm %>%
  .[rownames(.) %>% intersect(rownames(GPL6244))
    ,colnames(.) %>% .[!.%in%odata$GSE85307]] %>%
  exprs() %>%
  rowMedians() %>%
  setNames(intersect(rownames(GPL6244),rownames(edata$GSE85307$norm)))

ldata$GSE86200$row_median=
  edata$GSE86200$norm %>%
  .[rownames(.) %>% intersect(rownames(GPL10558))
    ,colnames(.) %>% .[!.%in%odata$GSE86200]] %>%
  exprs() %>%
  rowMedians() %>%
  setNames(intersect(rownames(GPL10558),rownames(edata$GSE86200$norm)))

ldata$GSE149437$row_median=
  edata$GSE149437$norm %>%
  .[rownames(.) %>% intersect(rownames(GPL28460))
    ,colnames(.) %>% .[!.%in%odata$GSE149437]] %>%
  exprs() %>%
  rowMedians() %>%
  setNames(intersect(rownames(GPL28460),rownames(edata$GSE149437$norm)))

ldata$GSE177477$row_median=
  edata$GSE177477$norm %>%
  .[rownames(.) %>% intersect(rownames(GPL23159))
    ,colnames(.) %>% .[!.%in%odata$GSE177477]] %>%
  exprs() %>%
  rowMedians() %>%
  setNames(intersect(rownames(GPL23159),rownames(edata$GSE177477$norm)))
```

```{r Distribution of the transcript-wise medians, eval=FALSE, include=FALSE}
hist(ldata$GSE108497$row_median,100)
abline(v=quantile(ldata$GSE108497$row_median,0.2),col=2,lwd=3)

hist(ldata$GSE73685$row_median,100)
abline(v=quantile(ldata$GSE73685$row_median,0.2),col=2,lwd=3)

hist(ldata$GSE85307$row_median,100)
abline(v=quantile(ldata$GSE85307$row_median,0.2),col=2,lwd=3)

hist(ldata$GSE86200$row_median,100)
abline(v=quantile(ldata$GSE86200$row_median,0.2),col=2,lwd=3)

hist(ldata$GSE149437$row_median,100)
abline(v=quantile(ldata$GSE149437$row_median,0.2),col=2,lwd=3)

hist(ldata$GSE177477$row_median,100)
abline(v=quantile(ldata$GSE177477$row_median,0.2),col=2,lwd=3)
```

```{r Identify the low-expressed transcripts, include=FALSE}
ldata$GSE108497$probe=
  ldata$GSE108497$row_median %>%
  .[. <= quantile(.,0.2)] %>%
  names()

ldata$GSE73685$probe=
  ldata$GSE73685$row_median %>%
  .[. <= quantile(.,0.2)] %>%
  names()

ldata$GSE85307$probe=
  ldata$GSE85307$row_median %>%
  .[. <= quantile(.,0.2)] %>%
  names()

ldata$GSE86200$probe=
  ldata$GSE86200$row_median %>%
  .[. <= quantile(.,0.2)] %>%
  names()

ldata$GSE149437$probe=
  ldata$GSE149437$row_median %>%
  .[. <= quantile(.,0.2)] %>%
  names()

ldata$GSE177477$probe=
  ldata$GSE177477$row_median %>%
  .[. <= quantile(.,0.2)] %>%
  names()
```

```{r}
timestamper[5]=as.character(now())
```

## Summarization from probe sets to genes

```{r Summarize to genes but no outliers and low-exp transcripts, include=FALSE}
# GSE108497
gdata$GSE108497$filtered=
  edata$GSE108497$norm %>%
  .[rownames(.) %>%
      .[.%in%intersect(.,rownames(GPL10558))
        & !.%in%ldata$GSE108497$probe]
    ,colnames(.) %>% .[!.%in%odata$GSE108497]]

gdata$GSE108497$hgnc=
  collapseRows(
    datET=exprs(gdata$GSE108497$filtered)
    ,rowID=rownames(gdata$GSE108497$filtered)
    ,rowGroup=
      fdata$GSE108497 %>%
      .[rownames(gdata$GSE108497$filtered),'hgnc_symbol',drop=F] %>%
      arrange(factor(hgnc_symbol,rownames(gdata$GSE108497$filtered))) %>%
      pull(hgnc_symbol)
  )

# GSE73685
gdata$GSE73685$filtered=
  edata$GSE73685$norm %>%
  .[rownames(.) %>%
      .[.%in%intersect(.,rownames(GPL6244))
        & !.%in%ldata$GSE73685$probe]
    ,colnames(.) %>% .[!.%in%odata$GSE73685]]

gdata$GSE73685$hgnc=
  collapseRows(
    datET=exprs(gdata$GSE73685$filtered)
    ,rowID=rownames(gdata$GSE73685$filtered)
    ,rowGroup=
      fdata$GSE73685 %>%
      .[rownames(gdata$GSE73685$filtered),'hgnc_symbol',drop=F] %>%
      arrange(factor(hgnc_symbol,rownames(gdata$GSE73685$filtered))) %>%
      pull(hgnc_symbol)
  )

# GSE85307
gdata$GSE85307$filtered=
  edata$GSE85307$norm %>%
  .[rownames(.) %>%
      .[.%in%intersect(.,rownames(GPL6244))
        & !.%in%ldata$GSE85307$probe]
    ,colnames(.) %>% .[!.%in%odata$GSE85307]]

gdata$GSE85307$hgnc=
  collapseRows(
    datET=exprs(gdata$GSE85307$filtered)
    ,rowID=rownames(gdata$GSE85307$filtered)
    ,rowGroup=
      fdata$GSE85307 %>%
      .[rownames(gdata$GSE85307$filtered),'hgnc_symbol',drop=F] %>%
      arrange(factor(hgnc_symbol,rownames(gdata$GSE85307$filtered))) %>%
      pull(hgnc_symbol)
  )

# GSE86200
gdata$GSE86200$filtered=
  edata$GSE86200$norm %>%
  .[rownames(.) %>%
      .[.%in%intersect(.,rownames(GPL10558))
        & !.%in%ldata$GSE86200$probe]
    ,colnames(.) %>% .[!.%in%odata$GSE86200]]

gdata$GSE86200$hgnc=
  collapseRows(
    datET=exprs(gdata$GSE86200$filtered)
    ,rowID=rownames(gdata$GSE86200$filtered)
    ,rowGroup=
      fdata$GSE86200 %>%
      .[rownames(gdata$GSE86200$filtered),'hgnc_symbol',drop=F] %>%
      arrange(factor(hgnc_symbol,rownames(gdata$GSE86200$filtered))) %>%
      pull(hgnc_symbol)
  )

# GSE149437
gdata$GSE149437$filtered=
  edata$GSE149437$norm %>%
  .[rownames(.) %>%
      .[.%in%intersect(.,rownames(GPL28460))
        & !.%in%ldata$GSE149437$probe]
    ,colnames(.) %>% .[!.%in%odata$GSE149437]]

gdata$GSE149437$hgnc=
  collapseRows(
    datET=exprs(gdata$GSE149437$filtered)
    ,rowID=rownames(gdata$GSE149437$filtered)
    ,rowGroup=
      fdata$GSE149437 %>%
      .[rownames(gdata$GSE149437$filtered),'hgnc_symbol',drop=F] %>%
      arrange(factor(hgnc_symbol,rownames(gdata$GSE149437$filtered))) %>%
      pull(hgnc_symbol)
  )

# GSE177477
gdata$GSE177477$filtered=
  edata$GSE177477$norm %>%
  .[rownames(.) %>%
      .[.%in%intersect(.,rownames(GPL23159))
        & !.%in%ldata$GSE177477$probe]
    ,colnames(.) %>% .[!.%in%odata$GSE177477]]

gdata$GSE177477$hgnc=
  collapseRows(
    datET=exprs(gdata$GSE177477$filtered)
    ,rowID=rownames(gdata$GSE177477$filtered)
    ,rowGroup=
      fdata$GSE177477 %>%
      .[rownames(gdata$GSE177477$filtered),'hgnc_symbol',drop=F] %>%
      arrange(factor(hgnc_symbol,rownames(gdata$GSE177477$filtered))) %>%
      pull(hgnc_symbol)
  )
```

```{r}
timestamper[6]=as.character(now())
```

# Selecting the common genes among all the microarray platforms

```{r Compile into new expression sets, include=FALSE}
# Compile each of the expression sets
GSE108497=
  ExpressionSet(
    assayData=
      gdata$GSE108497$hgnc$datETcollapsed
    ,phenoData=
      pdata$GSE108497$eset %>%
      .[,colnames(gdata$GSE108497$hgnc$datETcollapsed)] %>%
      phenoData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,featureData=
      gdata$GSE108497$hgnc$group2row %>%
      as.data.frame() %>%
      setNames(c('hgnc_symbol','probe_id')) %>%
      left_join(
        fdata$GSE108497 %>%
          rownames_to_column(var='probe_id')
        ,by=c('hgnc_symbol','probe_id')
      ) %>%
      column_to_rownames(var='hgnc_symbol') %>% 
      AnnotatedDataFrame()
    ,protocolData=
      pdata$GSE108497$eset %>%
      .[,colnames(gdata$GSE108497$hgnc$datETcollapsed)] %>%
      protocolData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,experimentData=experimentData(pdata$GSE108497$eset)
    ,annotation=annotation(pdata$GSE108497$eset)
  )

GSE73685=
  ExpressionSet(
    assayData=
      gdata$GSE73685$hgnc$datETcollapsed
    ,phenoData=
      pdata$GSE73685$eset %>%
      .[,colnames(gdata$GSE73685$hgnc$datETcollapsed)] %>%
      phenoData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,featureData=
      gdata$GSE73685$hgnc$group2row %>%
      as.data.frame() %>%
      setNames(c('hgnc_symbol','probe_id')) %>%
      left_join(
        fdata$GSE73685 %>%
          rownames_to_column(var='probe_id')
        ,by=c('hgnc_symbol','probe_id')
      ) %>%
      column_to_rownames(var='hgnc_symbol') %>% 
      AnnotatedDataFrame()
    ,protocolData=
      pdata$GSE73685$eset %>%
      .[,colnames(gdata$GSE73685$hgnc$datETcollapsed)] %>%
      protocolData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,experimentData=experimentData(pdata$GSE73685$eset)
    ,annotation=annotation(pdata$GSE73685$eset)
  )

GSE85307=
  ExpressionSet(
    assayData=
      gdata$GSE85307$hgnc$datETcollapsed
    ,phenoData=
      pdata$GSE85307$eset %>%
      .[,colnames(gdata$GSE85307$hgnc$datETcollapsed)] %>%
      phenoData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,featureData=
      gdata$GSE85307$hgnc$group2row %>%
      as.data.frame() %>%
      setNames(c('hgnc_symbol','probe_id')) %>%
      left_join(
        fdata$GSE85307 %>%
          rownames_to_column(var='probe_id')
        ,by=c('hgnc_symbol','probe_id')
      ) %>%
      column_to_rownames(var='hgnc_symbol') %>% 
      AnnotatedDataFrame()
    ,protocolData=
      pdata$GSE85307$eset %>%
      .[,colnames(gdata$GSE85307$hgnc$datETcollapsed)] %>%
      protocolData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,experimentData=experimentData(pdata$GSE85307$eset)
    ,annotation=annotation(pdata$GSE85307$eset)
  )

GSE86200=
  ExpressionSet(
    assayData=
      gdata$GSE86200$hgnc$datETcollapsed
    ,phenoData=
      pdata$GSE86200$eset %>%
      .[,colnames(gdata$GSE86200$hgnc$datETcollapsed)] %>%
      phenoData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,featureData=
      gdata$GSE86200$hgnc$group2row %>%
      as.data.frame() %>%
      setNames(c('hgnc_symbol','probe_id')) %>%
      left_join(
        fdata$GSE86200 %>%
          rownames_to_column(var='probe_id')
        ,by=c('hgnc_symbol','probe_id')
      ) %>%
      column_to_rownames(var='hgnc_symbol') %>% 
      AnnotatedDataFrame()
    ,protocolData=
      pdata$GSE86200$eset %>%
      .[,colnames(gdata$GSE86200$hgnc$datETcollapsed)] %>%
      protocolData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,experimentData=experimentData(pdata$GSE86200$eset)
    ,annotation=annotation(pdata$GSE86200$eset)
  )

GSE149437=
  ExpressionSet(
    assayData=
      gdata$GSE149437$hgnc$datETcollapsed
    ,phenoData=
      pdata$GSE149437$eset %>%
      .[,colnames(gdata$GSE149437$hgnc$datETcollapsed)] %>%
      phenoData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,featureData=
      gdata$GSE149437$hgnc$group2row %>%
      as.data.frame() %>%
      setNames(c('hgnc_symbol','probe_id')) %>%
      left_join(
        fdata$GSE149437 %>%
          rownames_to_column(var='probe_id')
        ,by=c('hgnc_symbol','probe_id')
      ) %>%
      column_to_rownames(var='hgnc_symbol') %>% 
      AnnotatedDataFrame()
    ,protocolData=
      pdata$GSE149437$eset %>%
      .[,colnames(gdata$GSE149437$hgnc$datETcollapsed)] %>%
      protocolData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,experimentData=experimentData(pdata$GSE149437$eset)
    ,annotation=annotation(pdata$GSE149437$eset)
  )

GSE177477=
  ExpressionSet(
    assayData=
      gdata$GSE177477$hgnc$datETcollapsed
    ,phenoData=
      pdata$GSE177477$eset %>%
      .[,colnames(gdata$GSE177477$hgnc$datETcollapsed)] %>%
      phenoData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,featureData=
      gdata$GSE177477$hgnc$group2row %>%
      as.data.frame() %>%
      setNames(c('hgnc_symbol','probe_id')) %>%
      left_join(
        fdata$GSE177477 %>%
          rownames_to_column(var='probe_id')
        ,by=c('hgnc_symbol','probe_id')
      ) %>%
      column_to_rownames(var='hgnc_symbol') %>% 
      AnnotatedDataFrame()
    ,protocolData=
      pdata$GSE177477$eset %>%
      .[,colnames(gdata$GSE177477$hgnc$datETcollapsed)] %>%
      protocolData() %>%
      pData() %>%
      AnnotatedDataFrame()
    ,experimentData=experimentData(pdata$GSE177477$eset)
    ,annotation=annotation(pdata$GSE177477$eset)
  )

# Take common genes among the expression sets, except that of COVID-19
source('R/take_common_genes-function.R')
GSE108497=
  GSE108497 %>%
  take_common_genes(list(GSE108497,GSE73685,GSE85307,GSE86200,GSE149437))

GSE73685=
  GSE73685 %>%
  take_common_genes(list(GSE108497,GSE73685,GSE85307,GSE86200,GSE149437))

GSE85307=
  GSE85307 %>%
  take_common_genes(list(GSE108497,GSE73685,GSE85307,GSE86200,GSE149437))

GSE86200=
  GSE86200 %>%
  take_common_genes(list(GSE108497,GSE73685,GSE85307,GSE86200,GSE149437))

GSE149437=
  GSE149437 %>%
  take_common_genes(list(GSE108497,GSE73685,GSE85307,GSE86200,GSE149437))

GSE177477=
  GSE177477 %>%
  take_common_genes(
    list(GSE108497,GSE73685,GSE85307,GSE86200,GSE149437,GSE177477)
  )
```

# Data exploration

```{r Summarize the available genes in other studies, eval=FALSE, include=FALSE}
rownames(GSE108497) %>%
  data.frame(gene=.) %>%
  mutate(
    ,gene_in_covid19_array=ifelse(gene%in%rownames(GSE177477),1,0)
    ,this_study=1
    ,common_predictor=ifelse(gene%in%c('ADAM12','FLT1','PAPPA2'),1,0)
    ,cfRNA=
      ifelse(
        gene
        %in%c(
          'CAMK2G'
          ,'DERA'
          ,'FAM46A'
          ,'KIAA1109'
          ,'LRRC58'
          ,'MYLIP'
          ,'NDUFV3'
          ,'NMRK1'
          ,'PI4KA'
          ,'PRTFDC1'
          ,'PYGO2'
          ,'RNF149'
          ,'TFIP11'
          ,'TRIM21'
          ,'USB1'
          ,'YWHAQP5'
          ,'Y_RNA'
        )
        ,1,0
      )
    ,meta_analysis=
      ifelse(
        gene
        %in%pull(
          readxl::read_xlsx(
            # Meta-analysis of gene expression profiles in preeclampsia
            # https://pubmed.ncbi.nlm.nih.gov/31901653/
            'data/1-s2.0-S2210778919304799-mmc3.xlsx'
            ,sheet=2
            ) %>%
            slice(-1)
          ,1
        )
        ,1,0
      )
  ) %>%
  arrange(
    desc(gene_in_covid19_array)
    ,desc(this_study)
    ,desc(common_predictor)
    ,desc(cfRNA)
    ,desc(meta_analysis)
  ) %>%
  lapply(X=1:2,Y=.,function(X,Y){
    if(X==1){
      Y
    }else{
      Y %>%
        group_by(
          gene_in_covid19_array
          ,this_study,common_predictor,cfRNA,meta_analysis
        ) %>%
        summarize(
          n=n()
          ,genes_if_20_or_less=ifelse(n<=20,paste0(gene,collapse=','),NA)
        ) %>%
        .[rev(seq(nrow(.))),]
    }
  })
```

```{r Summarize phenotype of the processed GSE108497, eval=FALSE, include=FALSE}
GSE108497 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(colnames(.) %>% str_remove_all(':ch1')) %>%
  filter(
    apl==0
    & lac==0
    & fd==0
    & nnd==0
  ) %>%
  mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
  mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
  select(sle,early,pe,sga,time_point) %>%
  # unite(outcome,sle,early,pe,sga) %>%
  table()%>%
  as.data.frame() %>%
  spread(time_point,Freq) %>%
  # group_by(sle,early,pe,sga) %>%
  select(-sle) %>%
  group_by(early,pe,sga) %>%
  summarize_all(sum) %>%
  mutate(total_preg=`<16 weeks`+`16-23 weeks`+`24-31 weeks`+`32-40 weeks`) %>%
  select(total_preg,everything()) %>%
  ungroup() %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  filter(
    (early==0 & pe==0 & sga==0)
    |
    (early==1 & pe==1 & sga==0)
    |
    (early==0 & pe==1 & sga==0)
    |
    (early==0 & pe==0 & sga==1)
  ) %>%
  arrange(factor(seq,c(1,7,3,2)))
```

```{r Characterize the development dataset, eval=FALSE, include=FALSE}
GSE108497 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(colnames(.) %>% str_remove_all(':ch1')) %>%
  filter(
    apl==0
    & lac==0
    & fd==0
    & nnd==0
  ) %>%
  mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
  mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
  filter(
    time_point%in%c('<16 weeks','16-23 weeks','24-31 weeks','32-40 weeks')
  ) %>%
  filter(
    (early==0 & pe==0 & sga==0)
    |
    (early==1 & pe==1 & sga==0)
    |
    (early==0 & pe==1 & sga==0)
    |
    (early==0 & pe==0 & sga==1)
  ) %>%
  mutate(
    ga_coll_num=str_remove_all(ga_at_collection,'[:alpha:]|\\s')
    ,ga_coll_chr=str_remove_all(ga_at_collection,'[:number:]|[:punct:]|\\s')
  ) %>%
  select(
    early,pe,sga
    ,age,ga_coll_num
    ,ethnicity,ga_coll_chr,pl_insuff,race,sle
  ) %>%
  mutate(
    outcome=
      case_when(
        early==0 & pe==0 & sga==0 ~ 'Normal'
        ,early==0 & pe==0 & sga==1 ~ 'Late Isolated FGR'
        ,early==1 & pe==1 & sga==0 ~ 'EOPE'
        ,early==0 & pe==1 & sga==0 ~ 'LOPE'
        ,early==1 & pe==0 & sga==0 ~ 'Early delivery'
        ,early==1 & pe==1 & sga==1 ~ 'EOPE-FGR'
        ,early==0 & pe==1 & sga==1 ~ 'LOPE-FGR'
        ,early==1 & pe==0 & sga==1 ~ 'Early Isolated FGR'
        ,TRUE ~ ''
      )
    ,group=ifelse(pe==1,'event','nonevent')
  ) %>%
  mutate(
    ethnicity=
      ifelse(
        ethnicity!='Hispanic or Latino'
        ,'0. Not Hispanic or Latino'
        ,'1. Hispanic or Latino'
      )
  ) %>%
  gather(key,value,-group) %>%
  lapply(X=1:3,Y=.,function(X,Y){
    if(X==1){
      Y %>%
        filter(key%in%c('age','ga_coll_num')) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=as.numeric(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            filter(term=='value') %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.01 ~ as.character(round(p.value,3))
              ,p.value>0.01 & p.value<=0.05 ~ as.character(round(p.value,2))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value) %>%
        summarize(
          var1=round(mean(as.numeric(value)),ifelse(n()<100,1,2))
          ,var2=round(sd(as.numeric(value)),ifelse(n()<100,1,2))
          ,var3=n()
        ) %>%
        ungroup() %>%
        mutate(category=NA) %>%
        mutate(var=paste0(var1,' (',var2,')'))
    }else if(X==2){
      Y %>%
        filter(key%in%c('ethnicity','sle')) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=factor(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            slice(2) %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.05 ~ as.character(round(p.value,3))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }else{
      Y %>%
        filter(!key%in%c('age','ga_coll_num','ethnicity','sle')) %>%
        group_by(key) %>%
        mutate(p.value=NA) %>%
        ungroup() %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }
  }) %>%
  do.call(rbind,.) %>%
  arrange(key,group) %>%
  select(key,everything()) %>%
  select(-var1,-var2,-var3) %>%
  spread(group,var)
```

```{r Summarize phenotype of the processed GSE73685, eval=FALSE, include=FALSE}
GSE73685 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  lapply(
      X=.$tissue_type %>%
        .[!duplicated(.)] %>%
        .[str_to_lower(.)!='maternal blood']
      ,Y=.
      ,function(X,Y){
    Z=group_by(.,subject_id) %>%
      filter(
        str_to_lower(tissue_type)==str_to_lower(X)
        | str_to_lower(tissue_type)=='maternal blood'
      ) %>%
      filter(n()==2) %>%
      ungroup() %>%
      select(subject_id,outcome) %>%
      filter(!duplicated(.)) %>%
      pull(outcome) %>%
      table()
    
    c(tissue=X,Z) %>%
      t() %>%
      as.data.frame() %>%
      gather(outcome,n,-tissue)
  }) %>%
  do.call(rbind,.) %>%
  mutate(sum_n=sum(as.integer(n))) %>%
  group_by(tissue) %>%
  mutate(n_tissue=sum(as.integer(n))) %>%
  spread(outcome,n,fill=0)
```

```{r Characterize the derivation dataset, eval=FALSE, include=FALSE}
GSE73685 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  lapply(
      X=.$tissue_type %>%
        .[!duplicated(.)] %>%
        .[str_to_lower(.)!='maternal blood']
      ,Y=.
      ,function(X,Y){
    Z=group_by(.,subject_id) %>%
      filter(
        str_to_lower(tissue_type)==str_to_lower(X)
        | str_to_lower(tissue_type)=='maternal blood'
      ) %>%
      filter(n()==2) %>%
      ungroup() %>%
      select(subject_id,outcome) %>%
      filter(!duplicated(.)) %>%
      pull(outcome) %>%
      table()
    
    c(tissue=X,Z) %>%
      t() %>%
      as.data.frame() %>%
      gather(outcome,n,-tissue)
  }) %>%
  do.call(rbind,.) %>%
  mutate(sum_n=sum(as.integer(n))) %>%
  group_by(outcome) %>%
  mutate(n_tissue=sum(as.integer(n))) %>%
  spread(tissue,n,fill=0) %>%
  mutate(p_tissue=round(n_tissue/sum_n*100,2)) %>%
  select(outcome,n_tissue,p_tissue,everything())
```

```{r Summarize phenotype of the processed GSE85307, eval=FALSE, include=FALSE}
GSE85307 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=
      as.numeric(blood_draw_at_gestation_week)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
    ,early=ifelse(gestation_weeks<=34,'Early_Onset','Late_Onset')
    ,trt=
      ifelse(
        `vitamind_25(0h)_ng/ml_whole_blood_baseline`<30
        ,'D<30','D30+'
      )
  ) %>%
  unite(outcome,early,pregnancy_condition,trt) %>%
  select(outcome,time_point) %>%
  table() %>%
  as.data.frame() %>%
  spread(time_point,Freq,fill=0) %>%
  mutate(outcome=str_replace_all(outcome,'_',' ')) %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  arrange(factor(seq,c(3,1,5,4,2,6)))
```

```{r Characterize the replication dataset 1, eval=FALSE, include=FALSE}
GSE85307 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=
      as.numeric(blood_draw_at_gestation_week)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
    ,early=ifelse(gestation_weeks<=34,'Early Onset','Late Onset')
    ,trt=
      ifelse(
        `vitamind_25(0h)_ng/ml_whole_blood_baseline`<30
        ,'1. D<30','0. D30+'
      )
    ,bmi_cat=
      case_when(
        `bmi_kg/m2`<18.5 ~'Underweight'
        ,`bmi_kg/m2`>=18.5 & `bmi_kg/m2`<25~'Normal'
        ,`bmi_kg/m2`>=25 & `bmi_kg/m2`<30~'Overweight'
        ,`bmi_kg/m2`>=30~'Obesity'
        ,TRUE~''
      )
  ) %>%
  unite(outcome,early,pregnancy_condition,sep=' ') %>%
  select(
    outcome
    ,age_years,blood_draw_at_gestation_week
    ,ethnicity
    ,`bmi_kg/m2`
    ,bmi_cat
    ,`asthma_0=healthy_1=asthmatic`
    ,`vitamind_25(0h)_ng/ml_whole_blood_baseline`
    ,trt
  ) %>%
  mutate(
    group=ifelse(str_detect(outcome,'Preeclampsia'),'event','nonevent')
    ,ethnicity=
      case_when(
        ethnicity==unique(ethnicity)[2]~paste0('0. ',ethnicity)
        ,ethnicity==unique(ethnicity)[3]~paste0('1. ',ethnicity)
        ,ethnicity==unique(ethnicity)[5]~paste0('2. ',ethnicity)
        ,ethnicity==unique(ethnicity)[1]~paste0('3. ',ethnicity)
        ,ethnicity==unique(ethnicity)[4]~paste0('4. ',ethnicity)
        ,TRUE~''
      )
    ,bmi_cat=
      case_when(
        bmi_cat==unique(bmi_cat)[2]~paste0('0. ',bmi_cat)
        ,bmi_cat==unique(bmi_cat)[4]~paste0('1. ',bmi_cat)
        ,bmi_cat==unique(bmi_cat)[3]~paste0('2. ',bmi_cat)
        ,bmi_cat==unique(bmi_cat)[1]~paste0('3. ',bmi_cat)
        ,TRUE~''
      )
  ) %>%
  gather(key,value,-group) %>%
  lapply(X=1:4,Y=.,function(X,Y){
    if(X==1){
      Y %>%
        filter(key%in%c('age_years'
                        ,'blood_draw_at_gestation_week'
                        ,'bmi_kg/m2'
                        ,'vitamind_25(0h)_ng/ml_whole_blood_baseline')) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=as.numeric(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            filter(term=='value') %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.01 ~ as.character(round(p.value,3))
              ,p.value>0.01 & p.value<=0.05 ~ as.character(round(p.value,2))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value) %>%
        summarize(
          var1=round(mean(as.numeric(value)),ifelse(n()<100,1,2))
          ,var2=round(sd(as.numeric(value)),ifelse(n()<100,1,2))
          ,var3=n()
        ) %>%
        ungroup() %>%
        mutate(category=NA) %>%
        mutate(var=paste0(var1,' (',var2,')'))
    }else if(X==2){
      Y %>%
        filter(key%in%c('bmi_cat','ethnicity')) %>%
        left_join(
          group_by(.,key) %>%
            mutate(
              group=factor(group,c('event','nonevent'))
              ,value=factor(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            slice(-1) %>%
            mutate(
              p.value=
                case_when(
                  p.value<=0.001 ~ 'p≤0.001'
                  ,p.value>0.001 & p.value<=0.05 ~ as.character(round(p.value,3))
                  ,TRUE~'p>0.05'
                )
            ) %>%
            mutate(value=str_remove_all(term,'value')) %>%
            select(key,value,p.value) %>%
            ungroup() 
          ,by=c('key','value')
        ) %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }else if(X==3){
      Y %>%
        filter(key%in%c('asthma_0=healthy_1=asthmatic','trt')) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=factor(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            slice(2) %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.05 ~ as.character(round(p.value,3))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }else{
      Y %>%
        filter(!key%in%c('age_years','blood_draw_at_gestation_week'
                         ,'ethnicity'
                         ,'bmi_kg/m2'
                         ,'bmi_cat'
                         ,'asthma_0=healthy_1=asthmatic'
                         ,'vitamind_25(0h)_ng/ml_whole_blood_baseline'
                         ,'trt')) %>%
        group_by(key) %>%
        mutate(p.value=NA) %>%
        ungroup() %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }
  }) %>%
  do.call(rbind,.) %>%
  arrange(key,group) %>%
  select(key,everything()) %>%
  select(-var1,-var2,-var3) %>%
  spread(group,var)
```

```{r Summarize phenotype of the processed GSE86200, eval=FALSE, include=FALSE}
GSE86200 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=
      floor(as.numeric(gestational_day_at_enrollment)/7)
    ,blood_draw_at_gestation_week=
      ifelse(str_detect(title,'32-38'),34,blood_draw_at_gestation_week)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
    ,trt=
      ifelse(
        str_detect(treatment_group,'Intervention')
        ,'Treated','Untreated'
      )
  ) %>%
  unite(outcome,pregnancy_condition,trt) %>%
  select(outcome,time_point) %>%
  table() %>%
  as.data.frame() %>%
  spread(time_point,Freq,fill=0) %>%
  mutate(outcome=str_replace_all(outcome,'_',' ')) %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  arrange(factor(seq,c(2,4,1,3)))
```

```{r Characterize the replication dataset 2, eval=FALSE, include=FALSE}
GSE86200 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=
      floor(as.numeric(gestational_day_at_enrollment)/7)
    ,blood_draw_at_gestation_week=
      ifelse(str_detect(title,'32-38'),34,blood_draw_at_gestation_week)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
    ,trt=
      ifelse(
        str_detect(treatment_group,'Intervention')
        ,'Treated','Untreated'
      )
    ,onset=
      ifelse(str_detect(title,'enrollment'),'pre','post')
  ) %>%
  rename(outcome=pregnancy_condition) %>%
  select(
    recruitment_id
    ,outcome,onset
    ,age_at_enrollment,gestational_day_at_enrollment
    ,ethnicity
    ,sex_of_child
    ,`vitamin_d_level_at_enrollment_(nmol/l)`
    ,`vitamin_d_level_at_32-38_weeks_of_gestation_(nmol/l)`
  ) %>%
  mutate(
    gestational_day_at_enrollment=
      as.numeric(gestational_day_at_enrollment)/7
    ,`vitamin_d_level(nmol/l)`=
      ifelse(
        onset=='pre'
        ,`vitamin_d_level_at_enrollment_(nmol/l)`
        ,`vitamin_d_level_at_32-38_weeks_of_gestation_(nmol/l)`
      )
  ) %>%
  rename(gestational_age_at_enrollment=gestational_day_at_enrollment) %>%
  select(
    -`vitamin_d_level_at_enrollment_(nmol/l)`
    ,-`vitamin_d_level_at_32-38_weeks_of_gestation_(nmol/l)`
  ) %>%
  mutate(onset=paste0('vitamin_d_level(nmol/l)_',onset)) %>%
  spread(onset,`vitamin_d_level(nmol/l)`) %>%
  mutate(
    group=ifelse(str_detect(outcome,'Preeclampsia'),'event','nonevent')
    ,ethnicity=
      case_when(
        ethnicity==unique(ethnicity)[1]~paste0('0. ',ethnicity)
        ,ethnicity==unique(ethnicity)[2]~paste0('1. ',ethnicity)
        ,TRUE~''
      )
    ,sex_of_child=
      case_when(
        sex_of_child=='Female'~'0. Female'
        ,sex_of_child=='Male'~'1. Male'
        ,TRUE~''
      )
    ,vit_d_improvement=
      as.numeric(`vitamin_d_level(nmol/l)_post`)
      -as.numeric(`vitamin_d_level(nmol/l)_pre`)
  ) %>%
  gather(key,value,-group,-recruitment_id) %>%
  lapply(X=1:4,Y=.,function(X,Y){
    if(X==1){
      Y %>%
        filter(key%in%c('age_at_enrollment'
                        ,'gestational_age_at_enrollment'
                        ,'vitamin_d_level(nmol/l)_pre'
                        ,'vitamin_d_level(nmol/l)_post'
                        ,'vit_d_improvement')) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=as.numeric(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            filter(term=='value') %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.01 ~ as.character(round(p.value,3))
              ,p.value>0.01 & p.value<=0.05 ~ as.character(round(p.value,2))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value) %>%
        summarize(
          var1=round(mean(as.numeric(value),na.rm=T),ifelse(n()<100,1,2))
          ,var2=round(sd(as.numeric(value),na.rm=T),ifelse(n()<100,1,2))
          ,var3=n()
        ) %>%
        ungroup() %>%
        mutate(category=NA) %>%
        mutate(var=paste0(var1,' (',var2,')'))
    }else if(X==2){
      Y %>%
        filter(key%in%c('ethnicity')) %>%
        left_join(
          group_by(.,key) %>%
            mutate(
              group=factor(group,c('event','nonevent'))
              ,value=factor(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            slice(-1) %>%
            mutate(
              p.value=
                case_when(
                  p.value<=0.001 ~ 'p≤0.001'
                  ,p.value>0.001 & p.value<=0.05 ~ as.character(round(p.value,3))
                  ,TRUE~'p>0.05'
                )
            ) %>%
            mutate(value=str_remove_all(term,'value')) %>%
            select(key,value,p.value) %>%
            ungroup() 
          ,by=c('key','value')
        ) %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }else if(X==3){
      Y %>%
        filter(key%in%c('sex_of_child')) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=factor(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            slice(2) %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.05 ~ as.character(round(p.value,3))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }else{
      Y %>%
        filter(!key%in%c('age_at_enrollment','gestational_age_at_enrollment'
                         ,'ethnicity'
                         ,'sex_of_child'
                         ,'vitamin_d_level(nmol/l)_pre'
                         ,'vitamin_d_level(nmol/l)_post'
                         ,'vit_d_improvement')) %>%
        group_by(key) %>%
        mutate(p.value=NA) %>%
        ungroup() %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }
  }) %>%
  do.call(rbind,.) %>%
  arrange(key,group) %>%
  select(key,everything()) %>%
  select(-var1,-var2,-var3) %>%
  spread(group,var)
```

```{r Summarize phenotype of the processed GSE149437, eval=FALSE, include=FALSE}
GSE149437 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=as.numeric(gestational_age)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
  ) %>%
  rename(outcome=group) %>%
  select(outcome,time_point) %>%
  table() %>%
  as.data.frame() %>%
  spread(time_point,Freq,fill=0) %>%
  mutate(outcome=str_replace_all(outcome,'_',' ')) %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  arrange(factor(seq,c(1,3,4,2)))
```

```{r Characterize the replication dataset 3, eval=FALSE, include=FALSE}
GSE149437 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=as.numeric(gestational_age)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
  ) %>%
  rename(outcome=group) %>%
  select(
    individual
    ,outcome
    ,blood_draw_at_gestation_week,diagnosisga,gadelivery
    ,d2interval,d3interval
  ) %>%
  mutate(
    group=ifelse(str_detect(outcome,'Preeclampsia'),'event','nonevent')
  ) %>%
  lapply(X=1:3,Y=.,function(X,Y){
    if(X==1){
      Y %>%
        select(-individual) %>%
        gather(key,value,-group) %>%
        filter(key%in%c('blood_draw_at_gestation_week')) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=as.numeric(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            filter(term=='value') %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.01 ~ as.character(round(p.value,3))
              ,p.value>0.01 & p.value<=0.05 ~ as.character(round(p.value,2))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value) %>%
        summarize(
          var1=round(mean(as.numeric(value),na.rm=T),ifelse(n()<100,1,2))
          ,var2=round(sd(as.numeric(value),na.rm=T),ifelse(n()<100,1,2))
          ,var3=n()
        ) %>%
        ungroup() %>%
        mutate(category=NA) %>%
        mutate(var=paste0(var1,' (',var2,')'))
    }else if(X==2){
      Y %>%
        gather(key,value,-group,-individual) %>%
        filter(key%in%c('diagnosisga','gadelivery')) %>%
        filter(!duplicated(.)) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=as.numeric(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            filter(term=='value') %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.01 ~ as.character(round(p.value,3))
              ,p.value>0.01 & p.value<=0.05 ~ as.character(round(p.value,2))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value) %>%
        summarize(
          var1=round(mean(as.numeric(value),na.rm=T),ifelse(n()<100,1,2))
          ,var2=round(sd(as.numeric(value),na.rm=T),ifelse(n()<100,1,2))
          ,var3=n()
        ) %>%
        ungroup() %>%
        mutate(category=NA) %>%
        mutate(var=paste0(var1,' (',var2,')'))
    }else{
      Y %>%
        gather(key,value,-group,-individual) %>%
        filter(!key%in%c('blood_draw_at_gestation_week'
                         ,'diagnosisga'
                         ,'gadelivery')) %>%
        group_by(key) %>%
        mutate(p.value=NA) %>%
        ungroup() %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }
  }) %>%
  do.call(rbind,.) %>%
  arrange(key,group) %>%
  select(key,everything()) %>%
  select(-var1,-var2,-var3) %>%
  spread(group,var)
```

```{r Summarize phenotype of the processed GSE177477, eval=FALSE, include=FALSE}
GSE177477 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  separate(
    comorbids
    ,c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
    ,sep='\\|'
  ) %>%
  mutate_at(
    c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
    ,function(x) ifelse(x=='',0,ifelse(is.na(x),0,1))
  ) %>%
  mutate(
    severity=
      factor(
        ifelse(
          is.na(severity)
          ,'uninfected'
          ,severity
        )
        ,c('uninfected','mild','severe')
      )
    ,sex=
      factor(
        ifelse(
          is.na(sex)
          ,'U'
          ,sex
        )
        ,c('U','F','M')
      )
  ) %>%
  rename(outcome=severity) %>%
  select(outcome,symptons,sex) %>%
  table() %>%
  as.data.frame() %>%
  spread(sex,Freq,fill=0) %>%
  mutate(seq=seq(nrow(.))) %>%
  select(seq,everything()) %>%
  filter(seq%in%c(2,1,6,9)) %>%
  arrange(factor(seq,c(2,1,6,9)))
```

```{r Pre-characterize the COVID-19 dataset, eval=FALSE, include=FALSE}
suppressWarnings(
  GSE177477 %>%
    phenoData() %>%
    pData() %>%
    select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
    setNames(
      colnames(.) %>%
        str_remove_all(':ch1') %>%
        str_replace_all('\\s','_')
    ) %>%
    separate(
      comorbids
      ,c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
      ,sep='\\|'
    ) %>%
    mutate_at(
      c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
      ,function(x) ifelse(x=='',0,ifelse(is.na(x),0,1))
    ) %>%
    mutate(
      age=
        case_when(
          age%in%c('20-24','44-49','55-59')~'<60'
          ,age%in%c('60-64','65-69','70-74','75-79','80-84')~'60+'
          ,is.na(age)~'NA'
          ,TRUE~''
        ) %>%
        str_replace_na()
      ,age=ifelse(symptons=='Healthy controls','<60',age)
      ,severity=
        factor(
          ifelse(
            is.na(severity)
            ,'uninfected'
            ,severity
          )
          ,c('uninfected','mild','severe')
        )
    ) %>%
    rename(individual=id,outcome=severity) %>%
    select(
      individual
      ,outcome
      ,disease_state
      ,symptons
      ,age
      ,sex
      ,who_score_admission
      ,d_dimer
      ,d_dimer_30
      ,crp
      ,ldh
      ,ferritin
      ,wbc
      ,lymphocytes
      ,nlr
      ,asthma
      ,cad
      ,caf
      ,t2dm
      ,htn
      ,ihd
      ,sp_cabg
      ,who_score_discharge
      ,endpoint
    ) %>%
    unite(symptons,symptons,outcome,sep=' ') %>%
    gather(variable,value,-symptons) %>%
    group_by(symptons,variable,value) %>%
    summarize(n=n()) %>%
    spread(symptons,n)
)
```

```{r Characterize the COVID-19 dataset, eval=FALSE, include=FALSE}
GSE177477 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  separate(
    comorbids
    ,c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
    ,sep='\\|'
  ) %>%
  mutate_at(
    c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
    ,function(x) ifelse(x=='',0,ifelse(is.na(x),0,1))
  ) %>%
  mutate(
    age=
      case_when(
        age=='20-24'~'<30'
        ,age%in%c('44-49','55-59')~'30-59'
        ,age%in%c('60-64','65-69','70-74','75-79','80-84')~'60+'
        ,is.na(age)~'NA'
        ,TRUE~''
      ) %>%
      str_replace_na()
    ,severity=
      factor(
        ifelse(
          is.na(severity)
          ,'uninfected'
          ,severity
        )
        ,c('uninfected','mild','severe')
      )
  ) %>%
  rename(individual=id,outcome=severity) %>%
  select(
    individual
    ,outcome
    ,disease_state
    ,symptons
    ,age
    ,sex
    ,who_score_admission
    ,d_dimer
    ,d_dimer_30
    ,crp
    ,ldh
    ,ferritin
    ,wbc
    ,lymphocytes
    ,nlr
    ,asthma
    ,cad
    ,caf
    ,t2dm
    ,htn
    ,ihd
    ,sp_cabg
    ,who_score_discharge
    ,endpoint
  ) %>%
  mutate(
    group=ifelse(str_detect(outcome,'severe'),'event','nonevent')
    ,d_dimer=ifelse(is.infinite(d_dimer),NA,d_dimer)
  ) %>%
  mutate_at(
    c('outcome','age','sex','d_dimer_30'
      ,'asthma','cad','caf','t2dm','htn','ihd','sp_cabg','endpoint')
    ,as.character
  ) %>%
  lapply(X=1:4,Y=.,function(X,Y){
    if(X==1){
      Y %>%
        select_at(colnames(.) %>% .[.%in%c('individual'
                                           ,'group'
                                           ,'who_score_admission'
                                           ,'d_dimer'
                                           ,'crp'
                                           ,'ldh'
                                           ,'ferritin'
                                           ,'wbc'
                                           ,'lymphocytes'
                                           ,'nlr'
                                           ,'who_score_discharge')]) %>%
        gather(key,value,-group,-individual) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=as.numeric(value)
            ) %>%
            do(broom::tidy(suppressWarnings(
              glm(group~value,data=.,family=binomial(link='logit'))
            ))) %>%
            filter(term=='value') %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.01 ~ as.character(round(p.value,3))
              ,p.value>0.01 & p.value<=0.05 ~ as.character(round(p.value,2))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value) %>%
        summarize(
          var1=round(mean(as.numeric(value),na.rm=T),ifelse(n()<100,1,2))
          ,var2=round(sd(as.numeric(value),na.rm=T),ifelse(n()<100,1,2))
          ,var3=n()
        ) %>%
        ungroup() %>%
        mutate(category=NA) %>%
        mutate(var=paste0(var1,' (',var2,')'))
    }else if(X==2){
      Y %>%
        select_at(colnames(.) %>% .[.%in%c('individual','group','age')]) %>%
        gather(key,value,-group,-individual) %>%
        left_join(
          group_by(.,key) %>%
            mutate(
              group=factor(group,c('event','nonevent'))
              ,value=factor(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            slice(-1) %>%
            mutate(
              p.value=
                case_when(
                  p.value<=0.001 ~ 'p≤0.001'
                  ,p.value>0.001 & p.value<=0.05 ~ as.character(round(p.value,3))
                  ,TRUE~'p>0.05'
                )
            ) %>%
            mutate(value=str_remove_all(term,'value')) %>%
            select(key,value,p.value) %>%
            ungroup() 
          ,by=c('key','value')
        ) %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }else if(X==3){
      Y %>%
        select_at(colnames(.) %>% .[.%in%c('individual','group','outcome','sex'
                                           ,'d_dimer_30','asthma','cad','caf'
                                           ,'t2dm','htn','ihd','sp_cabg'
                                           ,'endpoint')]) %>%
        gather(key,value,-group,-individual) %>%
        group_by(key) %>%
        mutate(
          p.value=
            data.frame(
              group=factor(group,c('event','nonevent'))
              ,value=factor(value)
            ) %>%
            do(broom::tidy(
              glm(group~value,data=.,family=binomial(link='logit'))
            )) %>%
            slice(2) %>%
            pull(p.value)
          ,p.value=
            case_when(
              p.value<=0.001 ~ 'p≤0.001'
              ,p.value>0.001 & p.value<=0.05 ~ as.character(round(p.value,3))
              ,TRUE~'p>0.05'
            )
        ) %>%
        ungroup() %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }else{
      Y %>%
        select_at(colnames(.) %>% .[!.%in%c('who_score_admission'
                                            ,'d_dimer'
                                            ,'crp'
                                            ,'ldh'
                                            ,'ferritin'
                                            ,'wbc'
                                            ,'lymphocytes'
                                            ,'nlr'
                                            ,'who_score_discharge'
                                            ,'age'
                                            ,'outcome','sex'
                                            ,'d_dimer_30','asthma','cad','caf'
                                            ,'t2dm','htn','ihd','sp_cabg'
                                            ,'endpoint')]) %>%
        gather(key,value,-group,-individual) %>%
        group_by(key) %>%
        mutate(p.value=NA) %>%
        ungroup() %>%
        group_by(group,key,p.value,value) %>%
        summarize(
          var1=''
          ,var2=''
          ,var3=n()
        ) %>%
        ungroup() %>%
        group_by(group,key) %>%
        mutate(
          var1=sum(var3)
          ,var2=round(var3/var1*100,2)
        ) %>%
        ungroup() %>%
        rename(category=value) %>%
        mutate(var=paste0(var3,' (',var2,')'))
    }
  }) %>%
  do.call(rbind,.) %>%
  arrange(key,group) %>%
  select(key,everything()) %>%
  select(-var1,-var2,-var3) %>%
  spread(group,var)
```

```{r}
timestamper[7]=as.character(now())
```

#	Derivation of maternal-fetal interface transcriptome from maternal blood

```{r Create a list of tissue-pairwise expression sets, include=FALSE}
surrogate=
  GSE73685 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  rownames_to_column(var='id') %>%
  lapply(
      X=.$tissue_type %>%
        .[!duplicated(.)] %>%
        .[str_to_lower(.)!='maternal blood'] %>%
        sort()
      ,Y=.
      ,function(X,Y){
    group_by(.,subject_id) %>%
      filter(
        str_to_lower(tissue_type)==str_to_lower(X)
        | str_to_lower(tissue_type)=='maternal blood'
      ) %>%
      filter(n()==2) %>%
      ungroup() %>%
      select(id,subject_id,tissue_type) %>%
      mutate(
        tissue_type=
          tissue_type %>%
          str_to_lower() %>%
          str_replace_all('\\s+','_')
      ) %>%
      spread(tissue_type,id) %>%
      column_to_rownames(var='subject_id')
  }) %>%
  setNames(
    GSE73685 %>%
      phenoData() %>%
      pData() %>%
      select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
      setNames(
        colnames(.) %>%
          str_remove_all(':ch1') %>%
          str_replace_all('\\s','_')
      ) %>%
      .$tissue_type %>%
      .[!duplicated(.)] %>%
      .[str_to_lower(.)!='maternal blood'] %>%
      sort() %>%
      str_to_lower() %>%
      str_replace_all('\\s+','_')
  ) %>%
  lapply(function(x){
    GSE73685 %>%
      .[,c(x[[1]],x[[2]])] %>%
      `phenoData<-`(
        phenoData(.) %>%
          pData() %>%
          rownames_to_column(var='id') %>%
          left_join(
            x %>%
              rownames_to_column(var='subject_id') %>%
              gather(tissue_type,id,-subject_id) %>%
              select(-subject_id)
            ,by='id'
          ) %>%
          mutate(
            tissue_type=
              tissue_type %>%
              factor(c('maternal_blood',unique(.) %>% .[.!='maternal_blood']))
          ) %>%
          select(tissue_type,everything()) %>%
          rename(outcome=tissue_type) %>%
          column_to_rownames(var='id') %>%
          AnnotatedDataFrame()
      )
  })
```

```{r Differential expression analysis per tissue pair, include=FALSE}
source('R/conduct_dea-function.R')

dea=
  surrogate %>%
  lapply(conduct_dea,nonevent='maternal_blood',paired=T)
```

```{r Summarize the DEGs, echo=FALSE}
dea %>%
  lapply(X=names(.),Y=.,function(X,Y){
    Y[[X]]$result %>%
      mutate(
        regulation=
          ifelse(
            adj.P.Val>=0.05
            ,'non-DEG'
            ,ifelse(
              logFC>=0
              ,ifelse(logFC>2,'strong up','weak up')
              ,ifelse(logFC<(-2),'strong down','weak down')
            )
          )
      ) %>%
      group_by(regulation) %>%
      summarize(n=n()) %>%
      ungroup() %>%
      mutate(comparison=paste0(X,' vs. maternal blood'))
  }) %>%
  do.call(rbind,.) %>%
  spread(regulation,n,fill=0) %>%
  mutate(
    total=
      `non-DEG`
      +`strong down`
      +`strong up`
      +`weak down`
      +`weak up`
  )
```

```{r An example for computing of a DEG, eval=FALSE, include=FALSE}
dea$amnion$result

dea$amnion$eset %>%
  exprs() %>%
  t() %>%
  as.data.frame() %>%
  select(AAAS) %>%
  rownames_to_column(var='id') %>%
  left_join(
    dea$amnion$eset %>%
      phenoData() %>%
      pData() %>%
      rownames_to_column(var='id') %>%
      separate(title,c('tissue_type','subject_id'),sep='_') %>%
      select(id,outcome,subject_id)
    ,by='id'
  ) %>%
  group_by(outcome) %>%
  mutate(avg=mean(AAAS)) %>%
  ungroup() %>%
  cbind(
    select(.,outcome,avg) %>%
      filter(!duplicated(.)) %>%
      mutate(
        outcome=ifelse(outcome=='maternal_blood','AveExpr1','AveExpr2')
      ) %>%
      spread(outcome,avg) %>%
      mutate(logFC=AveExpr2-AveExpr1)
  ) %>%
  mutate(
    AveExpr=mean(AAAS)
    ,FC=2^logFC
  ) %>%
  select(-id,-avg) %>%
  mutate(outcome=ifelse(outcome=='maternal_blood','Expr1','Expr2')) %>%
  spread(outcome,AAAS) %>%
  mutate(
    feature=Expr1-AveExpr1
    ,outcome=as.integer(Expr2>=AveExpr1)
  )
```

```{r Create a function to define feature and outcome of a DEG, include=FALSE}
deg=function(the_dea,tissue,side=c('feature','outcome'),direction=F){
  eset=
    GSE73685 %>%
    .[,colnames(the_dea$eset)]
  
  if(side=='feature'){
    eset=
      eset %>%
      `assayData<-`(
        ExpressionSet(
            `dimnames<-`(
              normalize.quantiles.use.target(
                x=exprs(.)
                ,target=
                  the_dea$result[rownames(.),'AveExpr1',drop=F] %>%
                    pull(AveExpr1)
              )
              ,dimnames(.)
            )
          ) %>%
          assayData()
      )
  }
  
  eset %>%
    phenoData() %>%
    pData() %>%
    select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
    setNames(
      colnames(.) %>%
        str_remove_all(':ch1') %>%
        str_replace_all('\\s','_')
    ) %>%
    rownames_to_column(var='id') %>%
    filter(tissue_type==tissue) %>%
    select(subject_id,everything()) %>%
    arrange(subject_id) %>%
    lapply(X=1,Y=.,function(X,Y){
      GSE73685[,Y$id] %>%
        `colnames<-`(Y$subject_id) %>%
        exprs() %>%
        as.data.frame() %>%
        rownames_to_column(var='feature') %>%
        mutate(tissue_type=Y$tissue_type[1]) %>%
        gather(subject_id,value,-feature,-tissue_type)
    }) %>%
    .[[1]] %>%
    inner_join(
      the_dea$result %>%
        rownames_to_column(var='feature') %>%
        filter(adj.P.Val<0.05) %>%
        select(feature,AveExpr,AveExpr1,AveExpr2)
      ,by='feature'
    ) %>%
    mutate(type=side,regulation=direction) %>%
    mutate(
      outcome=
        ifelse(
          type=='feature'
          ,ifelse(
            AveExpr1<AveExpr
            ,ifelse(
              value<AveExpr
              ,0-(value-AveExpr)/(AveExpr1-AveExpr)
              ,0+(value-AveExpr)/(AveExpr2-AveExpr)
            )
            ,ifelse(
              value>AveExpr
              ,0-(value-AveExpr)/(AveExpr1-AveExpr)
              ,0+(value-AveExpr)/(AveExpr2-AveExpr)
            )
          )
          ,ifelse(
            regulation
            ,as.integer(value>=AveExpr1)
            ,ifelse(
              AveExpr1<AveExpr
              ,as.integer(value>=AveExpr)
              ,as.integer(value<AveExpr)
            )
          )
        )
    ) %>%
    select(-value,-type,-regulation,-AveExpr,-AveExpr1,-AveExpr2) %>%
    spread(feature,outcome)
}
```

```{r An example for computing of an individual DEG, eval=FALSE, include=FALSE}
deg(dea$amnion,'Amnion','outcome',direction=F)
deg(dea$amnion,'Amnion','outcome',direction=T)
deg(dea$amnion,'Maternal Blood','feature')
```

```{r}
timestamper[8]=as.character(now())
```

```{r Compute outcomes of individual DEGs, include=FALSE}
# Compute the individual DEGs
outcome=
  dea %>%
  lapply(X=names(.),Y=.,function(X,Y){
    Y[[X]] %>%
      deg(
        X %>%
          str_replace_all('_',' ') %>%
          str_to_title()
        ,'outcome'
        ,direction=F
      )
  }) %>%
  setNames(names(dea)) %>%
  lapply(X=1,Y=.,function(X,Y){
    Z=Y %>%
      lapply(colnames) %>%
      unlist() %>%
      unique() %>%
      .[!.%in%c('tissue_type','subject_id')]
    
    Y %>%
      lapply(X=names(.),Y=.,Z=Z,function(X,Y,Z){
        K=expand.grid(subject_id=Y[[X]]$subject_id,feature=Z)
        
        Y[[X]] %>%
          gather(feature,value,-tissue_type,-subject_id) %>%
          right_join(K,by=c('subject_id','feature')) %>%
          spread(feature,value) %>%
          .[,c('tissue_type','subject_id',Z)] %>%
          filter(!is.na(tissue_type))
      })
  }) %>%
  .[[1]] %>%
  do.call(rbind,.) %>%
  unite(title,tissue_type,subject_id,sep='_')

# Join the individual DEGs to the expression set
outcome=
  GSE73685 %>%
  `phenoData<-`(
    phenoData(.) %>%
      pData() %>%
      rownames_to_column(var='id') %>%
      left_join(outcome,by='title') %>%
      column_to_rownames(var='id') %>%
      .[,colnames(.) %>% .[!.%in%colnames(phenoData(GSE73685))]] %>%
      AnnotatedDataFrame()
  ) %>%
  pData()

# Filter only genes with 80% of minor-outcome sizes, greater than or equal to 20
outcome=
  outcome %>%
  .[
    ,pull(
      outcome %>%
        rownames_to_column(var='id') %>%
        gather(feature,value,-id) %>%
        group_by(feature,value) %>%
        summarize(n=n()) %>%
        ungroup() %>%
        mutate(
          value=
            case_when(
              value==0~'nonevent'
              ,value==1~'event'
              ,TRUE~'missing'
            )
        ) %>%
        spread(value,n) %>%
        mutate(minimum=ifelse(event<=nonevent,event,nonevent)) %>%
        filter(floor(0.8*minimum)>=20) %>%
        arrange(desc(minimum))
      ,feature
    )
  ]
```

```{r Compute features of individual DEGs, include=FALSE}
# Compute the individual DEGs
features=
  dea %>%
  lapply(deg,'Maternal Blood','feature') %>%
  lapply(X=1,Y=.,function(X,Y){
    Z=Y %>%
      lapply(colnames) %>%
      unlist() %>%
      unique() %>%
      .[!.%in%c('tissue_type','subject_id')]
    
    Y %>%
      lapply(X=names(.),Y=.,Z=Z,function(X,Y,Z){
        K=expand.grid(subject_id=Y[[X]]$subject_id,feature=Z)
        
        Y[[X]] %>%
          gather(feature,value,-tissue_type,-subject_id) %>%
          right_join(K,by=c('subject_id','feature')) %>%
          spread(feature,value) %>%
          .[,c('tissue_type','subject_id',Z)] %>%
          filter(!is.na(tissue_type)) %>%
          mutate(
            tissue_type=
              X %>%
              str_replace_all('_',' ') %>%
              str_to_title()
          )
      })
  }) %>%
  .[[1]] %>%
  do.call(rbind,.) %>%
  unite(title,tissue_type,subject_id,sep='_')

# Join the individual DEGs to the expression set
features=
  GSE73685 %>%
  `phenoData<-`(
    phenoData(.) %>%
      pData() %>%
      rownames_to_column(var='id') %>%
      left_join(features,by='title') %>%
      column_to_rownames(var='id') %>%
      .[,colnames(.) %>% .[!.%in%colnames(phenoData(GSE73685))]] %>%
      AnnotatedDataFrame()
  ) %>%
  pData()
```

```{r}
timestamper[9]=as.character(now())
```

```{r Determine training parameters, include=FALSE}
# Create an empty list to save training parameters.
training_parameters=list()

# Define classification tuning to apply 5-fold cross validation.
training_parameters$tuning_trControl=
  trainControl(
    method='cv'
    ,number=5
    ,summaryFunction=twoClassSummary
    ,classProbs=T
    ,savePredictions=T
    ,allowParallel=T
  )

# Define classification training to apply 30-time bootstrapping.
training_parameters$final_trControl=
  trainControl(
    method='boot'
    ,number=30
    ,summaryFunction=twoClassSummary
    ,classProbs=T
    ,savePredictions=T
    ,allowParallel=T
  )
```

```{r Create functions for resampled dimensional reduction, include=FALSE}
# Create a function for fitting the resampled dimensional reduction weights
rsdr=function(data,sd_cutoff=0,state=33,sample.kind=sample.kind){
    
    set.seed(state,sample.kind=sample.kind)
    idx_list=
      round(seq(1,nrow(data),len=10)) %>%
      lapply(X=2:length(.),Y=.,function(X,Y){
        seq(ifelse(X==2,1,Y[X-1]+1),Y[X])
      })
    idx_list=
      idx_list%>%
      lapply(X=seq(10),Y=.,function(X,Y){
          Y[-X] %>% unlist
      })
    dimr_func=prcomp
    rotm_name='rotation'
    prefix='PC'
    
    dimr_exec=function(idx_list,data){
      rs_data=
        data %>%
        .[idx_list,,drop=F]
      
      rs_nzv=
        rs_data %>%
        gather() %>%
        group_by(key) %>%
        summarize(sd_value=sd(value,na.rm=T),.groups='drop') %>%
        filter(sd_value>sd_cutoff)
      
      rs_data=rs_data[,rs_nzv$key,drop=F]
      avg=summarize_all(rs_data,mean) %>% gather()
      avg=setNames(avg$value,avg$key)
      std=summarize_all(rs_data, sd) %>% gather()
      std=setNames(std$value,std$key)
      
      dmr=
        as.matrix(rs_data) %>%
        sweep(2,avg,'-') %>%
        sweep(2,std,'/') %>%
        dimr_func()
      
      dmr$rotm=
        dmr[[rotm_name]] %>%
        `dimnames<-`(list(names(avg),paste0(prefix,seq(ncol(.)))))
      
      list(avg=avg,std=std,dmr=dmr["rotm"])
    }
    
    rsdr_object=
      idx_list %>%
      pblapply(
        FUN=dimr_exec
        ,data
      )
    
    gc()
    
    range_dim=
      rsdr_object %>%
      sapply(function(x)dim(x$dmr$rotm)[2]) %>%
      range() %>%
      .[!is.na(.)] %>%
      unique()
    
    list(
      rsdr_object=rsdr_object
      ,input_dim=colnames(data), 
      rs_method='CV'
      ,rs_number='10'
      ,dr_method='PCA', 
      sd_cutoff=sd_cutoff
      ,state=state
      ,min_output_dim=range_dim[1]
      ,max_output_dim=
        range_dim %>%
        .[length(.)]
      ,dimr_func=dimr_func
      ,rotm_name=rotm_name
      ,prefix=prefix
    )
}

# Create a function to build inputs for dimension transformation
transformator=function(rsdr_object){
  avg_rotm=
    rsdr_object %>%
    composition() %>%
    as.matrix()
  
  scaler=
    rsdr_object$rsdr_object %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      data.frame(
          input_dim=names(Y[[X]]$avg)
          ,avg=Y[[X]]$avg
          ,std=Y[[X]]$std
        ) %>%
        `rownames<-`(NULL)
    }) %>%
    do.call(rbind, .)
  
  scaler=
    scaler %>%
    group_by(input_dim) %>%
    summarize_all(function(x)
      mean(x,na.rm=T)
    ) %>%
    ungroup() %>%
    column_to_rownames(var='input_dim')
  
  list(avg_rotm=avg_rotm,scaler=scaler)
}

# Create a function for dimension transformation
transformation=function(tidy_set,transformator_object){
  
  avg_rotm=transformator_object$avg_rotm
  scaler=transformator_object$scaler
  
  scaled_data=
    tidy_set %>%
    exprs() %>%
    t() %>%
    as.data.frame() %>%
    cbind(
      pData(tidy_set) %>%
        mutate(outcome=as.numeric(outcome=='event'))
    ) %>%
    select(outcome,everything()) %>%
    .[,rownames(avg_rotm)] %>%
    sweep(2,scaler$avg,'-') %>%
    sweep(2,scaler$std,'/') %>%
    as.matrix()
  
  converted_data=scaled_data %*% avg_rotm
  converted_data=as.data.frame(converted_data)
  
  top_n=ncol(converted_data)
  
  dr_var=apply(converted_data,2,var)
  
  pve=dr_var/sum(dr_var)
  
  top_n_pve_dimensions=
    pve %>%
    sort(decreasing=T) %>%
    names() %>%
    .[seq(top_n)]
  
  converted_data=converted_data[,top_n_pve_dimensions]
  
  ExpressionSet(
    assayData=
      converted_data %>% t()
    ,phenoData=
      tidy_set %>% 
      pData() %>%
      cbind(
        tidy_set %>%
          exprs() %>%
          t() %>%
          as.data.frame()
      ) %>% 
      AnnotatedDataFrame(
        varMetadata=
          tidy_set %>%
          phenoData() %>% 
          varMetadata() %>%
          rbind(
            tidy_set %>%
              rownames() %>%
              data.frame(rowname = ., labelDescription = NA) %>%
              column_to_rownames(var = "rowname"))
      )
    ,featureData=
      avg_rotm %>%
      as.data.frame() %>%
      .[rownames(tidy_set)[rownames(tidy_set) %in% rownames(.)]
        ,colnames(converted_data)[colnames(converted_data) %in% colnames(.)]
        ,drop=F] %>%
      t() %>%
      as.data.frame() %>% 
      AnnotatedDataFrame(
        varMetadata=
          tidy_set %>%
          fData()
      ) %>%
      `varMetadata<-`(
        varMetadata(.) %>%
          select(labelDescription,everything())
      )
    ,experimentData=
      tidy_set %>%
      `preproc<-`(
        preproc(.) %>%
          c(list(pve=pve))
      ) %>% 
      experimentData()
    ,annotation=tidy_set %>% annotation()
    ,protocolData=tidy_set %>% protocolData())
}
```

```{r Create a function to train a surrogate transcriptome model, include=FALSE}
deg_train=function(deg
                   ,method
                   ,calib_method
                   ,tuning_trControl
                   ,final_trControl
                   ,tuningGrid
                   ,path
                   ,GSE73685=GSE73685
                   ,outcome=outcome
                   ,features=features
                   ,sample.kind=sample.kind
                   ,rsdr=rsdr
                   ,transformator=transformator
                   ,transformation=transformation){
  
  # component=
  #   list(
  #     deg='BABAM2'
  #     ,method='glmnet'
  #     ,calib_method='gamLoess'
  #     ,training_parameters=training_parameters
  #     ,tuningGrid=expand.grid(alpha=seq(0,1,len=5),lambda=10^seq(-9,0,len=5))
  #     ,path='data/deg_models/'
  #   )
  
  component=list()

  component$deg=deg
  component$method=method
  component$calib_method=calib_method
  component$training_parameters=
    list(
      tuning_trControl=tuning_trControl
      ,final_trControl=final_trControl
    )
  component$tuning=tuningGrid
  component$path=path
  
  component$data=
    GSE73685 %>%
    phenoData() %>%
    pData() %>%
    rownames_to_column(var='id') %>%
    select(id,title) %>%
    separate(title,c('tissue_type','subject_id'),sep='_') %>%
    select(-subject_id) %>%
    right_join(
      features %>%
        rownames_to_column(var='id') %>%
        left_join(
          outcome[,component$deg,drop=F] %>%
            setNames('outcome') %>%
            rownames_to_column(var='id')
          ,by='id'
        ) %>%
        mutate(deg=component$deg) %>%
        select(id,outcome,everything())
      ,by='id'
    ) %>%
    mutate(
      outcome=
        case_when(
          outcome==0~'nonevent'
          ,outcome==1~'event'
          ,TRUE~'missing'
        )
    )
  
  component$class=
    component$data %>%
    select(outcome,tissue_type) %>%
    table() %>%
    as.data.frame() %>%
    rename(n=Freq) %>%
    left_join(
      group_by(.,outcome) %>%
        summarize(subtotal=sum(n))
      ,by='outcome'
    ) %>%
    cbind(summarize(.,total=sum(n))) %>%
    mutate(
      weight=
        ifelse(
          outcome=='missing'
          ,NA
          ,1/(subtotal/total)*0.5
        )
    )
  
  component$data=
    component$data %>%
    left_join(
      component$class %>%
        select(tissue_type,outcome,weight)
      ,by=c('tissue_type','outcome')
    ) %>%
    select(tissue_type,outcome,weight,everything()) %>%
    column_to_rownames(var='id')
  
  component$train=
    component$data %>%
    rownames_to_column(var='id') %>%
    filter(outcome!='missing')
  
  component$features=
    suppressWarnings(
      component$train %>%
        select(-id) %>%
        gather(feature,value,-tissue_type,-outcome,-weight) %>%
        group_by(feature,outcome,tissue_type) %>%
        summarize(sd=sd(value,na.rm=T)) %>%
        ungroup() %>%
        filter(!(is.na(sd) | sd==0)) %>%
        filter(
          feature %in%
            pull(
              group_by(.,feature,outcome) %>%
                summarize(n=n()) %>%
                spread(outcome,n) %>%
                # filter(event==7 & nonevent==7)
                filter(event>=1 & nonevent>=1)
              ,feature
            )
        ) %>%
        mutate(
          tissue_type=
            tissue_type %>%
            str_replace_all('\\s+','_') %>%
            str_to_lower()
        ) %>%
        # unite(strata,outcome,tissue_type,sep=':') %>%
        # spread(strata,sd)
        spread(outcome,sd)
    )
  
  component$train=
    component$train %>%
    mutate(outcome=factor(outcome,c('event','nonevent'))) %>%
    column_to_rownames(var='id') %>%
    select_at(
      c('tissue_type','outcome','weight'
        ,component$features$feature %>% .[!duplicated(.)])
    ) %>%
    .[,sapply(.,function(x)sum(is.na(x))==0)]
  
  component$features=
    component$features %>%
    filter(feature%in%colnames(component$train))
  
  component$train_tissue_type=
    component$train %>%
    select(outcome,tissue_type) %>%
    table() %>%
    as.data.frame() %>%
    spread(outcome,Freq) %>%
    filter(event>=3 & nonevent>=3) %>%
    pull(tissue_type) %>%
    str_replace_all('\\s+','_') %>%
    str_to_lower()
  
  if(length(component$features$feature %>% .[!duplicated(.)])>1
     & length(component$train_tissue_type)>=1
     & floor(min(table(component$train$outcome))/3)>=2){
    
    component$transformator=
      component$train %>%
      rownames_to_column(var='id') %>%
      select(-tissue_type,-weight) %>%
      column_to_rownames(var='id') %>%
      select(-outcome) %>%
      rsdr(sample.kind=sample.kind) %>%
      transformator()
    
    component$pc=
      ExpressionSet(
        assayData=
          component$train %>%
          select(-tissue_type,-outcome,-weight) %>%
          t()
        ,phenoData=
          component$train %>%
          select(tissue_type,outcome,weight) %>%
          AnnotatedDataFrame()
      ) %>%
      .[rownames(component$transformator$scaler),] %>%
      transformation(transformator_object=component$transformator)
    
    component$pc=
      component$pc %>%
      .[component$pc %>%
          preproc() %>%
          .$pve %>%
          sort(decreasing=T) %>%
          .[1:floor(min(table(component$train$outcome))/3)] %>%
          names()
        ,] %>%
      exprs() %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column(var='id') %>%
      left_join(
        component$pc %>%
          phenoData() %>%
          pData() %>%
          select(tissue_type,outcome,weight) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      select(outcome,tissue_type,weight,everything()) %>%
      column_to_rownames(var='id')
    
    component$model=
      suppressWarnings(caret::train(
        outcome~.
        ,data=
          component$pc %>%
          rownames_to_column(var='id') %>%
          select(-tissue_type,-weight) %>%
          column_to_rownames(var='id')
        ,method=component$method
        ,weights=
          component$pc %>%
          rownames_to_column(var='id') %>%
          select(id,weight) %>%
          column_to_rownames(var='id') %>%
          pull(weight) %>%
          setNames(rownames(component$train))
        ,metric='ROC'
        ,trControl=component$training_parameters$tuning_trControl
        ,tuneGrid=component$tuning
      ))

    component$model=
      suppressWarnings(caret::train(
        outcome~.
        ,data=
          component$pc %>%
          rownames_to_column(var='id') %>%
          select(-tissue_type,-weight) %>%
          column_to_rownames(var='id')
        ,method=component$method
        ,weights=
          component$pc %>%
          rownames_to_column(var='id') %>%
          select(id,weight) %>%
          column_to_rownames(var='id') %>%
          pull(weight) %>%
          setNames(rownames(component$train))
        ,metric='ROC'
        ,trControl=component$training_parameters$final_trControl
        ,tuneGrid=component$model$bestTune
      ))
    
    component$calib_model=
      component$train_tissue_type %>%
      lapply(function(x){
        x=x %>%
          str_replace_all('_',' ') %>%
          str_to_title()
        
        y=component$pc %>%
          rownames_to_column(var='id') %>%
          filter(tissue_type==x) %>%
          select(-tissue_type,-weight) %>%
          column_to_rownames(var='id') %>%
          cbind(predict(component$model,newdata=.,type='prob')) %>%
          rownames_to_column(var='id') %>%
          select(id,outcome,event) %>%
          rename(prob=event) %>%
          column_to_rownames(var='id')
        
        if(sd(y$prob)==0){
          NULL
        }else{
          suppressWarnings(caret::train(
            outcome~.
            ,data=y
            ,method=component$calib_method
            ,weights=
              component$pc %>%
              rownames_to_column(var='id') %>%
              filter(tissue_type==x) %>%
              select(id,weight) %>%
              column_to_rownames(var='id') %>%
              lapply(X=1,Y=.,function(X,Y){
                Y %>%
                  pull(weight) %>%
                  setNames(rownames(Y))
              }) %>%
              .[[1]]
            ,metric='ROC'
            ,trControl=component$training_parameters$final_trControl
            ,family=binomial(link='logit')
          ))
        }
      }) %>%
      setNames(component$train_tissue_type)

    component$eval_model=
      component$train_tissue_type %>%
      lapply(function(x){
        if(is.null(component$calib_model[[x]])){
          NULL
        }else{
          x2=x %>%
            str_replace_all('_',' ') %>%
            str_to_title()
          
          component$pc %>%
            rownames_to_column(var='id') %>%
            filter(tissue_type==x2) %>%
            select(-tissue_type,-weight) %>%
            column_to_rownames(var='id') %>%
            cbind(predict(component$model,newdata=.,type='prob')) %>%
            rownames_to_column(var='id') %>%
            select(id,outcome,event) %>%
            rename(prob=event) %>%
            column_to_rownames(var='id') %>%
            cbind(
              suppressWarnings(
                  predict(component$calib_model[[x]],newdata=.,type='prob')
                ) %>%
                rename_all(function(x)paste0(x,'2'))
            ) %>%
            select(-prob) %>%
            rename_all(function(x)str_remove_all(x,'2')) %>%
            mutate(obs=outcome) %>%
            select(-outcome) %>%
            evalm(silent=T,showplots=F)
        }
      }) %>%
      setNames(component$train_tissue_type)
    
    if(sum(component$eval_model %>% sapply(function(x)!is.null(x)))>0){
      
      saveRDS(component,paste0(component$path,component$deg,'.rds'))
      
      rm(component)
      gc()
      
      deg
    }else{
      rm(component)
      gc()
      
      NULL
    }
  }else{
    rm(component)
    gc()
    
    NULL
  }
  
}
```

```{r Train surrogate transcriptome models for the genes, include=FALSE}
if(run_heavy_computation){
  cat('Conduct DEG modeling\n')
  cat('Started:',as.character(now()),'\n')
  
  dir.create('data/deg_models/',showWarnings=F)
  
  # Define the method
  the_method='glmnet'
  
  # Create multiple CPU clusters for parallel computing.
  cl=detectCores()-2
  cl=makePSOCKcluster(cl)
  registerDoParallel(cl)
  
  # Close the CPU clusters and clean up memory at exit.
  on.exit(stopCluster(cl))
  on.exit(registerDoSEQ())
  on.exit(rm(cl))
  on.exit(rm(the_method))
  on.exit(gc())
  
  # Bring these packages to the CPU clusters.
  if(the_method=='glmnet'){
    clusterEvalQ(cl,{
      library('parallel')
      library('doParallel')
      library('tidyverse')
      library('Biobase')
      library('pbapply')
      library('lubridate')
      library('rsdr')
      library('caret')
      library('glmnet')
      library('gam')
      library('MLeval')
    })
  }else if(the_method=='Rborist'){
    clusterEvalQ(cl,{
      library('parallel')
      library('doParallel')
      library('tidyverse')
      library('Biobase')
      library('pbapply')
      library('lubridate')
      library('rsdr')
      library('caret')
      library('Rborist')
      library('gam')
      library('MLeval')
    })
  }else if(the_method=='gbm'){
    clusterEvalQ(cl,{
      library('parallel')
      library('doParallel')
      library('tidyverse')
      library('Biobase')
      library('pbapply')
      library('lubridate')
      library('rsdr')
      library('caret')
      library('gbm')
      library('gam')
      library('MLeval')
    })
  }
    
    deg_models=
      colnames(outcome) %>%
      pblapply(
        deg_train
        ,method='glmnet'
        ,calib_method='gamLoess'
        ,tuning_trControl=training_parameters$tuning_trControl
        ,final_trControl=training_parameters$final_trControl
        ,tuningGrid=expand.grid(alpha=seq(0,1,len=5),lambda=10^seq(-9,0,len=5))
        ,path='data/deg_models/'
        ,GSE73685=GSE73685
        ,outcome=outcome
        ,features=features
        ,sample.kind=sample.kind
        ,rsdr=rsdr
        ,transformator=transformator
        ,transformation=transformation
        ,cl=cl
      )
    
    deg_models %>%
      unlist() %>%
      saveRDS('data/deg_models.rds')
    
    cat('End:',as.character(now()))
    
    readRDS('data/deg_models.rds') %>%
      pblapply(function(x){
        readRDS(paste0('data/deg_models/',x,'.rds')) %>%
          .$eval_model %>%
          lapply(X=names(.),Y=.,function(X,Y){
            Y[[X]]$optres$Group1[c('AUC-ROC','MCC'),] %>%
              rownames_to_column(var='metric') %>%
              mutate(tissue_type=X) %>%
              rename_all(str_to_lower)
          }) %>%
          do.call(rbind,.) %>%
          mutate(deg=x)
      }) %>%
      do.call(rbind,.) %>%
      saveRDS('data/opt_pred_deg.rds')
    
    opt_pred_deg=
      readRDS('data/opt_pred_deg.rds') %>%
      filter(metric=='MCC') %>%
      select(-metric,-ci) %>%
      rename(mcc=score) %>%
      lapply(X=1,Y=.,function(X,Y){
        lapply(X=Y$tissue_type %>% .[!duplicated(.)],Y=.,function(X,Y){
            Y %>%
              filter(tissue_type==X) %>%
              select(-tissue_type)
          }) %>%
          setNames(Y$tissue_type %>% .[!duplicated(.)])
      }) %>%
      .[[1]]
    
    readRDS('data/deg_models.rds') %>%
      pblapply(function(x){
        
        y=readRDS(paste0('data/deg_models/',x,'.rds')) %>%
          .[c('transformator','model')]
        
        # Get weights from the elastic net regression model.
        z=coef(y$model$finalModel,y$model$bestTune$lambda) %>%
          as.matrix() %>%
          as.data.frame() %>%
          rownames_to_column(var='term') %>%
          
          # Clean up.
          mutate(term=str_remove_all(term,'\\`|\\(|\\)')) %>%
          setNames(c('term','estimate')) %>%
          filter(estimate!=0) %>%
          arrange(desc(abs(estimate)))
        
        k=z %>%
          filter(!term %in% c('Intercept')) %>%
          column_to_rownames(var='term')
        
        y$transformator$avg_rotm[,rownames(k),drop=F] %>%
          dimnames() %>%
          setNames(c('feature','dimension'))
        
      }) %>%
      setNames(readRDS('data/deg_models.rds')) %>%
      saveRDS('data/opt_feat_deg.rds')
    
    opt_feat_deg=
      readRDS('data/opt_feat_deg.rds') %>%
      lapply(function(x)x$feature) %>%
      Reduce(union,x=.) %>%
      data.frame(feature=.)
  
}else{
  cat(readRDS('data/log.rds')[['deg_models']])
  deg_models=readRDS('data/deg_models.rds')
  opt_pred_deg=
    readRDS('data/opt_pred_deg.rds') %>%
    filter(metric=='MCC') %>%
    select(-metric,-ci) %>%
    rename(mcc=score) %>%
    lapply(X=1,Y=.,function(X,Y){
      lapply(X=Y$tissue_type %>% .[!duplicated(.)],Y=.,function(X,Y){
          Y %>%
            filter(tissue_type==X) %>%
            select(-tissue_type)
        }) %>%
        setNames(Y$tissue_type %>% .[!duplicated(.)])
    }) %>%
    .[[1]]
  
  opt_feat_deg=
    readRDS('data/opt_feat_deg.rds') %>%
    lapply(function(x)x$feature) %>%
    Reduce(union,x=.) %>%
    data.frame(feature=.)
}
```

```{r MCCs of the models, echo=FALSE, fig.height=7.48031, fig.width=3.46457}
opt_pred_deg %>%
  lapply(X=names(.),Y=.,function(X,Y){
    Y[[X]] %>%
      mutate(tissue=X)
  }) %>%
  do.call(rbind,.) %>%
  mutate(
    tissue=
      case_when(
        tissue=='fundus'~'Fundus myometrium'
        ,tissue=='decidua'~'Decidua (maternal side)'
        ,tissue=='placenta'~'Placenta (fetal side)'
        ,tissue=='amnion'~'Amnion (inner)'
        ,tissue=='chorion'~'Chorion (outer)'
        ,tissue=='cord_blood'~'Cord (fetal) blood'
        ,tissue=='lower_segment'~'Lower-segment myometrium'
      ) %>%
      factor(c(
        'Fundus myometrium'
        ,'Decidua (maternal side)'
        ,'Placenta (fetal side)'
        ,'Amnion (inner)'
        ,'Chorion (outer)'
        ,'Cord (fetal) blood'
        ,'Lower-segment myometrium'
      ))
  ) %>%
  ggplot(aes(mcc)) +
  geom_col(
    data=
      opt_pred_deg %>%
      lapply(X=names(.),Y=.,function(X,Y){
        Y[[X]] %>%
          mutate(tissue=X)
      }) %>%
      do.call(rbind,.) %>%
      mutate(
        tissue=
          case_when(
            tissue=='fundus'~'Fundus myometrium'
            ,tissue=='decidua'~'Decidua (maternal side)'
            ,tissue=='placenta'~'Placenta (fetal side)'
            ,tissue=='amnion'~'Amnion (inner)'
            ,tissue=='chorion'~'Chorion (outer)'
            ,tissue=='cord_blood'~'Cord (fetal) blood'
            ,tissue=='lower_segment'~'Lower-segment myometrium'
          ) %>%
          factor(c(
            'Fundus myometrium'
            ,'Decidua (maternal side)'
            ,'Placenta (fetal side)'
            ,'Amnion (inner)'
            ,'Chorion (outer)'
            ,'Cord (fetal) blood'
            ,'Lower-segment myometrium'
          ))
      ) %>%
      mutate(mcc=round(mcc,2)) %>%
      group_by(tissue,mcc) %>%
      summarize(n=n()) %>%
      ungroup() %>%
      group_by(tissue) %>%
      mutate(n=n/mean(n))
    ,aes(y=n)
  ) +
  geom_density() +
  geom_text(
    data=
      opt_pred_deg %>%
      lapply(X=names(.),Y=.,function(X,Y){
        Y[[X]] %>%
          mutate(tissue=X)
      }) %>%
      do.call(rbind,.) %>%
      mutate(
        tissue=
          case_when(
            tissue=='fundus'~'Fundus myometrium'
            ,tissue=='decidua'~'Decidua (maternal side)'
            ,tissue=='placenta'~'Placenta (fetal side)'
            ,tissue=='amnion'~'Amnion (inner)'
            ,tissue=='chorion'~'Chorion (outer)'
            ,tissue=='cord_blood'~'Cord (fetal) blood'
            ,tissue=='lower_segment'~'Lower-segment myometrium'
          ) %>%
          factor(c(
            'Fundus myometrium'
            ,'Decidua (maternal side)'
            ,'Placenta (fetal side)'
            ,'Amnion (inner)'
            ,'Chorion (outer)'
            ,'Cord (fetal) blood'
            ,'Lower-segment myometrium'
          ))
      ) %>%
      filter(!is.nan(mcc)) %>%
      group_by(tissue) %>%
      summarize(n=n())
    ,aes(label=paste0(n,' genes'))
    ,x=0.3,y=3,hjust=0,vjust=0.5
    ,size=3
  ) +
  geom_errorbarh(
    data=
      opt_pred_deg %>%
      lapply(X=names(.),Y=.,function(X,Y){
        Y[[X]] %>%
          mutate(tissue=X)
      }) %>%
      do.call(rbind,.) %>%
      mutate(
        tissue=
          case_when(
            tissue=='fundus'~'Fundus myometrium'
            ,tissue=='decidua'~'Decidua (maternal side)'
            ,tissue=='placenta'~'Placenta (fetal side)'
            ,tissue=='amnion'~'Amnion (inner)'
            ,tissue=='chorion'~'Chorion (outer)'
            ,tissue=='cord_blood'~'Cord (fetal) blood'
            ,tissue=='lower_segment'~'Lower-segment myometrium'
          ) %>%
          factor(c(
            'Fundus myometrium'
            ,'Decidua (maternal side)'
            ,'Placenta (fetal side)'
            ,'Amnion (inner)'
            ,'Chorion (outer)'
            ,'Cord (fetal) blood'
            ,'Lower-segment myometrium'
          ))
      ) %>%
      filter(!is.nan(mcc)) %>%
      group_by(tissue) %>%
      summarize(
        lb=max(0,mean(mcc)-qnorm(0.975)*sd(mcc))
        ,ub=min(1,mean(mcc)+qnorm(0.975)*sd(mcc))
        ,mcc=mean(mcc)
        ,pos=-1
      )
    ,aes(y=pos,xmin=lb,xmax=ub)
  ) +
  geom_point(
    data=
      opt_pred_deg %>%
      lapply(X=names(.),Y=.,function(X,Y){
        Y[[X]] %>%
          mutate(tissue=X)
      }) %>%
      do.call(rbind,.) %>%
      mutate(
        tissue=
          case_when(
            tissue=='fundus'~'Fundus myometrium'
            ,tissue=='decidua'~'Decidua (maternal side)'
            ,tissue=='placenta'~'Placenta (fetal side)'
            ,tissue=='amnion'~'Amnion (inner)'
            ,tissue=='chorion'~'Chorion (outer)'
            ,tissue=='cord_blood'~'Cord (fetal) blood'
            ,tissue=='lower_segment'~'Lower-segment myometrium'
          ) %>%
          factor(c(
            'Fundus myometrium'
            ,'Decidua (maternal side)'
            ,'Placenta (fetal side)'
            ,'Amnion (inner)'
            ,'Chorion (outer)'
            ,'Cord (fetal) blood'
            ,'Lower-segment myometrium'
          ))
      ) %>%
      filter(!is.nan(mcc)) %>%
      group_by(tissue) %>%
      summarize(mcc=mean(mcc),pos=-1)
    ,aes(y=pos)
  ) +
  facet_wrap(~tissue,ncol=1) +
  scale_x_continuous("Matthew's correlation coefficient (MCC)") +
  scale_y_continuous('Frequency* bar and density† line') +
  theme(
    axis.title.x=element_text(size=unit(9,'pt'),family='sans')
    ,axis.text.x=element_text(size=unit(8,'pt'),family='sans')
    ,axis.title.y=element_text(size=unit(9,'pt'),family='sans')
    ,axis.text.y=element_text(size=unit(8,'pt'),family='sans')
    ,strip.text=element_text(size=unit(9,'pt'),family='sans')
  )

ggsave(
  filename="figure2.eps"
  ,device=cairo_ps
  ,width=90 # fig.height=7.48031, fig.width=3.46457
  ,height=190/88*90
  ,unit='mm'
)
```

```{r Characterize the number of surrogate vs. available genes, echo=FALSE}
dea %>%
  lapply(X=names(.),Y=.,function(X,Y){
    Y[[X]]$result %>%
      mutate(
        regulation=
          ifelse(
            adj.P.Val>=0.05
            ,'non-DEG'
            ,ifelse(
              logFC>=0
              ,ifelse(logFC>2,'strong up','weak up')
              ,ifelse(logFC<(-2),'strong down','weak down')
            )
          )
      ) %>%
      rownames_to_column(var='deg') %>%
      left_join(
        opt_pred_deg[[X]]
        ,by='deg'
      ) %>%
      mutate(pred=ifelse(!is.na(mcc),'pred1','pred0')) %>%
      select(-mcc) %>%
      group_by(regulation,pred) %>%
      summarize(n=n()) %>%
      ungroup() %>%
      mutate(comparison=paste0(X,' vs. maternal blood'))
  }) %>%
  do.call(rbind,.) %>%
  spread(pred,n,fill=0) %>%
  rbind(
    mutate(.,regulation='total') %>%
      group_by(regulation,comparison) %>%
      summarize_all(sum)
  ) %>%
  mutate(
    pred=pred0+pred1
    ,pred_p=paste0('(',round(pred1/pred*100,2),')')
  ) %>%
  select(-pred0) %>%
  unite(pred_f,pred1,pred,sep='/') %>%
  unite(pred,pred_f,pred_p,sep=' ') %>%
  spread(regulation,pred,fill=0) %>%
  select(
    comparison
    ,`strong up`,`weak up`
    ,`weak down`,`strong down`
    ,`total`
    ,everything()
  )
```

```{r Create a function for Q-Q normalization targeting controls, include=FALSE}
dea_qn=function(exprs_mat,the_dea_result){
  exprs_mat %>%
    normalize.quantiles.use.target(
      target=
        the_dea_result[rownames(.),'AveExpr1',drop=F] %>%
          pull(AveExpr1)
    ) %>%
    `dimnames<-`(dimnames(exprs_mat))
}
```

```{r Plot the datasets before the normalization, eval=FALSE, include=FALSE}
rbind(
    GSE108497 %>%
      exprs() %>%
      t() %>%
      as.data.frame() %>%
      .[1:24,,drop=F]
    
    ,GSE73685 %>%
      .[,phenoData(dea$amnion$eset) %>%
          pData() %>%
          filter(str_detect(str_to_lower(title),'maternal blood')) %>%
          rownames()] %>%
      exprs() %>%
      dea_qn(dea$amnion$result) %>%
      t() %>%
      as.data.frame()
    
    ,GSE85307 %>%
      exprs() %>%
      t() %>%
      as.data.frame() %>%
      .[1:24,,drop=F]
    
    ,GSE86200 %>%
      exprs() %>%
      t() %>%
      as.data.frame() %>%
      .[1:24,,drop=F]
    
    ,GSE149437 %>%
      exprs() %>%
      t() %>%
      as.data.frame() %>%
      .[1:24,,drop=F]
  ) %>%
  t() %>%
  boxplot()
```

```{r Plot the datasets after the normalization, eval=FALSE, include=FALSE}
rbind(
    GSE108497 %>%
      exprs() %>%
      dea_qn(dea$amnion$result) %>%
      t() %>%
      as.data.frame() %>%
      .[1:24,,drop=F]
    
    ,GSE73685 %>%
      .[,phenoData(dea$amnion$eset) %>%
          pData() %>%
          filter(str_detect(str_to_lower(title),'maternal blood')) %>%
          rownames()] %>%
      exprs() %>%
      dea_qn(dea$amnion$result) %>%
      t() %>%
      as.data.frame()
    
    ,GSE85307 %>%
      exprs() %>%
      dea_qn(dea$amnion$result) %>%
      t() %>%
      as.data.frame() %>%
      .[1:24,,drop=F]
  ) %>%
  t() %>%
  boxplot()
```

```{r Create a function to derive the surrogate transcriptome, include=FALSE}
surrogate_deg=function(exprs_mat,the_dea_list1,component){
  
  # exprs_mat=
  #   GSE73685 %>%
  #   .[,phenoData(dea$amnion$eset) %>%
  #       pData() %>%
  #       filter(str_detect(str_to_lower(title),'maternal blood')) %>%
  #       rownames()] %>%
  #   exprs()
  # the_dea_list1=dea['amnion']
  # component=readRDS('data/deg_models/DIRAS2.rds')
  
  the_dea=the_dea_list1[[1]]
  the_tissue=names(the_dea_list1)[1]
  
  exprs_mat %>%
    dea_qn(the_dea$result) %>%
    .[rownames(component$transformator$avg_rotm),] %>%
    as.data.frame() %>%
    rownames_to_column(var='feature') %>%
    left_join(
      the_dea$result %>%
        rownames_to_column(var='feature') %>%
        filter(adj.P.Val<0.05) %>%
        select(feature,AveExpr,AveExpr1,AveExpr2)
      ,by='feature'
    ) %>%
    gather(id,value,-feature,-AveExpr,-AveExpr1,-AveExpr2) %>%
    mutate(
      norm_value=
        ifelse(
          AveExpr1<AveExpr
          ,ifelse(
            value<AveExpr
            ,0-(value-AveExpr)/(AveExpr1-AveExpr)
            ,0+(value-AveExpr)/(AveExpr2-AveExpr)
          )
          ,ifelse(
            value>AveExpr
            ,0-(value-AveExpr)/(AveExpr1-AveExpr)
            ,0+(value-AveExpr)/(AveExpr2-AveExpr)
          )
        )
    ) %>%
    select(-value,-AveExpr,-AveExpr1,-AveExpr2) %>%
    spread(feature,norm_value) %>%
    column_to_rownames(var='id') %>%
    t() %>%
    ExpressionSet(
      phenoData=
        t(.) %>%
        as.data.frame() %>%
        mutate(outcome=NA) %>%
        select(outcome) %>%
        AnnotatedDataFrame()
    ) %>%
    transformation(transformator_object=component$transformator) %>%
    exprs() %>%
    t() %>%
    as.data.frame() %>%
    cbind(predict(component$model,newdata=.,type='prob')) %>%
    select(event) %>%
    rename(prob=event) %>%
    cbind(suppressWarnings(
      predict(component$calib_model[[the_tissue]],newdata=.,type='prob')
    )) %>%
    rename(calib_prob=event) %>%
    select(calib_prob) %>%
    setNames(paste0(the_tissue,':',component$deg))
}
```

```{r An example for deriving a surrogate transcript, eval=FALSE, include=FALSE}
GSE73685 %>%
  .[,phenoData(dea$amnion$eset) %>%
      pData() %>%
      filter(str_detect(str_to_lower(title),'maternal blood')) %>%
      rownames()] %>%
  exprs() %>%
  surrogate_deg(dea['amnion'],readRDS('data/deg_models/TRIM24.rds'))
```

```{r Create surrogate transcriptomes of GSE108497, include=FALSE}
if(run_heavy_computation){
  cat('Create surrogate transcriptomes of GSE108497\n')
  cat('Started:',as.character(now()),'\n')
  
    GSE108497_surrogate=
      opt_pred_deg %>%
      lapply(X=names(.),Y=.,Z=exprs(GSE108497),function(X,Y,Z){
        cat('Create surrogate transcriptome of',str_remove_all(X,'_'),'\n')
        L=Y[[X]] %>%
          pblapply(X=seq(nrow(.)),Y=.,Z=X,K=Z,function(X,Y,Z,K){
            L=surrogate_deg(
                K
                ,dea[Z]
                ,readRDS(paste0('data/deg_models/',Y$deg[X],'.rds'))
              )
            L[[1]]=(1+L[[1]]*Y$mcc[X])/2
            L
          }) %>%
          do.call(cbind,.)
        cat('\n')
        L
      }) %>%
      do.call(cbind,.)
  
  cat('End:',as.character(now()))
  saveRDS(GSE108497_surrogate,'data/GSE108497_surrogate.rds')
}else{
  cat(readRDS('data/log.rds')[['GSE108497_surrogate']])
  GSE108497_surrogate=readRDS('data/GSE108497_surrogate.rds')
}  
```

```{r Create surrogate transcriptomes of GSE85307, include=FALSE}
if(run_heavy_computation){
  cat('Create surrogate transcriptomes of GSE85307\n')
  cat('Started:',as.character(now()),'\n')
  
    GSE85307_surrogate=
      opt_pred_deg %>%
      lapply(X=names(.),Y=.,Z=exprs(GSE85307),function(X,Y,Z){
        cat('Create surrogate transcriptome of',str_remove_all(X,'_'),'\n')
        L=Y[[X]] %>%
          pblapply(X=seq(nrow(.)),Y=.,Z=X,K=Z,function(X,Y,Z,K){
            L=surrogate_deg(
                K
                ,dea[Z]
                ,readRDS(paste0('data/deg_models/',Y$deg[X],'.rds'))
              )
            L[[1]]=(1+L[[1]]*Y$mcc[X])/2
            L
          }) %>%
          do.call(cbind,.)
        cat('\n')
        L
      }) %>%
      do.call(cbind,.)
  
  cat('End:',as.character(now()))
  saveRDS(GSE85307_surrogate,'data/GSE85307_surrogate.rds')
}else{
  cat(readRDS('data/log.rds')[['GSE85307_surrogate']])
  GSE85307_surrogate=readRDS('data/GSE85307_surrogate.rds')
}  
```

```{r Create surrogate transcriptomes of GSE86200, include=FALSE}
if(run_heavy_computation){
  cat('Create surrogate transcriptomes of GSE86200\n')
  cat('Started:',as.character(now()),'\n')
  
    GSE86200_surrogate=
      opt_pred_deg %>%
      lapply(X=names(.),Y=.,Z=exprs(GSE86200),function(X,Y,Z){
        cat('Create surrogate transcriptome of',str_remove_all(X,'_'),'\n')
        L=Y[[X]] %>%
          pblapply(X=seq(nrow(.)),Y=.,Z=X,K=Z,function(X,Y,Z,K){
            L=surrogate_deg(
                K
                ,dea[Z]
                ,readRDS(paste0('data/deg_models/',Y$deg[X],'.rds'))
              )
            L[[1]]=(1+L[[1]]*Y$mcc[X])/2
            L
          }) %>%
          do.call(cbind,.)
        cat('\n')
        L
      }) %>%
      do.call(cbind,.)
  
  cat('End:',as.character(now()))
  saveRDS(GSE86200_surrogate,'data/GSE86200_surrogate.rds')
}else{
  cat(readRDS('data/log.rds')[['GSE86200_surrogate']])
  GSE86200_surrogate=readRDS('data/GSE86200_surrogate.rds')
}  
```

```{r Create surrogate transcriptomes of GSE149437, include=FALSE}
if(run_heavy_computation){
  cat('Create surrogate transcriptomes of GSE149437\n')
  cat('Started:',as.character(now()),'\n')
  
    GSE149437_surrogate=
      opt_pred_deg %>%
      lapply(X=names(.),Y=.,Z=exprs(GSE149437),function(X,Y,Z){
        cat('Create surrogate transcriptome of',str_remove_all(X,'_'),'\n')
        L=Y[[X]] %>%
          pblapply(X=seq(nrow(.)),Y=.,Z=X,K=Z,function(X,Y,Z,K){
            L=surrogate_deg(
                K
                ,dea[Z]
                ,readRDS(paste0('data/deg_models/',Y$deg[X],'.rds'))
              )
            L[[1]]=(1+L[[1]]*Y$mcc[X])/2
            L
          }) %>%
          do.call(cbind,.)
        cat('\n')
        L
      }) %>%
      do.call(cbind,.)
  
  cat('End:',as.character(now()))
  saveRDS(GSE149437_surrogate,'data/GSE149437_surrogate.rds')
}else{
  cat(readRDS('data/log.rds')[['GSE149437_surrogate']])
  GSE149437_surrogate=readRDS('data/GSE149437_surrogate.rds')
}  
```

```{r}
now()
```

```{r Conduct DEA for maternal blood genes of GSE108497, include=FALSE}
# Determine the true outcome from the phenotype data
GSE108497_phenotype=
  GSE108497 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(colnames(.) %>% str_remove_all(':ch1')) %>%
  filter(
    apl==0
    & lac==0
    & fd==0
    & nnd==0
  ) %>%
  mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
  mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
  filter(
    time_point
    %in%c('<16 weeks','16-23 weeks','24-31 weeks','32-40 weeks')
  ) %>%
  filter(
    (early==0 & pe==0 & sga==0)
    |
    (early==1 & pe==1 & sga==0)
    |
    (early==0 & pe==1 & sga==0)
    |
    (early==0 & pe==0 & sga==1)
  ) %>%
  mutate(
    diagnosis=
      case_when(
        (early==0 & pe==0 & sga==0)~'normotensive-AGA'
        ,(early==1 & pe==1 & sga==0)~'EOPE'
        ,(early==0 & pe==1 & sga==0)~'LOPE'
        ,(early==0 & pe==0 & sga==1)~'FGR-SGA'
        ,TRUE~'other'
      )
  ) %>%
  mutate(
    outcome=
      ifelse(
        diagnosis%in%c('EOPE','LOPE')
        ,'event'
        ,'nonevent'
      ) %>%
      factor(c('event','nonevent'))
  ) %>%
  select(outcome,diagnosis,time_point,everything())

# Conduct the DEA
GSE108497_maternal_blood=
  GSE108497 %>%
  .[rownames(.) %>% .[.%in%opt_feat_deg$feature]
    ,rownames(GSE108497_phenotype)] %>%
  `phenoData<-`(
    GSE108497_phenotype %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r Characterize the true outcome of GSE108497, echo=FALSE}
GSE108497_phenotype %>%
  group_by(outcome,diagnosis,time_point) %>%
  summarize(n=n()) %>%
  spread(time_point,n,fill=0)
```

```{r Get the average expressions of GSE108497 for all Q-Q norm, include=FALSE}
GSE108497_AveExpr1=
  GSE108497_maternal_blood$result %>%
  rownames_to_column(var='feature') %>%
  filter(adj.P.Val<0.05 & feature%in%opt_feat_deg$feature) %>%
  select(feature,AveExpr1) %>%
  mutate(feature=paste0('maternal_blood',':',feature)) %>%
  rbind(
    dea %>%
    lapply(X=names(.),Y=.,function(X,Y){
      Y[[X]]$result %>%
        .[opt_pred_deg[[X]]$deg,] %>%
        rownames_to_column(var='feature') %>%
        mutate(feature=paste0(X,':',feature)) %>%
        select(feature,AveExpr1)
    }) %>%
    do.call(rbind,.)
  ) %>%
  column_to_rownames(var='feature')
```

```{r Get the features of GSE108497, include=FALSE}
GSE108497_features=
  GSE108497 %>%
  .[,rownames(GSE108497_phenotype)] %>%
  .[GSE108497_maternal_blood$result %>%
      rownames_to_column(var='feature') %>%
      filter(adj.P.Val<0.05 & feature%in%opt_feat_deg$feature) %>%
      pull(feature),] %>%
  exprs() %>%
  t() %>%
  as.data.frame() %>%
  rename_all(function(x)paste0('maternal_blood',':',x)) %>%
  cbind(
    GSE108497_surrogate %>%
    .[rownames(GSE108497_phenotype),]
  )
```

```{r Construct the development dataset, include=FALSE}
train_set=
  GSE108497_phenotype %>%
  select(outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    GSE108497_features %>%
      t() %>%
      dea_qn(the_dea_result=GSE108497_AveExpr1) %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  column_to_rownames(var='id')
```

```{r Show processing of the development dataset, eval=FALSE, include=FALSE}
GSE108497_features %>%
  t() %>%
  .[,1:24] %>%
  boxplot()

GSE108497_features %>%
  t() %>%
  dea_qn(the_dea_result=GSE108497_AveExpr1) %>%
  .[,1:24] %>%
  boxplot()

train_set %>%
  select(-outcome) %>%
  t() %>%
  .[,1:24] %>%
  boxplot()
```

```{r DEA of all the features of the development dataset, include=FALSE}
GSE108497_all_tissue=
  ExpressionSet(
    assayData=
      train_set %>%
      select(-outcome) %>%
      t()
    ,phenoData=
      train_set %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE108497_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(train_set),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r DEA of blood transcripts of the development dataset, include=FALSE}
GSE108497_sel_mat_blood=
  ExpressionSet(
    assayData=
      train_set %>%
      select(-outcome) %>%
      .[,str_detect(colnames(.),'maternal_blood')] %>%
      t()
    ,phenoData=
      train_set %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE108497_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(train_set),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r DEA of surrogate transcripts of the development dataset, include=FALSE}
GSE108497_repro_tissue=
  ExpressionSet(
    assayData=
      train_set %>%
      select(-outcome) %>%
      .[,!str_detect(colnames(.),'maternal_blood')] %>%
      t()
    ,phenoData=
      train_set %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE108497_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(train_set),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r Conduct DEA for maternal blood genes of GSE85307, include=FALSE}
# Determine the true outcome from the phenotype data
GSE85307_phenotype=
  GSE85307 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=
      as.numeric(blood_draw_at_gestation_week)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
    ,early=ifelse(gestation_weeks<=34,'Early_Onset','Late_Onset')
  ) %>%
  unite(diagnosis,early,pregnancy_condition) %>%
  # filter(`vitamind_25(0h)_ng/ml_whole_blood_baseline`<30) %>%
  mutate(
    outcome=
      ifelse(
        diagnosis%in%c('Early_Onset_Preeclampsia','Late_Onset_Preeclampsia')
        ,'event'
        ,'nonevent'
      ) %>%
      factor(c('event','nonevent'))
  ) %>%
  select(outcome,diagnosis,time_point,everything())

# Conduct the DEA
GSE85307_maternal_blood=
  GSE85307 %>%
  .[rownames(.) %>% .[.%in%opt_feat_deg$feature]
    ,rownames(GSE85307_phenotype)] %>%
  `phenoData<-`(
    GSE85307_phenotype %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r Characterize the true outcome of GSE85307, echo=FALSE}
GSE85307_phenotype %>%
  group_by(outcome,diagnosis,time_point) %>%
  summarize(n=n()) %>%
  spread(time_point,n,fill=0)
```

```{r Get the features of GSE85307, include=FALSE}
GSE85307_features=
  GSE85307 %>%
  .[,rownames(GSE85307_phenotype)] %>%
  .[GSE108497_maternal_blood$result %>%
      rownames_to_column(var='feature') %>%
      filter(adj.P.Val<0.05 & feature%in%opt_feat_deg$feature) %>%
      pull(feature),] %>%
  exprs() %>%
  t() %>%
  as.data.frame() %>%
  rename_all(function(x)paste0('maternal_blood',':',x)) %>%
  cbind(
    GSE85307_surrogate %>%
    .[rownames(GSE85307_phenotype),]
  )
```

```{r Construct the replication dataset 1, include=FALSE}
test_set1=
  GSE85307_phenotype %>%
  select(outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    GSE85307_features %>%
      t() %>%
      dea_qn(the_dea_result=GSE108497_AveExpr1) %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  column_to_rownames(var='id')
```

```{r Show processing of the replication dataset 1, eval=FALSE, include=FALSE}
GSE85307_features %>%
  t() %>%
  .[,1:24] %>%
  boxplot()

GSE85307_features %>%
  t() %>%
  dea_qn(the_dea_result=GSE108497_AveExpr1) %>%
  .[,1:24] %>%
  boxplot()

test_set1 %>%
  select(-outcome) %>%
  t() %>%
  .[,1:24] %>%
  boxplot()
```

```{r DEA of all the features of the replication dataset 1, include=FALSE}
GSE85307_all_tissue=
  ExpressionSet(
    assayData=
      test_set1 %>%
      select(-outcome) %>%
      t()
    ,phenoData=
      test_set1 %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE85307_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(test_set1),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r DEA of blood transcripts of the replication dataset 1, include=FALSE}
GSE85307_sel_mat_blood=
  ExpressionSet(
    assayData=
      test_set1 %>%
      select(-outcome) %>%
      .[,str_detect(colnames(.),'maternal_blood')] %>%
      t()
    ,phenoData=
      test_set1 %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE85307_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(test_set1),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r DEA of surrogate transcripts of the replication dataset 1, include=FALSE}
GSE85307_repro_tissue=
  ExpressionSet(
    assayData=
      test_set1 %>%
      select(-outcome) %>%
      .[,!str_detect(colnames(.),'maternal_blood')] %>%
      t()
    ,phenoData=
      test_set1 %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE85307_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(test_set1),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r Conduct DEA for maternal blood genes of GSE86200, include=FALSE}
# Determine the true outcome from the phenotype data
GSE86200_phenotype=
  GSE86200 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=
      floor(as.numeric(gestational_day_at_enrollment)/7)
    ,blood_draw_at_gestation_week=
      ifelse(str_detect(title,'32-38'),34,blood_draw_at_gestation_week)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
    ,trt=
      ifelse(
        str_detect(treatment_group,'Intervention')
        ,'Treated','Untreated'
      )
  ) %>%
  unite(diagnosis,pregnancy_condition,trt) %>%
  mutate(
    outcome=
      ifelse(
        diagnosis%in%c('Preeclampsia_Untreated','Preeclampsia_Treated')
        ,'event'
        ,'nonevent'
      ) %>%
      factor(c('event','nonevent'))
  ) %>%
  select(outcome,diagnosis,time_point,everything())

# Conduct the DEA
GSE86200_maternal_blood=
  GSE86200 %>%
  .[rownames(.) %>% .[.%in%opt_feat_deg$feature]
    ,rownames(GSE86200_phenotype)] %>%
  `phenoData<-`(
    GSE86200_phenotype %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r Characterize the true outcome of GSE86200, echo=FALSE}
GSE86200_phenotype %>%
  group_by(outcome,diagnosis,time_point) %>%
  summarize(n=n()) %>%
  spread(time_point,n,fill=0)
```

```{r Get the features of GSE86200, include=FALSE}
GSE86200_features=
  GSE86200 %>%
  .[,rownames(GSE86200_phenotype)] %>%
  .[GSE108497_maternal_blood$result %>%
      rownames_to_column(var='feature') %>%
      filter(adj.P.Val<0.05 & feature%in%opt_feat_deg$feature) %>%
      pull(feature),] %>%
  exprs() %>%
  t() %>%
  as.data.frame() %>%
  rename_all(function(x)paste0('maternal_blood',':',x)) %>%
  cbind(
    GSE86200_surrogate %>%
    .[rownames(GSE86200_phenotype),]
  )
```

```{r Construct the replication dataset 2, include=FALSE}
test_set2=
  GSE86200_phenotype %>%
  select(outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    GSE86200_features %>%
      t() %>%
      dea_qn(the_dea_result=GSE108497_AveExpr1) %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  column_to_rownames(var='id')
```

```{r Show processing of the replication dataset 2, eval=FALSE, include=FALSE}
GSE86200_features %>%
  t() %>%
  .[,1:24] %>%
  boxplot()

GSE86200_features %>%
  t() %>%
  dea_qn(the_dea_result=GSE108497_AveExpr1) %>%
  .[,1:24] %>%
  boxplot()

test_set2 %>%
  select(-outcome) %>%
  t() %>%
  .[,1:24] %>%
  boxplot()
```

```{r DEA of all the features of the replication dataset 2, include=FALSE}
GSE86200_all_tissue=
  ExpressionSet(
    assayData=
      test_set2 %>%
      select(-outcome) %>%
      t()
    ,phenoData=
      test_set2 %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE86200_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(test_set2),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r DEA of blood transcripts of the replication dataset 2, include=FALSE}
GSE86200_sel_mat_blood=
  ExpressionSet(
    assayData=
      test_set2 %>%
      select(-outcome) %>%
      .[,str_detect(colnames(.),'maternal_blood')] %>%
      t()
    ,phenoData=
      test_set2 %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE86200_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(test_set2),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r DEA of surrogate transcripts of the replication dataset 2, include=FALSE}
GSE86200_repro_tissue=
  ExpressionSet(
    assayData=
      test_set2 %>%
      select(-outcome) %>%
      .[,!str_detect(colnames(.),'maternal_blood')] %>%
      t()
    ,phenoData=
      test_set2 %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE86200_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(test_set2),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r Conduct DEA for maternal blood genes of GSE149437, include=FALSE}
# Determine the true outcome from the phenotype data
GSE149437_phenotype=
  GSE149437 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  mutate(
    blood_draw_at_gestation_week=as.numeric(gestational_age)
    ,time_point=case_when(
      between(ceiling(blood_draw_at_gestation_week),0,15)~'<16 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),16,23)~'16-23 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),24,31)~'24-31 weeks'
      ,between(ceiling(blood_draw_at_gestation_week),32,40)~'32-40 weeks'
      ,TRUE~''
    )
  ) %>%
  rename(diagnosis=group) %>%
  # filter(diagnosis%in%c('Early_Preeclampsia','Control')) %>%
  mutate(
    outcome=
      ifelse(
        diagnosis%in%c('Early_Preeclampsia')
        ,'event'
        ,'nonevent'
      ) %>%
      factor(c('event','nonevent'))
  ) %>%
  select(outcome,diagnosis,time_point,everything())

# Conduct the DEA
GSE149437_maternal_blood=
  GSE149437 %>%
  .[rownames(.) %>% .[.%in%opt_feat_deg$feature]
    ,rownames(GSE149437_phenotype)] %>%
  `phenoData<-`(
    GSE149437_phenotype %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r Characterize the true outcome of GSE149437, echo=FALSE}
GSE149437_phenotype %>%
  group_by(outcome,diagnosis,time_point) %>%
  summarize(n=n()) %>%
  spread(time_point,n,fill=0)
```

```{r Get the features of GSE149437, include=FALSE}
GSE149437_features=
  GSE149437 %>%
  .[,rownames(GSE149437_phenotype)] %>%
  .[GSE108497_maternal_blood$result %>%
      rownames_to_column(var='feature') %>%
      filter(adj.P.Val<0.05 & feature%in%opt_feat_deg$feature) %>%
      pull(feature),] %>%
  exprs() %>%
  t() %>%
  as.data.frame() %>%
  rename_all(function(x)paste0('maternal_blood',':',x)) %>%
  cbind(
    GSE149437_surrogate %>%
    .[rownames(GSE149437_phenotype),]
  )
```

```{r Construct the replication dataset 3, include=FALSE}
test_set3=
  GSE149437_phenotype %>%
  select(outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    GSE149437_features %>%
      t() %>%
      dea_qn(the_dea_result=GSE108497_AveExpr1) %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  column_to_rownames(var='id')
```

```{r Show processing of the replication dataset 3, eval=FALSE, include=FALSE}
GSE149437_features %>%
  t() %>%
  .[,1:24] %>%
  boxplot()

GSE149437_features %>%
  t() %>%
  dea_qn(the_dea_result=GSE108497_AveExpr1) %>%
  .[,1:24] %>%
  boxplot()

test_set3 %>%
  select(-outcome) %>%
  t() %>%
  .[,1:24] %>%
  boxplot()
```

```{r DEA of all the features of the replication dataset 3, include=FALSE}
GSE149437_all_tissue=
  ExpressionSet(
    assayData=
      test_set3 %>%
      select(-outcome) %>%
      t()
    ,phenoData=
      test_set3 %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE149437_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(test_set3),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r DEA of blood transcripts of the replication dataset 3, include=FALSE}
GSE149437_sel_mat_blood=
  ExpressionSet(
    assayData=
      test_set3 %>%
      select(-outcome) %>%
      .[,str_detect(colnames(.),'maternal_blood')] %>%
      t()
    ,phenoData=
      test_set3 %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE149437_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(test_set3),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r DEA of surrogate transcripts of the replication dataset 3, include=FALSE}
GSE149437_repro_tissue=
  ExpressionSet(
    assayData=
      test_set3 %>%
      select(-outcome) %>%
      .[,!str_detect(colnames(.),'maternal_blood')] %>%
      t()
    ,phenoData=
      test_set3 %>%
      select(outcome) %>%
      rownames_to_column(var='id') %>%
      left_join(
        GSE149437_phenotype %>%
          select(time_point) %>%
          rownames_to_column(var='id')
        ,by='id'
      ) %>%
      column_to_rownames(var='id') %>%
      .[rownames(test_set3),] %>%
      AnnotatedDataFrame()
  ) %>%
  conduct_dea(ignore_timepoint=T)
```

```{r Determine the true outcome from the phenotype of GSE177477, include=FALSE}
GSE177477_phenotype=
  GSE177477 %>%
  phenoData() %>%
  pData() %>%
  select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
  setNames(
    colnames(.) %>%
      str_remove_all(':ch1') %>%
      str_replace_all('\\s','_')
  ) %>%
  separate(
    comorbids
    ,c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
    ,sep='\\|'
  ) %>%
  mutate_at(
    c('asthma','cad','caf','t2dm','htn','ihd','sp_cabg')
    ,function(x) ifelse(x=='',0,ifelse(is.na(x),0,1))
  ) %>%
  mutate(
    severity=
      factor(
        ifelse(
          is.na(severity)
          ,'uninfected'
          ,severity
        )
        ,c('uninfected','mild','severe')
      )
    ,sex=
      factor(
        ifelse(
          is.na(sex)
          ,'U'
          ,sex
        )
        ,c('U','F','M')
      )
  ) %>%
  mutate(
    outcome=
      ifelse(
        severity%in%c('severe')
        ,'event'
        ,'nonevent'
      ) %>%
      factor(c('event','nonevent'))
  ) %>%
  select(outcome,severity,sex,everything())
```

```{r Summarize DEGs of all the datasets, echo=FALSE}
list(
    GSE108497=GSE108497_maternal_blood
    ,GSE85307=GSE85307_maternal_blood
    ,GSE86200=GSE86200_maternal_blood
    ,GSE149437=GSE149437_maternal_blood
    ,`GSE108497 and GSE149437`=
      GSE108497_maternal_blood$result %>%
      rownames_to_column(var='gene') %>%
      filter(adj.P.Val<0.05) %>%
      filter(
        gene%in%rownames(filter(
          GSE149437_maternal_blood$result
          ,adj.P.Val<0.05
        ))
      ) %>%
      column_to_rownames(var='gene') %>%
      list(result=.)
  ) %>%
  lapply(X=names(.),Y=.,function(X,Y){
    Y[[X]]$result %>%
      mutate(
        regulation=
          ifelse(
            adj.P.Val>=0.05
            ,'non-DEG'
            ,ifelse(
              logFC>=0
              ,ifelse(logFC>2,'strong up','weak up')
              ,ifelse(logFC<(-2),'strong down','weak down')
            )
          )
      ) %>%
      rownames_to_column(var='deg') %>%
      group_by(regulation) %>%
      summarize(n=n()) %>%
      ungroup() %>%
      mutate(set=X)
  }) %>%
  do.call(rbind,.) %>%
  rbind(
    mutate(.,regulation='total') %>%
      group_by(regulation,set) %>%
      summarize_all(sum)
  ) %>%
  spread(regulation,n,fill=0) %>%
  select(
    set
    ,`weak up`
    ,`weak down`,`strong down`
    ,`total`
    ,everything()
  )
```

```{r Explore DEGs of all the datasets, include=FALSE}
tmp0=
  list(
    GSE108497=GSE108497_maternal_blood
    ,GSE85307=GSE85307_maternal_blood
    ,GSE86200=GSE86200_maternal_blood
    ,GSE149437=GSE149437_maternal_blood
    ,`GSE108497 and GSE149437`=
      GSE108497_maternal_blood$result %>%
      rownames_to_column(var='gene') %>%
      filter(adj.P.Val<0.05) %>%
      filter(
        gene%in%rownames(filter(
          GSE149437_maternal_blood$result
          ,adj.P.Val<0.05
        ))
      ) %>%
      column_to_rownames(var='gene') %>%
      list(result=.)
  ) %>%
  lapply(X=names(.),Y=.,function(X,Y){
    Y[[X]]$result %>%
      mutate(
        regulation=
          ifelse(
            adj.P.Val>=0.05
            ,'non-DEG'
            ,ifelse(
              logFC>=0
              ,ifelse(logFC>2,'strong up','weak up')
              ,ifelse(logFC<(-2),'strong down','weak down')
            )
          )
      ) %>%
      rownames_to_column(var='deg') %>%
      filter(regulation!='non-DEG') %>%
      mutate(set=X)
  }) %>%
  setNames(
    c('GSE108497'
      ,'GSE85307'
      ,'GSE86200'
      ,'GSE149437'
      ,'GSE108497 and GSE149437')
  )
```

```{r}
timestamper[10]=as.character(now())
```

```{r Create a function to train a non-DIVNN prediction model, include=FALSE}
disease_train=function(data
                       ,method
                       ,calib_method
                       ,tuning_trControl
                       ,final_trControl
                       ,tuningGrid
                       ,calib_tuningGrid
                       ,epv=20
                       ,sample.kind=sample.kind){
  
  component=list()
  
  component$data=data
  component$method=method
  component$calib_method=calib_method
  component$training_parameters=
    list(
      tuning_trControl=tuning_trControl
      ,final_trControl=final_trControl
    )
  component$tuning=tuningGrid
  component$calib_tuning=calib_tuningGrid
  
  suppressWarnings(set.seed(33,sample.kind=sample.kind))
  component$data=
    component$data %>%
    rownames_to_column(var='id') %>%
    mutate(
      partition=
        suppressWarnings(
          sapply(X=seq(nrow(.)),Y=outcome,function(X,Y){
            Z=createDataPartition(Y,p=0.8)[[1]]
            ifelse(X%in%Z,'train','calib')
          })
        )
    ) %>%
    select(id,partition,everything())
  
  component$class=
    component$data %>%
    select(outcome,partition) %>%
    table() %>%
    as.data.frame() %>%
    rename(n=Freq) %>%
    left_join(
      group_by(.,outcome) %>%
        summarize(subtotal=sum(n))
      ,by='outcome'
    ) %>%
    cbind(summarize(.,total=sum(n))) %>%
    mutate(
      weight=
        ifelse(
          outcome=='missing'
          ,NA
          ,1/(subtotal/total)*0.5
        )
    )
  
  component$data=
    component$data %>%
    left_join(
      component$class %>%
        select(partition,outcome,weight)
      ,by=c('partition','outcome')
    ) %>%
    select(partition,outcome,weight,everything()) %>%
    column_to_rownames(var='id')
  
  component$train=
    component$data %>%
    rownames_to_column(var='id') %>%
    filter(partition=='train' & outcome!='missing') %>%
    select(-partition) %>%
    column_to_rownames(var='id')
  
  component$calib=
    component$data %>%
    rownames_to_column(var='id') %>%
    filter(partition=='calib' & outcome!='missing') %>%
    select(-partition) %>%
    column_to_rownames(var='id')
  
  component$transformator=
    component$train %>%
    rownames_to_column(var='id') %>%
    select(-weight) %>%
    column_to_rownames(var='id') %>%
    select(-outcome) %>%
    rsdr(sample.kind=sample.kind) %>%
    transformator()
  
  component$pc=
    ExpressionSet(
      assayData=
        component$train %>%
        select(-outcome,-weight) %>%
        t()
      ,phenoData=
        component$train %>%
        select(outcome,weight) %>%
        AnnotatedDataFrame()
    ) %>%
    .[rownames(component$transformator$scaler),] %>%
    transformation(transformator_object=component$transformator)
  
  component$pc=
    component$pc %>%
    .[component$pc %>%
        preproc() %>%
        .$pve %>%
        sort(decreasing=T) %>%
        .[1:floor(min(table(component$train$outcome))/epv)] %>%
        names()
      ,] %>%
    exprs() %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column(var='id') %>%
    left_join(
      component$pc %>%
        phenoData() %>%
        pData() %>%
        select(outcome,weight) %>%
        rownames_to_column(var='id')
      ,by='id'
    ) %>%
    select(outcome,weight,everything()) %>%
    column_to_rownames(var='id')
  
  component$pc2=
    ExpressionSet(
      assayData=
        component$calib %>%
        select(-outcome,-weight) %>%
        t()
      ,phenoData=
        component$calib %>%
        select(outcome,weight) %>%
        AnnotatedDataFrame()
    ) %>%
    .[rownames(component$transformator$scaler),] %>%
    transformation(transformator_object=component$transformator)
  
  component$pc2=
    component$pc2 %>%
    .[component$pc %>%
        colnames() %>%
        .[!.%in%c('outcome','weight')]
      ,] %>%
    exprs() %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column(var='id') %>%
    left_join(
      component$pc2 %>%
        phenoData() %>%
        pData() %>%
        select(outcome,weight) %>%
        rownames_to_column(var='id')
      ,by='id'
    ) %>%
    select(outcome,weight,everything()) %>%
    column_to_rownames(var='id')
  
  data=
    component$pc %>%
    rownames_to_column(var='id') %>%
    select(-weight) %>%
    column_to_rownames(var='id')
  
  data2=
    component$pc2 %>%
    rownames_to_column(var='id') %>%
    select(-weight) %>%
    column_to_rownames(var='id')
  
  component$model=
    suppressWarnings(caret::train(
      outcome~.
      ,data=data
      ,method=component$method
      ,weights=
        component$train %>%
        rownames_to_column(var='id') %>%
        select(id,weight) %>%
        column_to_rownames(var='id') %>%
        pull(weight) %>%
        setNames(rownames(component$train))
      ,metric='ROC'
      ,trControl=component$training_parameters$tuning_trControl
      ,tuneGrid=component$tuning
    ))
  
  component$model=
    suppressWarnings(caret::train(
      outcome~.
      ,data=data
      ,method=component$method
      ,weights=
        component$train %>%
        rownames_to_column(var='id') %>%
        select(id,weight) %>%
        column_to_rownames(var='id') %>%
        pull(weight) %>%
        setNames(rownames(component$train))
      ,metric='ROC'
      ,trControl=component$training_parameters$final_trControl
      ,tuneGrid=component$model$bestTune
    ))
  
  component$calib_model=
    suppressWarnings(caret::train(
      outcome~.
      ,data=
        data2 %>%
          cbind(predict(component$model,newdata=.,type='prob')) %>%
          rownames_to_column(var='id') %>%
          select(id,outcome,event) %>%
          rename(prob=event) %>%
          column_to_rownames(var='id')
      ,method=component$calib_method
      ,weights=
        component$calib %>%
        rownames_to_column(var='id') %>%
        select(id,weight) %>%
        column_to_rownames(var='id') %>%
        pull(weight) %>%
        setNames(rownames(component$calib))
      ,metric='ROC'
      ,trControl=component$training_parameters$final_trControl
      ,tuneGrid=component$calib_tuning
      ,family=binomial(link='logit')
    ))
  
  component$pc3=
    ExpressionSet(
      assayData=
        component$data[,colnames(component$train)] %>%
        rownames_to_column(var='id') %>%
        filter(outcome!='missing') %>%
        select(-weight) %>%
        column_to_rownames(var='id') %>%
        select(-outcome) %>%
        t()
      ,phenoData=
        component$data[,colnames(component$train)] %>%
        select(outcome,weight) %>%
        AnnotatedDataFrame()
    ) %>%
    .[rownames(component$transformator$scaler),] %>%
    transformation(transformator_object=component$transformator)
  
  component$pc3=
    component$pc3 %>%
    .[component$pc %>%
        colnames() %>%
        .[!.%in%c('outcome','weight')]
      ,] %>%
    exprs() %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column(var='id') %>%
    left_join(
      component$pc3 %>%
        phenoData() %>%
        pData() %>%
        select(outcome,weight) %>%
        rownames_to_column(var='id')
      ,by='id'
    ) %>%
    select(outcome,weight,everything()) %>%
    column_to_rownames(var='id')
  
  component$eval_model=
    component$pc3 %>%
    rownames_to_column(var='id') %>%
    filter(outcome!='missing') %>%
    select(-weight) %>%
    column_to_rownames(var='id') %>%
    cbind(predict(component$model,newdata=.,type='prob')) %>%
    rownames_to_column(var='id') %>%
    select(id,outcome,event) %>%
    rename(prob=event) %>%
    column_to_rownames(var='id') %>%
    cbind(
      suppressWarnings(predict(component$calib_model,newdata=.,type='prob')) %>%
        rename_all(function(x)paste0(x,'2'))
    ) %>%
    select(-prob) %>%
    rename_all(function(x)str_remove_all(x,'2')) %>%
    mutate(obs=outcome) %>%
    select(-outcome) %>%
    select(nonevent,everything()) %>%
    evalm(silent=T,showplots=F)
  
  rm(data,data2)
  
  component
}
```

```{r Create a function to infer clique-extracted ontology, include=FALSE}
clixo=function(similarity
               ,alpha=0.01
               ,beta=0.5
               ,feature_name='feature'
               ,onto_prefix='CliXO'
               ,os='windows'){
  
  if(os=='windows'){
    clixo_path='clixo_0.3/clixo_0.3_windows'
  }else{
    clixo_path='clixo_0.3'
  }

  system(paste(c(
    'git clone https://github.com/herdiantrisufriyana/clixo_0.3/'
  ),collapse=' '))
  on.exit({
    system('rm -r clixo_0.3')
  })
  
  similarity %>%
    as.data.frame() %>%
    rownames_to_column(var='source') %>%
    gather(target,similarity,-source) %>%
    .[lower.tri(similarity) %>%
        as.logical()
      ,] %>%
    arrange(desc(similarity)) %>%
    write_tsv(paste0('clixo_0.3/input.tsv',collapse=''),col_names=F)
  
  if(os=='windows'){
    filecon=file('clixo_0.3/ontology.cx')
    on.exit({
        system('rm -r clixo_0.3')
        close(filecon)
    })
    
    suppressWarnings(system(paste(c(
        paste0(clixo_path,'/clixo')
        ,'clixo_0.3/input.tsv'
        ,alpha
        ,beta
        ,feature_name
      ),collapse=' '),intern=T)) %>%
      paste(collapse='\n') %>% 
      writeLines(filecon)
  }else{
    system(paste(c(
      'bash -c'
      ,paste0('\"',clixo_path,'/clixo')
      ,'clixo_0.3/input.tsv'
      ,alpha
      ,beta
      ,feature_name
      ,'> clixo_0.3/ontology.cx\"'
    ),collapse=' '))
  }
  
  cx=
    suppressMessages(suppressWarnings(read_tsv(
      'clixo_0.3/ontology.cx'
      ,col_names=
        c('target','source','relation','similarity')
      ,col_types=
        list(col_character(),col_character(),col_character(),col_double())
      ,skip=
        sum(
          1
          ,suppressMessages(suppressWarnings(read_tsv(
              'clixo_0.3/ontology.cx'
            ))) %>%
            setNames("column") %>%
            pull(column) %>%
            str_detect('#') %>% 
            which() %>%
            max()
        )
    ))) %>%
    mutate(
      relation=ifelse(relation=='default','is_a',feature_name)
      ,source=ifelse(relation=='is_a',paste0('CliXO:',source),source)
      ,target=paste0(onto_prefix,':',target)
    ) %>%
    select(source,target,similarity,relation)
  
  cx$target=
    paste0(
      onto_prefix
      ,':'
      ,str_pad(
        as.numeric(gsub(paste0(onto_prefix,':'),'',cx$target))
        ,str_count(max(as.numeric(gsub(paste0(onto_prefix,':'),'',cx$target))))
        ,'left'
        ,'0'
      )
    )
  
  cx$source[cx$relation=='is_a']=
    paste0(
      onto_prefix
      ,':'
      ,str_pad(
        as.numeric(gsub(
          paste0(onto_prefix,':')
          ,''
          ,cx$source[cx$relation=='is_a']
        ))
        ,str_count(max(as.numeric(gsub(
          paste0(onto_prefix,':'),'',cx$source[cx$relation=='is_a']
        ))))
        ,'left'
        ,'0'
      )
    )
  
  cx
}
```

```{r Create a function to prepare for a predictive modeling, include=FALSE}
model_component=function(data
                         ,method
                         ,calib_method
                         ,tuning_trControl
                         ,final_trControl
                         ,tuningGrid
                         ,calib_tuningGrid){
  
  component=list()

  component$data=data
  component$method=method
  component$calib_method=calib_method
  component$training_parameters=
    list(
      tuning_trControl=tuning_trControl
      ,final_trControl=final_trControl
    )
  component$tuning=tuningGrid
  component$calib_tuning=calib_tuningGrid
  
  suppressWarnings(set.seed(33,sample.kind=sample.kind))
  component$data=
    component$data %>%
    rownames_to_column(var='id') %>%
    mutate(
      partition=
        suppressWarnings(
          sapply(X=seq(nrow(.)),Y=outcome,function(X,Y){
            Z=createDataPartition(Y,p=0.8)[[1]]
            ifelse(X%in%Z,'train','calib')
          })
        )
    ) %>%
    select(id,partition,everything())
  
  component$class=
    component$data %>%
    select(outcome,partition) %>%
    table() %>%
    as.data.frame() %>%
    rename(n=Freq) %>%
    left_join(
      group_by(.,outcome) %>%
        summarize(subtotal=sum(n))
      ,by='outcome'
    ) %>%
    cbind(summarize(.,total=sum(n))) %>%
    mutate(
      weight=
        ifelse(
          outcome=='missing'
          ,NA
          ,1/(subtotal/total)*0.5
        )
    )
  
  component$data=
    component$data %>%
    left_join(
      component$class %>%
        select(partition,outcome,weight)
      ,by=c('partition','outcome')
    ) %>%
    select(partition,outcome,weight,everything()) %>%
    column_to_rownames(var='id')
  
  component$train=
    component$data %>%
    rownames_to_column(var='id') %>%
    filter(partition=='train' & outcome!='missing') %>%
    select(-partition) %>%
    column_to_rownames(var='id')
  
  component$calib=
    component$data %>%
    rownames_to_column(var='id') %>%
    filter(partition=='calib' & outcome!='missing') %>%
    select(-partition) %>%
    column_to_rownames(var='id')
  
  component
}
```

```{r DI-VNN feature selection and representation, include=FALSE}
divnn_pre_object=function(data,save=T,path=NULL){
  # Set input list
  input=list()
  
  # Use provider-wise, pre-calibrated, training set for modeling.
  input$data=data
  
  # Set output list
  output=list()
  
  # Get uncensored outcome.
  output$outcome=
    input$data %>%
    select(outcome)
  
  # Get features corresponding to uncensored outcome.
  output$raw=
    input$data %>%
    select(-outcome) %>%
    as.matrix()
  
  # Check if feature selection was conducted utilizing differential analysis
  # by moderated-t stats with Benjamini-Hochberg multiple testing correction.
  output$fit=
    t(output$raw) %>%
    normalize.quantiles() %>%
    `dimnames<-`(dimnames(t(output$raw))) %>%
    lmFit(model.matrix(~outcome,output$outcome)) %>%
    eBayes() %>%
    topTable(coef=2,nrow(.$coefficients),adjust.method='BH',sort.by='none') %>%
    filter(adj.P.Val<0.05)
  
  # Get the selected features as unnormalized ones.
  output$unnorm=
    output$raw %>%
    .[,rownames(output$fit)]
  
  # Normalize features quantile-to-quantile
  # using differential average of features as target.
  output$norm=
    output$unnorm %>%
    t() %>%
    normalize.quantiles.use.target(output$fit$AveExpr) %>%
    t() %>%
    `dimnames<-`(dimnames(output$unnorm))
  
  # Transform features by 1-bit stochastic gradient descent (SGD)
  # using the differential average.
  cat('Transform features by 1-bit stochastic gradient descent (SGD)\n')
  output$predictor_v=
    output$norm %>%
    sweep(2,output$fit$AveExpr,'-') %>%
    pbsapply(function(x){ifelse(x==0,0,ifelse(x>0,1,-1))}) %>%
    matrix(
      ncol=ncol(output$unnorm)
      ,byrow=F
      ,dimnames=dimnames(output$unnorm)
    ) %>%
    as.data.frame()
  
  # Get feature-wise mean of unnormalized features.
  output$predictor_m=
    output$unnorm %>%
    matrixStats::colMeans2()
  
  # Get feature-wise standard deviation (SD) of unnormalized features.
  output$predictor_s=
    output$unnorm %>%
    matrixStats::colSds()
  
  # Scale unnormalized features by the mean and SD.
  output$scaled=
    output$unnorm %>%
    sweep(2,output$predictor_m,'-') %>%
    sweep(2,output$predictor_s,'/')
  
  # Compute feature-feature Pearson correlation matrix
  # of unnormalized features.
  cat('Compute feature-feature Pearson correlation matrix\n')
  output$predictor_p=
    colnames(output$unnorm) %>%
    lapply(X=seq(length(.)-1),Y=.,function(X,Y){
      data.frame(
        predictor1=Y[X]
        ,predictor2=Y[(X+1):length(.)]
      )
    }) %>%
    do.call(rbind,.) %>%
    mutate(
      pearson=
        pbsapply(X=seq(nrow(.)),Y=.,Z=output$scaled,function(X,Y,Z){
          cor(
            Z[,Y$predictor1[X]]
            ,Z[,Y$predictor2[X]]
          )
        })
    ) %>%
    rbind(
      setNames(select(.,predictor2,predictor1,everything()),colnames(.))
      ,data.frame(
        predictor1=colnames(output$unnorm)
        ,predictor2=colnames(output$unnorm)
        ,pearson=1
      )
    ) %>%
    spread(predictor2,pearson) %>%
    column_to_rownames(var='predictor1') %>%
    as.matrix()
  
  # Conduct Barnes-Hut t-moderated stochastic neighbor embedding (t-SNE).
  cat('Conduct Barnes-Hut t-moderated stochastic neighbor embedding (t-SNE)\n')
  suppressWarnings(set.seed(33,sample.kind=sample.kind))
  output$predictor_tsne=
    output$predictor_p %>%
    Rtsne(dims=3,verbose=T,is_distance=T)
  
  # Construct clique-extracted ontology (CliXO) (choose your own OS).
  output$predictor_c=
    output$predictor_p %>%
    clixo(os='windows')
  
  # Compile input.
  input=list()
  
  input$value=output$predictor_v
  
  input$outcome=
    output$outcome %>%
    mutate(outcome=as.integer(outcome=='event')) %>%
    pull(outcome) %>%
    setNames(rownames(input$value))
  
  input$similarity=output$predictor_p
  
  input$mapping=
    output$predictor_tsne$Y %>%
    `rownames<-`(colnames(input$value))
  
  input$ontology=
    output$predictor_c %>%
    mutate(seq=seq(nrow(.))) %>%
    gather(key,value,-similarity,-relation,-seq) %>%
    mutate(
      value2=
        ifelse(
          str_detect(value,'CliXO:')
          ,str_remove_all(value,'CliXO:')
          ,NA
        ) %>%
        as.integer()
    ) %>%
    mutate(
      value=
        ifelse(
          str_detect(value,'CliXO:')
          ,paste0(
            'ONT:'
            ,str_pad(value2,str_count(max(value2,na.rm=T)),'left','0')
          )
          ,value
        )
    ) %>%
    select(-value2) %>%
    spread(key,value) %>%
    arrange(seq) %>%
    select(-seq) %>%
    select(source,target,everything())
  
  input$fit=output$fit
  
  # Compile into a TidySet.
  cat('Compile into a TidySet\n')
  output=
    TidySet.compile(
      value=input$value
      ,outcome=input$outcome
      ,similarity=input$similarity
      ,mapping=input$mapping
      ,ontology=input$ontology
      ,ranked=T
      ,dims=7
      ,decreasing=F
      ,seed_num=33
    )
  
  if(save){
    saveRDS(input,paste0(path,'_input.rds'))
    saveRDS(output,paste0(path,'_output.rds'))
  }
  
  list(input=input,output=output)
}
```

```{r Create a function to create training function given lambda, include=FALSE}
source('R/trainer_generator-function.R')
```

```{r Create a function to transform test data for DI-VNN input, include=FALSE}
source('R/test_transformer-function.R')
```

```{r Build a function for DI-VNN evaluation, include=FALSE}
source('R/eval_divnn-function.R')
```

```{r Build a function for DI-VNN calibration and evaluation, include=FALSE}
source('R/divnn_calibrator_evaluator-function.R')
```

```{r Create a function to refresh keras backend session, include=FALSE}
source('R/refresh_session-function.R')
refresh_session()
```

```{r Create empty lists for non-DIVNN prediction models, include=FALSE}
disease_model=list()
disease_model2=list()
disease_model3=list()
```

```{r Train PC-ENR using all the features, include=FALSE}
if(run_heavy_computation){
  disease_model$pc_elnet=
    train_set %>%
    .[,c('outcome'
         ,GSE108497_all_tissue$result %>%
            rownames_to_column(var='feature') %>%
            filter(adj.P.Val<0.05) %>%
            pull(feature))] %>%
    disease_train(
      method='glmnet'
      ,calib_method='gamLoess'
      ,tuning_trControl=training_parameters$tuning_trControl
      ,final_trControl=training_parameters$final_trControl
      ,tuningGrid=expand.grid(alpha=seq(0,1,len=5),lambda=10^seq(-9,0,len=5))
      ,calib_tuningGrid=
        expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))
      ,epv=10
      ,sample.kind=sample.kind
    )
  
  saveRDS(disease_model$pc_elnet,'data/pc_elnet.rds')
}else{
  cat(readRDS('data/log.rds')[['pc_elnet']])
  disease_model$pc_elnet=readRDS('data/pc_elnet.rds')
}
```

```{r Train PC-ENR using blood transcripts, include=FALSE}
if(run_heavy_computation){
  disease_model2$pc_elnet=
    train_set %>%
    .[,c('outcome'
         ,GSE108497_sel_mat_blood$result %>%
            rownames_to_column(var='feature') %>%
            filter(adj.P.Val<0.05) %>%
            pull(feature))] %>%
    disease_train(
      method='glmnet'
      ,calib_method='gamLoess'
      ,tuning_trControl=training_parameters$tuning_trControl
      ,final_trControl=training_parameters$final_trControl
      ,tuningGrid=expand.grid(alpha=seq(0,1,len=5),lambda=10^seq(-9,0,len=5))
      ,calib_tuningGrid=
        expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))
      ,epv=10
      ,sample.kind=sample.kind
    )
  
  saveRDS(disease_model2$pc_elnet,'data/pc_elnet2.rds')
}else{
  cat(readRDS('data/log.rds')[['pc_elnet2']])
  disease_model2$pc_elnet=readRDS('data/pc_elnet2.rds')
}
```

```{r Train PC-ENR using surrogate transcripts, include=FALSE}
if(run_heavy_computation){
  disease_model3$pc_elnet=
    train_set %>%
    .[,c('outcome'
         ,GSE108497_repro_tissue$result %>%
            rownames_to_column(var='feature') %>%
            filter(adj.P.Val<0.05) %>%
            pull(feature))] %>%
    disease_train(
      method='glmnet'
      ,calib_method='gamLoess'
      ,tuning_trControl=training_parameters$tuning_trControl
      ,final_trControl=training_parameters$final_trControl
      ,tuningGrid=expand.grid(alpha=seq(0,1,len=5),lambda=10^seq(-9,0,len=5))
      ,calib_tuningGrid=
        expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))
      ,epv=10
      ,sample.kind=sample.kind
    )
    # |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=02s
  
  saveRDS(disease_model3$pc_elnet,'data/pc_elnet3.rds')
}else{
  cat(readRDS('data/log.rds')[['pc_elnet3']])
  disease_model3$pc_elnet=readRDS('data/pc_elnet3.rds')
}
```

```{r Train PC-RF using all the features, include=FALSE}
if(run_heavy_computation){
  disease_model$pc_rf=
    train_set %>%
    .[,c('outcome'
         ,GSE108497_all_tissue$result %>%
            rownames_to_column(var='feature') %>%
            filter(adj.P.Val<0.05) %>%
            pull(feature))] %>%
    disease_train(
      method='Rborist'
      ,calib_method='gamLoess'
      ,tuning_trControl=training_parameters$tuning_trControl
      ,final_trControl=training_parameters$final_trControl
      ,tuningGrid=
        expand.grid(predFixed=seq(5,45,len=5),minNode=seq(20,100,len=5))
      ,calib_tuningGrid=
        expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))[-1,]
      ,epv=5
      ,sample.kind=sample.kind
    )
  
  saveRDS(disease_model$pc_rf,'data/pc_rf.rds')
}else{
  cat(readRDS('data/log.rds')[['pc_rf']])
  disease_model$pc_rf=readRDS('data/pc_rf.rds')
}
```

```{r Train PC-RF using blood transcripts, include=FALSE}
if(run_heavy_computation){
  disease_model2$pc_rf=
    train_set %>%
    .[,c('outcome'
         ,GSE108497_sel_mat_blood$result %>%
            rownames_to_column(var='feature') %>%
            filter(adj.P.Val<0.05) %>%
            pull(feature))] %>%
    disease_train(
      method='Rborist'
      ,calib_method='gamLoess'
      ,tuning_trControl=training_parameters$tuning_trControl
      ,final_trControl=training_parameters$final_trControl
      ,tuningGrid=expand.grid(predFixed=seq(5,45,len=5),minNode=seq(20,100,len=5))
      ,calib_tuningGrid=
        expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))[-1,]
      ,epv=5
      ,sample.kind=sample.kind
    )
  
  saveRDS(disease_model2$pc_rf,'data/pc_rf2.rds')
}else{
  cat(readRDS('data/log.rds')[['pc_rf2']])
  disease_model2$pc_rf=readRDS('data/pc_rf2.rds')
}
```

```{r Train PC-RF using surrogate transcripts, include=FALSE}
if(run_heavy_computation){
  disease_model3$pc_rf=
    train_set %>%
    .[,c('outcome'
         ,GSE108497_repro_tissue$result %>%
            rownames_to_column(var='feature') %>%
            filter(adj.P.Val<0.05) %>%
            pull(feature))] %>%
    disease_train(
      method='Rborist'
      ,calib_method='gamLoess'
      ,tuning_trControl=training_parameters$tuning_trControl
      ,final_trControl=training_parameters$final_trControl
      ,tuningGrid=
        expand.grid(predFixed=seq(5,45,len=5),minNode=seq(20,100,len=5))
      ,calib_tuningGrid=
        expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))[-1,]
      ,epv=5
      ,sample.kind=sample.kind
    )
  
  saveRDS(disease_model3$pc_rf,'data/pc_rf3.rds')
}else{
  cat(readRDS('data/log.rds')[['pc_rf3']])
  disease_model3$pc_rf=readRDS('data/pc_rf3.rds')
}
```

```{r Train PC-GBM using all the features, include=FALSE}
if(run_heavy_computation){
  disease_model$pc_gbm=
    train_set %>%
    .[,c('outcome'
         ,GSE108497_all_tissue$result %>%
            rownames_to_column(var='feature') %>%
            filter(adj.P.Val<0.05) %>%
            pull(feature))] %>%
    disease_train(
      method='gbm'
      ,calib_method='gamLoess'
      ,tuning_trControl=training_parameters$tuning_trControl
      ,final_trControl=training_parameters$final_trControl
      ,tuningGrid=
        expand.grid(
          interaction.depth=1
          ,shrinkage=seq(0.0005,0.05,len=25)
          ,n.minobsinnode=20
        ) %>%
        mutate(n.trees=seq(0,2500,len=26) %>% .[-1] %>% rev()) %>%
        select(n.trees,everything())
      ,calib_tuningGrid=
        expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))
      ,epv=10
      ,sample.kind=sample.kind
    )
  
  saveRDS(disease_model$pc_gbm,'data/pc_gbm.rds')
}else{
  cat(readRDS('data/log.rds')[['pc_gbm']])
  disease_model$pc_gbm=readRDS('data/pc_gbm.rds')
}
```

```{r Train PC-GBM using blood transcripts, include=FALSE}
if(run_heavy_computation){
  disease_model2$pc_gbm=
    train_set %>%
    .[,c('outcome'
         ,GSE108497_sel_mat_blood$result %>%
            rownames_to_column(var='feature') %>%
            filter(adj.P.Val<0.05) %>%
            pull(feature))] %>%
    disease_train(
      method='gbm'
      ,calib_method='gamLoess'
      ,tuning_trControl=training_parameters$tuning_trControl
      ,final_trControl=training_parameters$final_trControl
      ,tuningGrid=
        expand.grid(
          interaction.depth=1
          ,shrinkage=seq(0.0005,0.05,len=25)
          ,n.minobsinnode=20
        ) %>%
        mutate(n.trees=seq(0,2500,len=26) %>% .[-1] %>% rev()) %>%
        select(n.trees,everything())
      ,calib_tuningGrid=
        expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))
      ,epv=10
      ,sample.kind=sample.kind
    )
  
  saveRDS(disease_model2$pc_gbm,'data/pc_gbm2.rds')
}else{
  cat(readRDS('data/log.rds')[['pc_gbm2']])
  disease_model2$pc_gbm=readRDS('data/pc_gbm2.rds')
}
```

```{r Train PC-GBM using surrogate transcripts, include=FALSE}
if(run_heavy_computation){
  disease_model3$pc_gbm=
    train_set %>%
    .[,c('outcome'
         ,GSE108497_repro_tissue$result %>%
            rownames_to_column(var='feature') %>%
            filter(adj.P.Val<0.05) %>%
            pull(feature))] %>%
    disease_train(
      method='gbm'
      ,calib_method='gamLoess'
      ,tuning_trControl=training_parameters$tuning_trControl
      ,final_trControl=training_parameters$final_trControl
      ,tuningGrid=
        expand.grid(
          interaction.depth=1
          ,shrinkage=seq(0.0005,0.05,len=25)
          ,n.minobsinnode=20
        ) %>%
        mutate(n.trees=seq(0,2500,len=26) %>% .[-1] %>% rev()) %>%
        select(n.trees,everything())
      ,calib_tuningGrid=
        expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))
      ,epv=10
      ,sample.kind=sample.kind
    )
  
  saveRDS(disease_model3$pc_gbm,'data/pc_gbm3.rds')
}else{
  cat(readRDS('data/log.rds')[['pc_gbm3']])
  disease_model3$pc_gbm=readRDS('data/pc_gbm3.rds')
}
```

```{r Create empty lists for DIVNN prediction models, include=FALSE}
disease_divnn=list()
disease_divnn2=list()
disease_divnn3=list()
lambda=10^seq(-10,-1,len=10)
```

```{r Prepare DIVNN using all the features, include=FALSE}
disease_divnn$component=
  model_component(
    data=
      train_set %>%
      .[,c('outcome'
           ,GSE108497_all_tissue$result %>%
              rownames_to_column(var='feature') %>%
              filter(adj.P.Val<0.05) %>%
              pull(feature))]
    ,method='glmnet'
    ,calib_method='gamLoess'
    ,tuning_trControl=training_parameters$tuning_trControl
    ,final_trControl=training_parameters$final_trControl
    ,tuningGrid=expand.grid(alpha=seq(0,1,len=5),lambda=10^seq(-9,0,len=5))
    ,calib_tuningGrid=
      expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))
  )

if(run_heavy_computation){
  disease_divnn$pre_object=
    divnn_pre_object(
      data=
        disease_divnn$component$train %>%
        rownames_to_column(var='id') %>%
        select(-weight) %>%
        column_to_rownames(var='id')
      ,save=T
      ,path='data/disease_divnn'
    )
}else{
  cat(readRDS('data/log.rds')[['disease_divnn_pre_object']])
  disease_divnn$pre_object=
    list(
      input=readRDS('data/disease_divnn_input.rds')
      ,output=readRDS('data/disease_divnn_output.rds')
    )
}
```

```{r Show ontonet using all the features, eval=FALSE, include=FALSE}
disease_divnn$pre_object$output %>%
  viz.ontonet(feature=F) %>%
  plot(
    node.size=5
    ,edge.arrow.size=0.25
    ,main='Untrained Ontonet'
  )
```

```{r Hyperparameter tuning for a DI-VNN using all the features, include=FALSE}
if(run_heavy_computation){
  cat('Conduct hyperparameter tuning for DI-VNN model\n')
  cat('Started:',as.character(now()),'\n')
    
    surrogate_model=
      trainer_generator(
        disease_divnn$pre_object$output
        ,class_weight=
          disease_divnn$component$class %>%
          filter(partition=='train') %>%
          mutate(outcome=as.integer(outcome=='event')) %>%
          arrange(outcome) %>%
          lapply(X=1,Y=.,function(X,Y){
            setNames(Y$weight,Y$outcome)
          }) %>%
          .[[1]]
        ,epochs=5
        ,patience=round(5/2)
        ,batch_size=8
        ,warm_up=0.05
        ,lr=2^-6
        ,min_lr=2^-6/8
        ,tuning_mode=T
        ,verbose=1
      )
    
    disease_divnn$tuning_divnn=list()
    i=1
    for(j in seq(i,length(lambda))){
      suppressWarnings(set.seed(33,sample.kind=sample.kind))
      disease_divnn$tuning_divnn[[j]]=surrogate_model(lambda[j])
    }
  
  rm(i,j,surrogate_model)
  cat('End:',as.character(now()))
  saveRDS(disease_divnn$tuning_divnn,'data/tuning_divnn.rds')
}else{
  cat(readRDS('data/log.rds')[['tuning_divnn']])
  disease_divnn$tuning_divnn=readRDS('data/tuning_divnn.rds')
}
```

```{r DI-VNN tuning results using all the features, eval=FALSE, include=FALSE}
disease_divnn$tuning_divnn %>%
  sapply(function(x)min(x$Score)) %>%
  data.frame(mean_AUROC=.) %>%
  mutate(which_param=seq(nrow(.))) %>%
  mutate(best=which.max(mean_AUROC)) %>%
  qplot(which_param,mean_AUROC,color=which_param==best,data=.) +
  geom_smooth(method='loess',color='black',formula=y~x) +
  scale_x_continuous(breaks=seq(lambda))
```

```{r Conduct modeling for a DI-VNN using all the features, include=FALSE}
if(run_heavy_computation){
  cat('Conduct modeling for DI-VNN\n')
  cat('Started:',as.character(now()),'\n')
    
    surrogate_model=
      trainer_generator(
        disease_divnn$pre_object$output
        ,path='data/ontonet'
        ,class_weight=
          disease_divnn$component$class %>%
          filter(partition=='train') %>%
          mutate(outcome=as.integer(outcome=='event')) %>%
          arrange(outcome) %>%
          lapply(X=1,Y=.,function(X,Y){
            setNames(Y$weight,Y$outcome)
          }) %>%
          .[[1]]
        ,epochs=500
        ,patience=100
        ,batch_size=8
        ,warm_up=0.05
        ,lr=2^-6
        ,min_lr=2^-6/8
        ,tuning_mode=F
        ,checkpoint=T
        ,verbose=1
      )
    suppressWarnings(set.seed(33,sample.kind=sample.kind))
    disease_divnn$modeling_divnn=surrogate_model(lambda[8])
  
  rm(surrogate_model)
  cat('End:',as.character(now()))
  save_model_weights_hdf5(
    disease_divnn$modeling_divnn$ontonet
    ,'data/ontonet.h5'
  )
  saveRDS(disease_divnn$modeling_divnn,'data/modeling_divnn.rds')
}else{
  cat(readRDS('data/log.rds')[['modeling_divnn']])
  disease_divnn$modeling_divnn=readRDS('data/modeling_divnn.rds')
  refresh_session()
  disease_divnn$modeling_divnn$ontonet=
    disease_divnn$pre_object$output %>%
    generator.ontonet(l2_norm=lambda[8]) %>%
    load_model_weights_hdf5('data/ontonet.h5') %>%
    compile(
      optimizer=optimizer_sgd(
        lr=disease_divnn$modeling_divnn$history$metrics$lr %>%
          .[which.max(
              disease_divnn$modeling_divnn$history$metrics$val_root_roc
            )]
        ,momentum=0.9
      )
      ,loss='mean_squared_error'
      ,loss_weights=c(rep(
          0.3/(0.3*(length(.$outputs)-1)+1),length(.$outputs)-1)
          ,1/(0.3*(length(.$outputs)-1)+1)
        )
      ,metrics=c(
          tf$keras$metrics$AUC(name='roc')
          ,tf$keras$metrics$TruePositives(name='tp')
          ,tf$keras$metrics$FalseNegatives(name='fn')
          ,tf$keras$metrics$FalsePositives(name='fp')
          ,tf$keras$metrics$TrueNegatives(name='tn')
        )
    )
}
```

```{r Iteration plot 1, fig.height=7, fig.width=7, eval=FALSE, include=FALSE}
disease_divnn$modeling_divnn$history$metrics %>%
  .[c('root_loss','val_root_loss','root_roc','val_root_roc','lr')] %>%
  do.call(cbind,.) %>%
  as.data.frame() %>%
  mutate(iteration=seq(nrow(.))) %>%
  gather(metric,value,-iteration) %>%
  mutate(set=str_remove_all(metric,'_loss|_roc')) %>%
  mutate(
    metric=case_when(
      metric=='lr'~'1. LR'
      ,str_detect(metric,'loss')~'2. Loss'
      ,str_detect(metric,'roc')~'3. AUROC'
      ,TRUE~''
    )
  ) %>%
  qplot(iteration,value,color=set,data=.,geom='line') +
  geom_smooth(method='loess',formula=y~x,se=F) +
  facet_wrap(~metric,scales='free_y',ncol=1) +
  theme_minimal()
```

```{r Prepare pre-objects of the datasets using all the features, include=FALSE}
if(run_heavy_computation){
  disease_divnn$other_outputs=
    list(
      calib=
        disease_divnn$component$calib %>%
        rownames_to_column(var='id') %>%
        select(-weight) %>%
        column_to_rownames(var='id')
      ,int_val=
        disease_divnn$component$data %>%
        .[,colnames(disease_divnn$component$train)] %>%
        rownames_to_column(var='id') %>%
        select(-weight) %>%
        column_to_rownames(var='id')
      ,test_set1=test_set1
      ,test_set2=test_set2
      ,test_set3=test_set3
    ) %>%
    lapply(X=names(.),Y=.,function(X,Y){
      cat('Create tidy set for',X,'\n')
      Z=Y[[X]] %>%
        test_transformer(
          SGD1bit_fit=disease_divnn$pre_object$input$fit
          ,similarity=disease_divnn$pre_object$input$similarity
          ,mapping=disease_divnn$pre_object$input$mapping
          ,ontology=disease_divnn$pre_object$input$ontology
          ,ranked=T
          ,dims=7
          ,decreasing=F
          ,seed_num=33
        )
      cat('\n')
      Z
    }) %>%
    setNames(c('calib','int_val','test_set1','test_set2','test_set3'))
  
  saveRDS(disease_divnn$other_outputs,'data/disease_divnn_other_outputs.rds')
}else{
  cat(readRDS('data/log.rds')[['disease_divnn_other_outputs']])
  disease_divnn$other_outputs=
    readRDS('data/disease_divnn_other_outputs.rds')
}
```

```{r Calibrate and evaluate DI-VNN using all the features, include=FALSE}
if(run_heavy_computation){
  disease_model$divnn=
    divnn_calibrator_evaluator(
      data=disease_divnn$other_outputs
      ,modeling=disease_divnn$modeling_divnn
      ,component=disease_divnn$component
      ,batch_size=8
      ,title='Calibrate and evaluate DI-VNN'
      ,dir='data'
      ,file='divnn'
    )
}else{
  cat(readRDS('data/log.rds')[['divnn']])
  disease_model$divnn=
    list(
      model=readRDS('data/divnn.rds')
      ,calib_model=readRDS('data/calib_divnn.rds')
      ,eval_model=readRDS('data/eval_divnn.rds')
    )
}
```

```{r Visualization tables using all the features, include=FALSE}
if(run_heavy_computation){
  disease_divnn$visualization=list()
  
  cat('Get ontonet visualisation table.\n')
  disease_divnn$visualization$ontonet=
    disease_divnn$pre_object$output %>%
    viz.ontonet(
      feature=F
      ,eval.results=disease_divnn$modeling_divnn$evaluation
      ,eval.metric='roc'
      ,eval.pal=c('#E64B35FF','#00A087FF')
    )
  
  cat('Get ontoarray visualisation table.\n')
  disease_divnn$visualization$ontoarray=
    disease_divnn$pre_object$output %>%
    viz.ontoarray(disease_divnn$modeling_divnn$ontonet,batch_size=8,verbose=T)
  
  saveRDS(disease_divnn$visualization,'data/visualization.rds')
}else{
  cat(readRDS('data/log.rds')[['visualization']])
  disease_divnn$visualization=readRDS('data/visualization.rds')
}
```

```{r Prepare DIVNN using blood transcripts, include=FALSE}
disease_divnn2$component=
  model_component(
    data=
      train_set %>%
      .[,c('outcome'
           ,GSE108497_sel_mat_blood$result %>%
              rownames_to_column(var='feature') %>%
              filter(adj.P.Val<0.05) %>%
              pull(feature))]
    ,method='glmnet'
    ,calib_method='gamLoess'
    ,tuning_trControl=training_parameters$tuning_trControl
    ,final_trControl=training_parameters$final_trControl
    ,tuningGrid=expand.grid(alpha=seq(0,1,len=5),lambda=10^seq(-9,0,len=5))
    ,calib_tuningGrid=
      expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))
  )

if(run_heavy_computation){
  disease_divnn2$pre_object=
    divnn_pre_object(
      data=
        disease_divnn2$component$train %>%
        rownames_to_column(var='id') %>%
        select(-weight) %>%
        column_to_rownames(var='id')
      ,save=T
      ,path='data/disease_divnn2'
    )
}else{
  cat(readRDS('data/log.rds')[['disease_divnn2_pre_object']])
  disease_divnn2$pre_object=
    list(
      input=readRDS('data/disease_divnn2_input.rds')
      ,output=readRDS('data/disease_divnn2_output.rds')
    )
}
```

```{r Show ontonet using blood transcripts, eval=FALSE, include=FALSE}
disease_divnn2$pre_object$output %>%
  viz.ontonet(feature=F) %>%
  plot(
    node.size=5
    ,edge.arrow.size=0.25
    ,main='Untrained Ontonet'
  )
```

```{r Hyperparameter tuning for a DI-VNN using blood transcripts, include=FALSE}
if(run_heavy_computation){
  cat('Conduct hyperparameter tuning for DI-VNN 2 model\n')
  cat('Started:',as.character(now()),'\n')
    
    surrogate_model=
      trainer_generator(
        disease_divnn2$pre_object$output
        ,class_weight=
          disease_divnn2$component$class %>%
          filter(partition=='train') %>%
          mutate(outcome=as.integer(outcome=='event')) %>%
          arrange(outcome) %>%
          lapply(X=1,Y=.,function(X,Y){
            setNames(Y$weight,Y$outcome)
          }) %>%
          .[[1]]
        ,epochs=5
        ,patience=round(5/2)
        ,batch_size=8
        ,warm_up=0.05
        ,lr=2^-6
        ,min_lr=2^-6/8
        ,tuning_mode=T
        ,verbose=1
      )
    
    disease_divnn2$tuning_divnn=list()
    i=1
    for(j in seq(i,length(lambda))){
      suppressWarnings(set.seed(33,sample.kind=sample.kind))
      disease_divnn2$tuning_divnn[[j]]=surrogate_model(lambda[j])
    }
  
  rm(i,j,surrogate_model)
  cat('End:',as.character(now()))
  saveRDS(disease_divnn2$tuning_divnn,'data/tuning_divnn2.rds')
}else{
  cat(readRDS('data/log.rds')[['tuning_divnn2']])
  disease_divnn2$tuning_divnn=readRDS('data/tuning_divnn2.rds')
}
```

```{r DI-VNN tuning results using blood transcripts, eval=FALSE, include=FALSE}
disease_divnn2$tuning_divnn %>%
  sapply(function(x)min(x$Score)) %>%
  data.frame(mean_AUROC=.) %>%
  mutate(which_param=seq(nrow(.))) %>%
  mutate(best=which.max(mean_AUROC)) %>%
  qplot(which_param,mean_AUROC,color=which_param==best,data=.) +
  geom_smooth(method='loess',color='black',formula=y~x) +
  scale_x_continuous(breaks=seq(lambda))
```

```{r Conduct modeling for a DI-VNN using blood transcripts, include=FALSE}
if(run_heavy_computation){
  cat('Conduct modeling for DI-VNN 2\n')
  cat('Started:',as.character(now()),'\n')
    
    surrogate_model=
      trainer_generator(
        disease_divnn2$pre_object$output
        ,path='data/ontonet2'
        ,class_weight=
          disease_divnn2$component$class %>%
          filter(partition=='train') %>%
          mutate(outcome=as.integer(outcome=='event')) %>%
          arrange(outcome) %>%
          lapply(X=1,Y=.,function(X,Y){
            setNames(Y$weight,Y$outcome)
          }) %>%
          .[[1]]
        ,epochs=500
        ,patience=100
        ,batch_size=8
        ,warm_up=0.05
        ,lr=2^-6
        ,min_lr=2^-6/8
        ,tuning_mode=F
        ,checkpoint=T
        ,verbose=1
      )
    suppressWarnings(set.seed(33,sample.kind=sample.kind))
    disease_divnn2$modeling_divnn=surrogate_model(lambda[10])
  
  rm(surrogate_model)
  cat('End:',as.character(now()))
  save_model_weights_hdf5(
    disease_divnn2$modeling_divnn$ontonet
    ,'data/ontonet2.h5'
  )
  saveRDS(disease_divnn2$modeling_divnn,'data/modeling_divnn2.rds')
}else{
  cat(readRDS('data/log.rds')[['modeling_divnn2']])
  disease_divnn2$modeling_divnn=readRDS('data/modeling_divnn2.rds')
  refresh_session()
  disease_divnn2$modeling_divnn$ontonet=
    disease_divnn2$pre_object$output %>%
    generator.ontonet(l2_norm=lambda[10]) %>%
    load_model_weights_hdf5('data/ontonet2.h5') %>%
    compile(
      optimizer=optimizer_sgd(
        lr=disease_divnn2$modeling_divnn$history$metrics$lr %>%
          .[which.max(
              disease_divnn2$modeling_divnn$history$metrics$val_root_roc
            )]
        ,momentum=0.9
      )
      ,loss='mean_squared_error'
      ,loss_weights=c(rep(
          0.3/(0.3*(length(.$outputs)-1)+1),length(.$outputs)-1)
          ,1/(0.3*(length(.$outputs)-1)+1)
        )
      ,metrics=c(
          tf$keras$metrics$AUC(name='roc')
          ,tf$keras$metrics$TruePositives(name='tp')
          ,tf$keras$metrics$FalseNegatives(name='fn')
          ,tf$keras$metrics$FalsePositives(name='fp')
          ,tf$keras$metrics$TrueNegatives(name='tn')
        )
    )
}
```

```{r Iteration plot 2, fig.height=7, fig.width=7, eval=FALSE, include=FALSE}
disease_divnn2$modeling_divnn$history$metrics %>%
  .[c('root_loss','val_root_loss','root_roc','val_root_roc','lr')] %>%
  do.call(cbind,.) %>%
  as.data.frame() %>%
  mutate(iteration=seq(nrow(.))) %>%
  gather(metric,value,-iteration) %>%
  mutate(set=str_remove_all(metric,'_loss|_roc')) %>%
  mutate(
    metric=case_when(
      metric=='lr'~'1. LR'
      ,str_detect(metric,'loss')~'2. Loss'
      ,str_detect(metric,'roc')~'3. AUROC'
      ,TRUE~''
    )
  ) %>%
  qplot(iteration,value,color=set,data=.,geom='line') +
  geom_smooth(method='loess',formula=y~x,se=F) +
  facet_wrap(~metric,scales='free_y',ncol=1) +
  theme_minimal()
```

```{r Prepare pre-objects of the datasets using blood transcripts,include=FALSE}
if(run_heavy_computation){
  disease_divnn2$other_outputs=
    list(
      calib=
        disease_divnn2$component$calib %>%
        rownames_to_column(var='id') %>%
        select(-weight) %>%
        column_to_rownames(var='id')
      ,int_val=
        disease_divnn2$component$data %>%
        .[,colnames(disease_divnn2$component$train)] %>%
        rownames_to_column(var='id') %>%
        select(-weight) %>%
        column_to_rownames(var='id')
      ,test_set1=test_set1
      ,test_set2=test_set2
      ,test_set3=test_set3
    ) %>%
    lapply(X=names(.),Y=.,function(X,Y){
      cat('Create tidy set for',X,'\n')
      Z=Y[[X]] %>%
        test_transformer(
          SGD1bit_fit=disease_divnn2$pre_object$input$fit
          ,similarity=disease_divnn2$pre_object$input$similarity
          ,mapping=disease_divnn2$pre_object$input$mapping
          ,ontology=disease_divnn2$pre_object$input$ontology
          ,ranked=T
          ,dims=7
          ,decreasing=F
          ,seed_num=33
        )
      cat('\n')
      Z
    }) %>%
    setNames(c('calib','int_val','test_set1','test_set2','test_set3'))
  
  saveRDS(disease_divnn2$other_outputs,'data/disease_divnn2_other_outputs.rds')
}else{
  cat(readRDS('data/log.rds')[['disease_divnn2_other_outputs']])
  disease_divnn2$other_outputs=
    readRDS('data/disease_divnn2_other_outputs.rds')
}
```

```{r Calibrate and evaluate DI-VNN using blood transcripts, include=FALSE}
if(run_heavy_computation){
  disease_model2$divnn=
    divnn_calibrator_evaluator(
      data=disease_divnn2$other_outputs
      ,modeling=disease_divnn2$modeling_divnn
      ,component=disease_divnn2$component
      ,batch_size=8
      ,title='Calibrate and evaluate DI-VNN 2'
      ,dir='data'
      ,file='divnn2'
    )
}else{
  cat(readRDS('data/log.rds')[['divnn2']])
  disease_model2$divnn=
    list(
      model=readRDS('data/divnn2.rds')
      ,calib_model=readRDS('data/calib_divnn2.rds')
      ,eval_model=readRDS('data/eval_divnn2.rds')
    )
}
```

```{r Visualization tables using blood transcripts, include=FALSE}
if(run_heavy_computation){
  disease_divnn2$visualization=list()
  
  cat('Get ontonet 2 visualisation table.\n')
  disease_divnn2$visualization$ontonet=
    disease_divnn2$pre_object$output %>%
    viz.ontonet(
      feature=F
      ,eval.results=disease_divnn2$modeling_divnn$evaluation
      ,eval.metric='roc'
      ,eval.pal=c('#E64B35FF','#00A087FF')
    )
  
  cat('Get ontoarray 2 visualisation table.\n')
  disease_divnn2$visualization$ontoarray=
    disease_divnn2$pre_object$output %>%
    viz.ontoarray(disease_divnn2$modeling_divnn$ontonet,batch_size=8,verbose=T)
  
  saveRDS(disease_divnn2$visualization,'data/visualization2.rds')
}else{
  cat(readRDS('data/log.rds')[['visualization2']])
  disease_divnn2$visualization=readRDS('data/visualization2.rds')
}
```

```{r Prepare DIVNN using surrogate transcripts, include=FALSE}
disease_divnn3$component=
  model_component(
    data=
      train_set %>%
      .[,c('outcome'
           ,GSE108497_repro_tissue$result %>%
              rownames_to_column(var='feature') %>%
              filter(adj.P.Val<0.05) %>%
              pull(feature))]
    ,method='glmnet'
    ,calib_method='gamLoess'
    ,tuning_trControl=training_parameters$tuning_trControl
    ,final_trControl=training_parameters$final_trControl
    ,tuningGrid=expand.grid(alpha=seq(0,1,len=5),lambda=10^seq(-9,0,len=5))
    ,calib_tuningGrid=
      expand.grid(span=seq(0.15,0.65,len=10),degree=seq(0,1,len=2))
  )

if(run_heavy_computation){
  disease_divnn3$pre_object=
    divnn_pre_object(
      data=
        disease_divnn3$component$train %>%
        rownames_to_column(var='id') %>%
        select(-weight) %>%
        column_to_rownames(var='id')
      ,save=T
      ,path='data/disease_divnn3'
    )
}else{
  cat(readRDS('data/log.rds')[['disease_divnn3_pre_object']])
  disease_divnn3$pre_object=
    list(
      input=readRDS('data/disease_divnn3_input.rds')
      ,output=readRDS('data/disease_divnn3_output.rds')
    )
}
```

```{r Show ontonet using surrogate transcripts, eval=FALSE, include=FALSE}
disease_divnn3$pre_object$output %>%
  viz.ontonet(feature=F) %>%
  plot(
    node.size=5
    ,edge.arrow.size=0.25
    ,main='Untrained Ontonet'
  )
```

```{r Hyperparameter tuning for DIVNN using surrogate transcript, include=FALSE}
if(run_heavy_computation){
  cat('Conduct hyperparameter tuning for DI-VNN 3 model\n')
  cat('Started:',as.character(now()),'\n')
    
    surrogate_model=
      trainer_generator(
        disease_divnn3$pre_object$output
        ,class_weight=
          disease_divnn3$component$class %>%
          filter(partition=='train') %>%
          mutate(outcome=as.integer(outcome=='event')) %>%
          arrange(outcome) %>%
          lapply(X=1,Y=.,function(X,Y){
            setNames(Y$weight,Y$outcome)
          }) %>%
          .[[1]]
        ,epochs=5
        ,patience=round(5/2)
        ,batch_size=8
        ,warm_up=0.05
        ,lr=2^-6
        ,min_lr=2^-6/8
        ,tuning_mode=T
        ,verbose=1
      )
    
    disease_divnn3$tuning_divnn=list()
    i=1
    for(j in seq(i,length(lambda))){
      suppressWarnings(set.seed(33,sample.kind=sample.kind))
      disease_divnn3$tuning_divnn[[j]]=surrogate_model(lambda[j])
    }
  
  rm(i,j,surrogate_model)
  cat('End:',as.character(now()))
  saveRDS(disease_divnn3$tuning_divnn,'data/tuning_divnn3.rds')
}else{
  cat(readRDS('data/log.rds')[['tuning_divnn3']])
  disease_divnn3$tuning_divnn=readRDS('data/tuning_divnn3.rds')
}
```

```{r DIVNN tuning result using surrogate transcript, eval=FALSE, include=FALSE}
disease_divnn3$tuning_divnn %>%
  sapply(function(x)min(x$Score)) %>%
  data.frame(mean_AUROC=.) %>%
  mutate(which_param=seq(nrow(.))) %>%
  mutate(best=which.max(mean_AUROC)) %>%
  qplot(which_param,mean_AUROC,color=which_param==best,data=.) +
  geom_smooth(method='loess',color='black',formula=y~x) +
  scale_x_continuous(breaks=seq(lambda))
```

```{r Conduct modeling for a DI-VNN using surrogate transcripts, include=FALSE}
if(run_heavy_computation){
  cat('Conduct modeling for DI-VNN 3\n')
  cat('Started:',as.character(now()),'\n')
    
    surrogate_model=
      trainer_generator(
        disease_divnn3$pre_object$output
        ,path='data/ontonet3'
        ,class_weight=
          disease_divnn3$component$class %>%
          filter(partition=='train') %>%
          mutate(outcome=as.integer(outcome=='event')) %>%
          arrange(outcome) %>%
          lapply(X=1,Y=.,function(X,Y){
            setNames(Y$weight,Y$outcome)
          }) %>%
          .[[1]]
        ,epochs=500
        ,patience=100
        ,batch_size=8
        ,warm_up=0.05
        ,lr=2^-6
        ,min_lr=2^-6/8
        ,tuning_mode=F
        ,checkpoint=T
        ,verbose=1
      )
    suppressWarnings(set.seed(33,sample.kind=sample.kind))
    disease_divnn3$modeling_divnn=surrogate_model(lambda[8])
  
  rm(surrogate_model)
  cat('End:',as.character(now()))
  save_model_weights_hdf5(
    disease_divnn3$modeling_divnn$ontonet
    ,'data/ontonet3.h5'
  )
  saveRDS(disease_divnn3$modeling_divnn,'data/modeling_divnn3.rds')
}else{
  cat(readRDS('data/log.rds')[['modeling_divnn3']])
  disease_divnn3$modeling_divnn=readRDS('data/modeling_divnn3.rds')
  refresh_session()
  disease_divnn3$modeling_divnn$ontonet=
    disease_divnn3$pre_object$output %>%
    generator.ontonet(l2_norm=lambda[8]) %>%
    load_model_weights_hdf5('data/ontonet3.h5') %>%
    compile(
      optimizer=optimizer_sgd(
        lr=disease_divnn3$modeling_divnn$history$metrics$lr %>%
          .[which.max(
              disease_divnn3$modeling_divnn$history$metrics$val_root_roc
            )]
        ,momentum=0.9
      )
      ,loss='mean_squared_error'
      ,loss_weights=c(rep(
          0.3/(0.3*(length(.$outputs)-1)+1),length(.$outputs)-1)
          ,1/(0.3*(length(.$outputs)-1)+1)
        )
      ,metrics=c(
          tf$keras$metrics$AUC(name='roc')
          ,tf$keras$metrics$TruePositives(name='tp')
          ,tf$keras$metrics$FalseNegatives(name='fn')
          ,tf$keras$metrics$FalsePositives(name='fp')
          ,tf$keras$metrics$TrueNegatives(name='tn')
        )
    )
}
```

```{r Iteration plot 3, fig.height=7, fig.width=7, eval=FALSE, include=FALSE}
disease_divnn3$modeling_divnn$history$metrics %>%
  .[c('root_loss','val_root_loss','root_roc','val_root_roc','lr')] %>%
  do.call(cbind,.) %>%
  as.data.frame() %>%
  mutate(iteration=seq(nrow(.))) %>%
  gather(metric,value,-iteration) %>%
  mutate(set=str_remove_all(metric,'_loss|_roc')) %>%
  mutate(
    metric=case_when(
      metric=='lr'~'1. LR'
      ,str_detect(metric,'loss')~'2. Loss'
      ,str_detect(metric,'roc')~'3. AUROC'
      ,TRUE~''
    )
  ) %>%
  qplot(iteration,value,color=set,data=.,geom='line') +
  geom_smooth(method='loess',formula=y~x,se=F) +
  facet_wrap(~metric,scales='free_y',ncol=1) +
  theme_minimal()
```

```{r Prepare pre-objects using surrogate transcripts, include=FALSE}
if(run_heavy_computation){
  disease_divnn3$other_outputs=
    list(
      calib=
        disease_divnn3$component$calib %>%
        rownames_to_column(var='id') %>%
        select(-weight) %>%
        column_to_rownames(var='id')
      ,int_val=
        disease_divnn3$component$data %>%
        .[,colnames(disease_divnn3$component$train)] %>%
        rownames_to_column(var='id') %>%
        select(-weight) %>%
        column_to_rownames(var='id')
      ,test_set1=test_set1
      ,test_set2=test_set2
      ,test_set3=test_set3
    ) %>%
    lapply(X=names(.),Y=.,function(X,Y){
      cat('Create tidy set for',X,'\n')
      Z=Y[[X]] %>%
        test_transformer(
          SGD1bit_fit=disease_divnn3$pre_object$input$fit
          ,similarity=disease_divnn3$pre_object$input$similarity
          ,mapping=disease_divnn3$pre_object$input$mapping
          ,ontology=disease_divnn3$pre_object$input$ontology
          ,ranked=T
          ,dims=7
          ,decreasing=F
          ,seed_num=33
        )
      cat('\n')
      Z
    }) %>%
    setNames(c('calib','int_val','test_set1','test_set2','test_set3'))
  
  saveRDS(disease_divnn3$other_outputs,'data/disease_divnn3_other_outputs.rds')
}else{
  cat(readRDS('data/log.rds')[['disease_divnn3_other_outputs']])
  disease_divnn3$other_outputs=
    readRDS('data/disease_divnn3_other_outputs.rds')
}
```

```{r Calibrate and evaluate DI-VNN using surrogate transcripts, include=FALSE}
if(run_heavy_computation){
  disease_model3$divnn=
    divnn_calibrator_evaluator(
      data=disease_divnn3$other_outputs
      ,modeling=disease_divnn3$modeling_divnn
      ,component=disease_divnn3$component
      ,batch_size=8
      ,title='Calibrate and evaluate DI-VNN 3'
      ,dir='data'
      ,file='divnn3'
    )
}else{
  cat(readRDS('data/log.rds')[['divnn3']])
  disease_model3$divnn=
    list(
      model=readRDS('data/divnn3.rds')
      ,calib_model=readRDS('data/calib_divnn3.rds')
      ,eval_model=readRDS('data/eval_divnn3.rds')
    )
}
```

```{r Visualization tables using surrogate transcripts, include=FALSE}
if(run_heavy_computation){
  disease_divnn3$visualization=list()
  
  cat('Get ontonet 3 visualisation table.\n')
  disease_divnn3$visualization$ontonet=
    disease_divnn3$pre_object$output %>%
    viz.ontonet(
      feature=F
      ,eval.results=disease_divnn3$modeling_divnn$evaluation
      ,eval.metric='roc'
      ,eval.pal=c('#E64B35FF','#00A087FF')
    )
  
  cat('Get ontoarray 3 visualisation table.\n')
  disease_divnn3$visualization$ontoarray=
    disease_divnn3$pre_object$output %>%
    viz.ontoarray(disease_divnn3$modeling_divnn$ontonet,batch_size=8,verbose=T)
  
  saveRDS(disease_divnn3$visualization,'data/visualization3.rds')
}else{
  cat(readRDS('data/log.rds')[['visualization3']])
  disease_divnn3$visualization=readRDS('data/visualization3.rds')
}
```

```{r}
timestamper[11]=as.character(now())
```

```{r Create a function to compare calibration of the models, include=FALSE}
calib_comparison=function(calib){
  list(
      all=disease_model
      ,blood=disease_model2
      ,repro=disease_model3
    ) %>%
    lapply(X=names(.),Y=.,Z=calib,function(X,Y,Z){
      Y[[X]] %>%
        lapply(X=names(.),Y=.,Z=X,K=Z,function(X,Y,Z,K){
          if(X=='divnn'){
            Y[[X]][[K]]$pred=
              Y[[X]][[K]]$pred %>%
              mutate(Resample='Resample01')
          }
          Y[[X]][[K]]$pred %>%
            mutate(pred=round(event*10)/10) %>%
            group_by(Resample,pred) %>%
            summarize(
              event=sum(obs=='event')
              ,nonevent=sum(obs=='nonevent')
              ,total=n()
              ,obs=mean(obs=='event',na.rm=T)
            ) %>%
            ungroup() %>%
            group_by(pred) %>%
            summarize(
              avg=mean(obs,na.rm=T)
              ,lb=mean(obs,na.rm=T)-qnorm(0.975)*sd(obs,na.rm=T)/sqrt(n())
              ,ub=mean(obs,na.rm=T)+qnorm(0.975)*sd(obs,na.rm=T)/sqrt(n())
              ,event=sum(event)
              ,nonevent=sum(nonevent)
              ,total=sum(total)
            ) %>%
            ungroup() %>%
            cbind(
              lm(avg~pred,data=.) %>%
                broom::tidy() %>%
                mutate(term=term %>% str_remove_all('[:punct:]') %>% str_to_lower()) %>%
                select(term,estimate,std.error) %>%
                mutate(term=ifelse(term=='intercept','i','s')) %>%
                rename(avg=estimate,se=std.error) %>%
                pivot_wider(names_from=term,values_from=c(avg,se)) %>%
                mutate_all(round,2)
            ) %>%
            mutate(fit=avg_i+avg_s*pred) %>%
            mutate(
              event=event/max(event)
              ,nonevent=nonevent/max(nonevent)
              ,event_from=-0.2-0.25
              ,event_to=-0.2-0.25+0.25*event
              ,nonevent_from=-0.2-0.25-0.25*nonevent
              ,nonevent_to=-0.2-0.25
            ) %>%
            mutate(model=X)
        }) %>%
        do.call(rbind,.) %>%
        mutate(tissue=X)
    }) %>%
    do.call(rbind,.) %>%
      mutate(
        model=
          model %>%
          factor(c('pc_elnet','pc_rf','pc_gbm','divnn'))
        ,tissue=
          tissue %>%
          factor(c('all','blood','repro'))
      )
}
```

```{r Calibration of the no-calib models, echo=FALSE, fig.height=7, fig.width=7}
calib_comparison('model') %>%
  filter(tissue!='all') %>%
  ggplot(aes(pred)) +
  geom_abline(intercept=0,slope=1,lty=2) +
  geom_smooth(aes(y=avg),method='lm',formula=y~x,color='darkgrey',size=2) +
  geom_line(aes(y=fit)) +
  geom_linerange(aes(ymin=lb,ymax=ub)) +
  geom_linerange(
    aes(ymin=event_from,ymax=event_to,color='Event')
    ,size=5
  ) +
  geom_linerange(
    aes(ymin=nonevent_from,ymax=nonevent_to,color='Nonevent')
    ,size=5
  ) +
  geom_point(aes(y=avg)) +
  geom_text(
    aes(label=ifelse(pred==0.1,paste0('intercept ',avg_i,' ± ',se_i),NA))
    ,x=0,y=1.2,vjust=1,hjust=0,size=3,na.rm=T
  ) +
  geom_text(
    aes(label=ifelse(pred==0.1,paste0('slope ',avg_s,'± ',se_s),NA))
    ,x=0,y=1.1,vjust=1,hjust=0,size=3,na.rm=T
  ) +
  facet_grid(tissue~model) +
  coord_equal() +
  scale_x_continuous(limits=c(0,1),breaks=seq(0,1,0.1)) +
  scale_y_continuous(limits=c(-0.7,1.2),breaks=seq(0,1,0.1)) +
  scale_color_manual('',values=c('red','blue')) +
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
```

```{r Calibration of the calib models, echo=FALSE, fig.height=7, fig.width=7}
calib_comparison('calib_model') %>%
  filter(tissue!='all') %>%
  ggplot(aes(pred)) +
  geom_abline(intercept=0,slope=1,lty=2) +
  geom_smooth(aes(y=avg),method='lm',formula=y~x,color='darkgrey',size=2) +
  geom_line(aes(y=fit)) +
  geom_linerange(aes(ymin=lb,ymax=ub)) +
  geom_linerange(
    aes(ymin=event_from,ymax=event_to,color='Event')
    ,size=5
  ) +
  geom_linerange(
    aes(ymin=nonevent_from,ymax=nonevent_to,color='Nonevent')
    ,size=5
  ) +
  geom_point(aes(y=avg)) +
  geom_text(
    aes(label=ifelse(pred==0,paste0('intercept ',avg_i,' ± ',se_i),NA))
    ,x=0,y=1.2,vjust=1,hjust=0,size=3,na.rm=T
  ) +
  geom_text(
    aes(label=ifelse(pred==0,paste0('slope ',avg_s,'± ',se_s),NA))
    ,x=0,y=1.1,vjust=1,hjust=0,size=3,na.rm=T
  ) +
  facet_grid(tissue~model) +
  coord_equal() +
  scale_x_continuous(limits=c(0,1),breaks=seq(0,1,0.1)) +
  scale_y_continuous(limits=c(-0.7,1.2),breaks=seq(0,1,0.1)) +
  scale_color_manual('',values=c('red','blue')) +
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
```

```{r Create a function to get model weights, include=FALSE}
opt_feat_disease_fn=function(X,Y,Z){
  if(X=='divnn'){
    # Get the classification DI-VNN weights of the representation layers.
    Z$visualization$ontoarray %>%
    lapply(function(x){
      x$ontotype %>%
        lapply(X=seq(nrow(.)),Y=.,Z=x$output,function(X,Y,Z){
          K=Z[Y$x[X],Y$y[X],Y$z[X]]
          if(K==0){
            NULL
          }else{
            K
          }
        }) %>%
        setNames(x$ontotype$feature)
    }) %>%
    setNames(names(Z$visualization$ontoarray)) %>%
    unlist() %>%
    data.frame(output=.) %>%
    rownames_to_column(var='ontology') %>%
    rename(classification_output=output) %>%
    separate(ontology,c('ontology','predictor'),sep='\\.') %>%
    
    # Get the predictor position.
    left_join(
      Z$pre_object$output %>%
        fData() %>%
        rownames_to_column(var='xyz') %>%
        rename(predictor=feature) %>%
        select(xyz,predictor) %>%
        filter(!is.na(predictor)) %>%
        mutate(xyz=str_remove_all(xyz,'x')) %>%
        separate(xyz,c('dimension_1','yz'),sep='y') %>%
        separate(yz,c('dimension_2','dimension_3'),sep='z')
      ,by='predictor'
    ) %>%
    
    # Wrap up.
    arrange(ontology,dimension_3,dimension_2,dimension_1,predictor) %>%
    select(
      ontology
      ,dimension_1
      ,dimension_2
      ,dimension_3
      ,predictor
      ,everything()
    )
  }else{
    if(Y[[X]]$method=='glmnet'){
      # Get weights from the elastic net regression model.
      Z=coef(Y[[X]]$model$finalModel,Y[[X]]$model$bestTune$lambda) %>%
        as.matrix() %>%
        as.data.frame()
    }else{
      # Get weights from the tree-based models.
      Z=varImp(Y[[X]]$model) %>%
        .[[1]]
    }
    
    Z %>%
      rownames_to_column(var='term') %>%
      
      # Clean up.
      mutate(term=str_remove_all(term,'\\`|\\(|\\)')) %>%
      setNames(c('term','estimate')) %>%
      filter(estimate!=0) %>%
      arrange(desc(abs(estimate))) %>%
      filter(!term %in% c('Intercept')) %>%
      right_join(
        Y[[X]]$transformator$avg_rotm %>%
          .[,Y[[X]]$pc %>% colnames() %>% .[!.%in%c('outcome','weight')]] %>%
          as.data.frame() %>%
          rownames_to_column(var='tissue_feature') %>%
          separate(tissue_feature,c('tissue_type','feature'),sep=':') %>%
          gather(term,value,-feature,-tissue_type)
        ,by='term'
      ) %>%
      arrange(
        factor(tissue_type,c('maternal_blood',names(dea)))
        ,desc(estimate)
      )
  }
}
```

```{r Get weights of the models using all the features, include=FALSE}
opt_feat_disease=
  disease_model %>%
  lapply(X=names(.),Y=.,Z=disease_divnn,opt_feat_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Get weights of the models using blood transcripts, include=FALSE}
opt_feat_disease2=
  disease_model2 %>%
  lapply(X=names(.),Y=.,Z=disease_divnn2,opt_feat_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Get weights of the models using surrogate features, include=FALSE}
opt_feat_disease3=
  disease_model3 %>%
  lapply(X=names(.),Y=.,Z=disease_divnn3,opt_feat_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Create a function to evaluate the models, include=FALSE}
opt_pred_disease_fn=function(X,Y,Z,Z2){
  if(X=='divnn'){
    K=Y[[X]]$eval_model[[paste0('calib_',Z2)]]
    Y[[X]]$eval_model=Y[[X]]$eval_model$calib_int_val
  }else{
    K=Z %>%
      lapply(X=1,Y=.,Z=Y[[X]],function(X,Y,Z){
        K=ExpressionSet(
            assayData=
              Y %>%
              select(-outcome) %>%
              t()
            ,phenoData=
              Y %>%
              select(outcome) %>%
              AnnotatedDataFrame()
          ) %>%
          .[rownames(Z$transformator$scaler),] %>%
          transformation(transformator_object=Z$transformator)
        
        K %>%
          .[Z$pc %>%
              colnames() %>%
              .[!.%in%c('outcome','weight')]
            ,] %>%
          exprs() %>%
          t() %>%
          as.data.frame() %>%
          rownames_to_column(var='id') %>%
          left_join(
            K %>%
              phenoData() %>%
              pData() %>%
              select(outcome) %>%
              rownames_to_column(var='id')
            ,by='id'
          ) %>%
          select(outcome,everything()) %>%
          column_to_rownames(var='id')
      }) %>%
      .[[1]] %>%
      cbind(predict(Y[[X]]$model,newdata=.,type='prob')) %>%
      rownames_to_column(var='id') %>%
      select(id,outcome,event) %>%
      rename(prob=event) %>%
      column_to_rownames(var='id') %>%
      cbind(
        suppressWarnings(
            predict(Y[[X]]$calib_model,newdata=.,type='prob')
          ) %>%
          rename_all(function(x)paste0(x,'2'))
      ) %>%
      select(-prob) %>%
      rename_all(function(x)str_remove_all(x,'2')) %>%
      mutate(obs=outcome) %>%
      select(-outcome) %>%
      select(nonevent,event,obs) %>%
      evalm(silent=T,showplots=F)
  }
  
  L=list(
    Y[[X]]$eval_model$optres$Group1 %>%
      setNames(paste0(colnames(.),'_train'))

    ,K %>%
      .$optres %>%
      .$Group1 %>%
      setNames(paste0(colnames(.),'_test'))
    ) %>%
    do.call(cbind,.)

  list(all_conclusions=L,test_eval=K)
}
```

```{r Evaluate using all the features in the replication 1, include=FALSE}
opt_pred_disease1=
  disease_model %>%
  lapply(X=names(.),Y=.,Z=test_set1,Z2='test_set1',opt_pred_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Evaluate using blood transcripts in the replication 1, include=FALSE}
opt_pred2_disease1=
  disease_model2 %>%
  lapply(X=names(.),Y=.,Z=test_set1,Z2='test_set1',opt_pred_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Evaluate using surrogate transcripts in the replication 1, include=FALSE}
opt_pred3_disease1=
  disease_model3 %>%
  lapply(X=names(.),Y=.,Z=test_set1,Z2='test_set1',opt_pred_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Evaluate using all the features in the replication 2, include=FALSE}
opt_pred_disease2=
  disease_model %>%
  lapply(X=names(.),Y=.,Z=test_set2,Z2='test_set2',opt_pred_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Evaluate using blood transcripts in the replication 2, include=FALSE}
opt_pred2_disease2=
  disease_model2 %>%
  lapply(X=names(.),Y=.,Z=test_set2,Z2='test_set2',opt_pred_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Evaluate using surrogate transcripts in the replication 2, include=FALSE}
opt_pred3_disease2=
  disease_model3 %>%
  lapply(X=names(.),Y=.,Z=test_set2,Z2='test_set2',opt_pred_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Evaluate using all the features in the replication 3, include=FALSE}
opt_pred_disease3=
  disease_model %>%
  lapply(X=names(.),Y=.,Z=test_set3,Z2='test_set3',opt_pred_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Evaluate using blood transcripts in the replication 3, include=FALSE}
opt_pred2_disease3=
  disease_model2 %>%
  lapply(X=names(.),Y=.,Z=test_set3,Z2='test_set3',opt_pred_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Evaluate using surrogate transcripts in the replication 3, include=FALSE}
opt_pred3_disease3=
  disease_model3 %>%
  lapply(X=names(.),Y=.,Z=test_set3,Z2='test_set3',opt_pred_disease_fn) %>%
  setNames(names(disease_model))
```

```{r Create a function to compare results of the evaluation, include=FALSE}
result_comparison=function(metric){
  list(
      all_GSE85307=opt_pred_disease1
      ,all_GSE86200=opt_pred_disease2
      ,all_GSE149437=opt_pred_disease3
      
      ,blood_GSE85307=opt_pred2_disease1
      ,blood_GSE86200=opt_pred2_disease2
      ,blood_GSE149437=opt_pred2_disease3
      
      ,repro_GSE85307=opt_pred3_disease1
      ,repro_GSE86200=opt_pred3_disease2
      ,repro_GSE149437=opt_pred3_disease3
    ) %>%
    lapply(X=names(.),Y=.,Z=metric,function(X,Y,Z){
      Y[[X]] %>%
        lapply(X=names(.),Y=.,Z=X,K=Z,function(X,Y,Z,K){
          Y[[X]]$all_conclusions[K,] %>%
            separate(CI_train,c('LB_train','UB_train'),sep='-') %>%
            separate(CI_test,c('LB_test','UB_test'),sep='-') %>%
            mutate_all(as.numeric) %>%
            gather(AUC,value) %>%
            separate(AUC,c('AUC','set'),sep='_') %>%
            spread(AUC,value) %>%
            mutate(model=X) %>%
            mutate(
              set=
                ifelse(
                  set=='train'
                  ,'GSE108497'
                  ,str_split_fixed(Z,'_',2)[,2]
                )
              ,tissue=str_split_fixed(Z,'_',2)[,1]
            )
        }) %>%
        do.call(rbind,.)
    }) %>%
    do.call(rbind,.) %>%
    filter(!duplicated(.)) %>%
    mutate(
      model=
        model %>%
        factor(c('pc_elnet','pc_rf','pc_gbm','divnn'))
      ,tissue=
        tissue %>%
        factor(c('all','blood','repro'))
      ,set=
        set %>%
        factor(c('GSE108497','GSE85307','GSE86200','GSE149437'))
    )
}
```

```{r Compare the AUC-ROCs, echo=FALSE, fig.height=5.34309, fig.width=7.48031}
auc_comparison=result_comparison('AUC-ROC')

auc_comparison %>%
  filter(tissue!='all') %>%
  mutate(
    model=
      case_when(
        model=='pc_elnet'~'PC-ENR'
        ,model=='pc_rf'~'PC-RF'
        ,model=='pc_gbm'~'PC-GBM'
        ,model=='divnn'~'DI-VNN'
        ,TRUE~''
      ) %>%
      factor(c('PC-ENR','PC-RF','PC-GBM','DI-VNN'))
    ,tissue=
      case_when(
        tissue=='all'~'Both'
        ,tissue=='blood'~'Blood\ntranscriptome'
        ,tissue=='repro'~'Blood-derived\nsurrogate\ntranscriptome'
        ,TRUE~''
      ) %>%
      factor(c(
        'Blood-derived\nsurrogate\ntranscriptome'
        ,'Blood\ntranscriptome'
        ,'Both'
      ))
  ) %>%
  group_by(tissue,set) %>%
  mutate(best=ifelse(Score>=mean(Score) & LB>=0.5,'Yes','No')) %>%
  ungroup() %>%
  lapply(X=1,Y=.,function(X,Y){
    Y=Y %>%
      mutate(
        set=
          case_when(
            set=='GSE108497'~'Development\nGSE108497\n(no intervention)'
            ,set=='GSE85307'~'Replication\nGSE85307\n(vitamin D +/-)'
            ,set=='GSE86200'~'Replication\nGSE86200\n(vitamin D +/-)'
            ,set=='GSE149437'~'Replication\nGSE149437\n(no intervention)'
            ,TRUE~''
          ) %>%
          factor(c(
            'Development\nGSE108497\n(no intervention)'
            ,'Replication\nGSE85307\n(vitamin D +/-)'
            ,'Replication\nGSE86200\n(vitamin D +/-)'
            ,'Replication\nGSE149437\n(no intervention)'
          ))
      )
    Z=filter(
        Y,model=='PC-GBM'
        & tissue=='Blood-derived\nsurrogate\ntranscriptome'
      ) %>%
      arrange(set)
    print(Z)
    Y
  }) %>%
  .[[1]] %>%
  ggplot(aes(x=model,color=best)) +
  geom_hline(yintercept=0.5,lty=2) +
  geom_hline(
    data=
      auc_comparison %>%
      filter(tissue!='all') %>%
      # filter(model!='divnn') %>%
      mutate(
        model=
          case_when(
            model=='pc_elnet'~'PC-ENR'
            ,model=='pc_rf'~'PC-RF'
            ,model=='pc_gbm'~'PC-GBM'
            ,model=='divnn'~'DI-VNN'
            ,TRUE~''
          ) %>%
          factor(c('PC-ENR','PC-RF','PC-GBM','DI-VNN'))
        ,tissue=
          case_when(
            tissue=='all'~'Both'
            ,tissue=='blood'~'Blood\ntranscriptome'
            ,tissue=='repro'~'Blood-derived\nsurrogate\ntranscriptome'
            ,TRUE~''
          ) %>%
          factor(c(
            'Blood-derived\nsurrogate\ntranscriptome'
            ,'Blood\ntranscriptome'
            ,'Both'
          ))
        ,set=
          case_when(
            set=='GSE108497'~'Development\nGSE108497\n(no intervention)'
            ,set=='GSE85307'~'Replication\nGSE85307\n(vitamin D +/-)'
            ,set=='GSE86200'~'Replication\nGSE86200\n(vitamin D +/-)'
            ,set=='GSE149437'~'Replication\nGSE149437\n(no intervention)'
            ,TRUE~''
          ) %>%
          factor(c(
            'Development\nGSE108497\n(no intervention)'
            ,'Replication\nGSE85307\n(vitamin D +/-)'
            ,'Replication\nGSE86200\n(vitamin D +/-)'
            ,'Replication\nGSE149437\n(no intervention)'
          ))
      ) %>%
      group_by(tissue,set) %>%
      summarize(avg=mean(Score))
    ,aes(yintercept=avg)
    ,lty=2
  ) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0.25) +
  geom_point(aes(y=Score)) +
  facet_grid(tissue~set) +
  coord_flip() +
  scale_x_discrete('Model') +
  scale_y_continuous('AUROC (95% CI)') +
  scale_color_manual('Best',values=c('black','red')) +
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)) +
  theme(
    axis.title.x=element_text(size=unit(9,'pt'),family='sans')
    ,axis.text.x=
      element_text(size=unit(8,'pt'),family='sans',angle=90,hjust=1,vjust=0.5)
    ,axis.title.y=element_text(size=unit(9,'pt'),family='sans')
    ,axis.text.y=element_text(size=unit(8,'pt'),family='sans')
    ,strip.text.x=element_text(size=unit(9,'pt'),family='sans')
    ,strip.text.y=element_text(size=unit(9,'pt'),family='sans',angle=0)
    ,legend.title=element_text(size=unit(9,'pt'),family='sans')
  )

ggsave(
  filename="figure3.eps"
  ,device=cairo_ps
  ,width=190
  ,height=117.602/165.1*190
  ,unit='mm'
)
```

```{r Training evaluation of the best model, eval=FALSE, include=FALSE}
disease_model3$pc_gbm$pc3 %>%
  rownames_to_column(var='id') %>%
  filter(outcome!='missing') %>%
  select(-weight) %>%
  column_to_rownames(var='id') %>%
  cbind(predict(disease_model3$pc_gbm$model,newdata=.,type='prob')) %>%
  rownames_to_column(var='id') %>%
  select(id,outcome,event) %>%
  rename(prob=event) %>%
  column_to_rownames(var='id') %>%
  cbind(
    suppressWarnings(
        predict(disease_model3$pc_gbm$calib_model,newdata=.,type='prob')
      ) %>%
      rename_all(function(x)paste0(x,'2'))
  ) %>%
  select(-prob) %>%
  rename_all(function(x)str_remove_all(x,'2')) %>%
  mutate(obs=outcome) %>%
  select(-outcome) %>%
  select(nonevent,everything()) %>%
  evalm(silent=T,showplots=F)
```

```{r Optimal training-thresholds the best model, eval=FALSE, include=FALSE}
disease_model3$pc_gbm$pc3 %>%
  rownames_to_column(var='id') %>%
  filter(outcome!='missing') %>%
  select(-weight) %>%
  column_to_rownames(var='id') %>%
  cbind(predict(disease_model3$pc_gbm$model,newdata=.,type='prob')) %>%
  rownames_to_column(var='id') %>%
  select(id,outcome,event) %>%
  rename(prob=event) %>%
  column_to_rownames(var='id') %>%
  cbind(
    suppressWarnings(predict(
        disease_model3$pc_gbm$calib_model,newdata=.,type='prob')
      ) %>%
      rename_all(function(x)paste0(x,'2'))
  ) %>%
  select(-prob) %>%
  rename_all(function(x)str_remove_all(x,'2')) %>%
  mutate(obs=outcome) %>%
  select(-outcome) %>%
  select(nonevent,everything()) %>%
  cbind(
    data.frame(
        th=paste0('th_',str_pad(seq(0,100),3,'left','0'))
        ,value=seq(0,1,0.01)
      ) %>%
      spread(th,value)
  ) %>%
  gather(key,th,-nonevent,-event,-obs) %>%
  select(-key) %>%
  mutate(pred=factor(ifelse(event>=th,'event','nonevent'),levels(obs))) %>%
  mutate(
    cm=
      case_when(
        obs=='event' & pred=='event' ~ 'tp'
        ,obs=='event' & pred=='nonevent' ~ 'fn'
        ,obs=='nonevent' & pred=='event' ~ 'fp'
        ,obs=='nonevent' & pred=='nonevent' ~ 'tn'
      )
  ) %>%
  group_by(th,cm) %>%
  summarize(n=n()) %>%
  spread(cm,n,fill=0) %>%
  mutate(
    sens=round(tp/(tp+fn),2)
    ,spec=round(tn/(tn+fp),2)
    ,ppv=round(tp/(tp+fp),2)
    ,npv=round(tn/(tn+fn),2)
  ) %>%
  # Based on the 95% CI of the optimal threshold
  # in the evaluation of the best model
  filter(sens>=0.64 & sens<=0.90) %>%
  filter(spec>=0.72 & spec<=0.81) %>%
  filter(ppv>=0.21 & ppv<=0.38) %>%
  filter(npv>=0.94 & npv<=0.99)
```

```{r Other evaluations, eval=FALSE, fig.height=10, fig.width=10, include=FALSE}
disease_model3$pc_gbm$pc3 %>%
  rownames_to_column(var='id') %>%
  filter(outcome!='missing') %>%
  select(-weight) %>%
  column_to_rownames(var='id') %>%
  cbind(predict(disease_model3$pc_gbm$model,newdata=.,type='prob')) %>%
  rownames_to_column(var='id') %>%
  select(id,outcome,event) %>%
  rename(prob=event) %>%
  column_to_rownames(var='id') %>%
  cbind(
    suppressWarnings(predict(
        disease_model3$pc_gbm$calib_model,newdata=.,type='prob')
      ) %>%
      rename_all(function(x)paste0(x,'2'))
  ) %>%
  select(-prob) %>%
  rename_all(function(x)str_remove_all(x,'2')) %>%
  mutate(obs=outcome) %>%
  select(-outcome) %>%
  select(nonevent,everything()) %>%
  cbind(
    data.frame(
        th=paste0('th_',str_pad(seq(0,100),3,'left','0'))
        ,value=seq(0,1,0.01)
      ) %>%
      spread(th,value)
  ) %>%
  gather(key,th,-nonevent,-event,-obs) %>%
  select(-key) %>%
  mutate(pred=factor(ifelse(event>=th,'event','nonevent'),levels(obs))) %>%
  mutate(
    cm=
      case_when(
        obs=='event' & pred=='event' ~ 'tp'
        ,obs=='event' & pred=='nonevent' ~ 'fn'
        ,obs=='nonevent' & pred=='event' ~ 'fp'
        ,obs=='nonevent' & pred=='nonevent' ~ 'tn'
      )
  ) %>%
  group_by(th,cm) %>%
  summarize(n=n()) %>%
  spread(cm,n,fill=0) %>%
  mutate(
    sens=tp/(tp+fn)
    ,spec=tn/(tn+fp)
    ,ppv=tp/(tp+fp)
    ,npv=tn/(tn+fn)
    ,nb=tp/(tp+fn+fp+tn)-fp/(tp+fn+fp+tn)*th/(1-th)
    ,p=(tp+fn)/(tp+fn+fp+tn)
    ,pg=(ppv-p)/(ppv*(1-p))
    ,rg=(sens-p)/(sens*(1-p))
    ,inn=sens-spec
    ,mrn=ppv-npv
  ) %>%
  ungroup() %>%
  lapply(X=1,Y=.,function(X,Y){
    print(Y)
    Y %>%
      lapply(X=c('roc','prc','dca','prg','imc'),Y=.,function(X,Y){
        if(X=='roc'){
          Y %>%
            select(spec,sens,th) %>%
            mutate(fpr=1-spec) %>%
            select(-spec) %>%
            rename(x=fpr,y=sens,z=th) %>%
            mutate(metrics='roc(1-spec,sens)')
        }else if(X=='prc'){
          Y %>%
            select(sens,ppv,th) %>%
            rename(x=sens,y=ppv,z=th) %>%
            mutate(metrics='prc(sens,ppv)') %>%
            mutate(y=ifelse(is.nan(y),1,y))
        }else if(X=='dca'){
          Y %>%
            select(th,nb) %>%
            rename(x=th,y=nb) %>%
            mutate(z=x) %>%
            mutate(metrics='dca(th,nb)')
        }else if(X=='prg'){
          Y %>%
            select(rg,pg,th) %>%
            rename(x=rg,y=pg,z=th) %>%
            mutate(metrics='prg(rg,pg)') %>%
            filter(x>=0 & y>=0)
        }else if(X=='imc'){
          Y %>%
            select(inn,mrn,th) %>%
            rename(x=inn,y=mrn,z=th) %>%
            mutate(metrics='imc(inn,mrn)')
        }
      }) %>%
      do.call(rbind,.) %>%
      filter(!is.nan(x) & !is.nan(y) & is.finite(x) & is.finite(y)) %>%
      ggplot(aes(x,y,label=x)) +
      geom_line() +
      geom_point() +
      geom_text(aes(label=round(z,2)),size=2,hjust=0,vjust=0) +
      geom_abline(
        aes(
          slope=
            ifelse(
              metrics!='dca(th,nb)'
              ,ifelse(
                metrics!='roc(1-spec,sens)'
                ,NA
                ,1
              )
              ,-1
            )
          ,intercept=
            ifelse(
              metrics!='dca(th,nb)'
              ,ifelse(
                metrics!='roc(1-spec,sens)'
                ,NA
                ,0
              )
              ,mean(Y$p)
            )
        )
        ,lty=2,color='red'
        ,na.rm=T
      ) +
      geom_hline(
        aes(
         yintercept=
            ifelse(
              metrics!='dca(th,nb)'
              ,NA
              ,0
            )
        )
        ,lty=2,color='red'
        ,na.rm=T
      ) +
      facet_wrap(~metrics,scales='free')
  }) %>%
  .[[1]]
```

```{r ROC curves of the best model in all the datasets, echo=FALSE}
disease_model3$pc_gbm$eval_model$roc
opt_pred3_disease1$pc_gbm$test_eval$roc
opt_pred3_disease2$pc_gbm$test_eval$roc
opt_pred3_disease3$pc_gbm$test_eval$roc
```

```{r Compare the PPVs, include=FALSE}
ppv_comparison=result_comparison('PREC')

tmp=
  ppv_comparison %>%
  filter(tissue!='all') %>%
  filter(model=='pc_gbm' & tissue=='repro') %>%
  select(-model,-tissue)
```

```{r Compare the NPVs, include=FALSE}
npv_comparison=result_comparison('NPV')

tmp2=
  npv_comparison %>%
  filter(tissue!='all') %>%
  filter(model=='pc_gbm' & tissue=='repro') %>%
  select(-model,-tissue)
```

```{r Show the PPVs and NPVs, echo=FALSE}
tmp %>%
  select(set,Score,LB,UB) %>%
  unite(CI,LB,UB,sep=' to ') %>%
  rename_at(c('Score','CI'),paste0,'_PPV') %>%
  left_join(
    tmp2 %>%
      select(set,Score,LB,UB) %>%
      unite(CI,LB,UB,sep=' to ') %>%
      rename_at(c('Score','CI'),paste0,'_NPV')
    ,by='set'
  ) %>%
  mutate(
    set=
      set %>%
      factor(c('GSE108497','GSE85307','GSE86200','GSE149437'))
  ) %>%
  arrange(set)
```

```{r}
timestamper[12]=as.character(now())
```

```{r Get PC weights from the best model, include=FALSE}
if(run_heavy_computation){
  tmp3=
    disease_model3$pc_gbm$model %>%
    varImp() %>%
    .[[1]] %>%
    rownames_to_column(var='term') %>%
    
    # Clean up.
    filter(Overall!=0) %>%
    arrange(desc(Overall)) %>%
    
    left_join(
      disease_model3$pc_gbm$transformator$avg_rotm %>%
        as.data.frame() %>%
        rownames_to_column(var='feature') %>%
        gather(term,weight,-feature)
      ,by='term'
    ) %>%
    select(-Overall) %>%
    separate(feature,c('tissue','gene'),sep=':') %>%
    mutate(w_direction=ifelse(weight>=0,'positive','negative')) %>%
    group_by(term,tissue,w_direction) %>%
    arrange(term,tissue,w_direction,desc(abs(weight))) %>%
    ungroup() %>%
    left_join(
      pull(.,gene) %>%
        .[!duplicated(.)] %>%
        pblapply(function(x){
          y=readRDS(paste0('data/deg_models/',x,'.rds'))
          
          coef(y$model$finalModel,y$model$bestTune$lambda) %>%
            as.matrix() %>%
            as.data.frame() %>%
            rownames_to_column(var='term') %>%
            
            # Clean up.
            mutate(term=str_remove_all(term,'\\`|\\(|\\)')) %>%
            setNames(c('term','estimate')) %>%
            filter(estimate!=0) %>%
            arrange(desc(abs(estimate))) %>%
            filter(term!='Intercept') %>%
            left_join(
              y$transformator$avg_rotm %>%
                t() %>%
                as.data.frame() %>%
                rownames_to_column(var='term')
              ,by='term'
            ) %>%
            gather(blood_t,bt_weight,-term,-estimate) %>%
            rename(surro_pc=term,spc_weight=estimate) %>%
            mutate(gene=x) %>%
            select(gene,blood_t,bt_weight,surro_pc,spc_weight)
        }) %>%
        do.call(rbind,.)
      ,by='gene'
    ) %>%
    left_join(
      opt_pred_deg %>%
        do.call(rbind,.) %>%
        rownames_to_column(var='tissue') %>%
        separate(tissue,c('tissue','seq'),sep='\\.') %>%
        select(-seq) %>%
        rename(gene=deg)
      ,by=c('tissue','gene')
    )
  
  tmp3 %>%
    pull(tissue) %>%
    .[!duplicated(.)] %>%
    sapply(function(x){
      tmp3 %>%
        filter(tissue==x) %>%
        saveRDS(paste0('data/tmp3_',x,'.rds'))
    })
}else{
  cat(readRDS('data/log.rds')[['tmp3']])
  tmp3=
    list.files('data/',pattern='tmp3_',full.names=T) %>%
    lapply(readRDS) %>%
    do.call(rbind,.) %>%
    arrange(term,tissue)
}
```

```{r Get PC weights from the best model, include=FALSE}
tmp3b=
  tmp3 %>%
  group_by(term,tissue,gene,weight,w_direction,mcc,blood_t) %>%
  summarize(accu_w=sum(bt_weight*spc_weight,na.rm=T),n=n()) %>%
  ungroup() %>%
  group_by(tissue,blood_t) %>%
  summarize(accu_w=sum(accu_w*mcc*weight,na.rm=T),n=sum(n,na.rm=T)) %>%
  ungroup() %>%
  mutate(avg_accu_w=accu_w/n) %>%
  arrange(desc(abs(avg_accu_w))) %>%
  group_by(tissue) %>%
  mutate(rank=seq(n())) %>%
  ungroup()

# 5 biomarkers is considerably low-cost
# 20 is minimum rank per tissue to include all of them
# after filtering top 5 in any tissues
if(run_heavy_computation){
  tmp3c=
    seq(20) %>%
    pblapply(function(x){
      seq(5) %>%
        lapply(X=.,Y=x,function(X,Y){
          tmp3 %>%
            left_join(tmp3b,by=c('tissue','blood_t')) %>%
            filter(rank%in%seq(Y)) %>%
            select(tissue,gene,blood_t,avg_accu_w,rank) %>%
            filter(!duplicated(.)) %>%
            group_by(blood_t) %>%
            mutate(n_tissue=length(tissue %>% .[!duplicated(.)])) %>%
            arrange(desc(n_tissue),tissue,rank) %>%
            ungroup() %>%
            left_join(
              select(.,blood_t) %>%
                filter(!duplicated(.)) %>%
                mutate(global_rank=seq(n()))
              ,by='blood_t'
            ) %>%
            filter(global_rank<=X) %>%
            mutate(grank=X)
        }) %>%
        do.call(rbind,.) %>%
        mutate(rank=x)
    }) %>%
    do.call(rbind,.)
  saveRDS(tmp3c,'data/tmp3c.rds')
}else{
  cat(readRDS('data/log.rds')[['tmp3c']])
  tmp3c=readRDS('data/tmp3c.rds')
}
```


```{r Ensure all tissues included by the criteria, eval=FALSE, include=FALSE}
tmp3c %>%
  group_by(tissue) %>%
  summarize(n=n()) %>%
  mutate(total=sum(n))
```

```{r Show all the combinations, eval=FALSE, include=FALSE}
tmp3c %>%
  pull(rank) %>%
  .[!duplicated(.)] %>%
  sort() %>%
  lapply(function(x){
    tmp3c %>%
      filter(rank==x) %>%
      pull(grank) %>%
      .[!duplicated(.)] %>%
      sort() %>%
      lapply(X=.,Y=x,function(X,Y){
        tmp3c %>%
          filter(grank==X & rank==Y) %>%
          unite(tissue_gene_bloodt,tissue,gene,blood_t,sep=':') %>%
          pull(tissue_gene_bloodt) %>%
          paste0(collapse=', ') %>%
          as.data.frame() %>%
          setNames(paste0('grank',X))
      }) %>%
      do.call(cbind,.) %>%
      `rownames<-`(paste0('rank',x))
  }) %>%
  do.call(rbind,.) %>%
  knitr::kable()
```

```{r Create a function to preprocess the features, include=FALSE}
mb_predictor=function(exprs_mat,the_dea_list1){
  
  the_dea=the_dea_list1[[1]]
  the_tissue=names(the_dea_list1)[1]
  
  exprs_mat %>%
    dea_qn(the_dea$result) %>%
    as.data.frame() %>%
    rownames_to_column(var='feature') %>%
    left_join(
      the_dea$result %>%
        rownames_to_column(var='feature') %>%
        select(feature,AveExpr,AveExpr1,AveExpr2)
      ,by='feature'
    ) %>%
    gather(id,value,-feature,-AveExpr,-AveExpr1,-AveExpr2) %>%
    mutate(
      norm_value=
        ifelse(
          AveExpr1<AveExpr
          ,ifelse(
            value<AveExpr
            ,0-(value-AveExpr)/(AveExpr1-AveExpr)
            ,0+(value-AveExpr)/(AveExpr2-AveExpr)
          )
          ,ifelse(
            value>AveExpr
            ,0-(value-AveExpr)/(AveExpr1-AveExpr)
            ,0+(value-AveExpr)/(AveExpr2-AveExpr)
          )
        )
    ) %>%
    select(-value,-AveExpr,-AveExpr1,-AveExpr2) %>%
    spread(feature,norm_value) %>%
    column_to_rownames(var='id') %>%
    setNames(paste0(the_tissue,':',colnames(.)))
}
```

```{r Create a function to standardize features ignoring outlier, include=FALSE}
std_no_outlier=function(input,scaler){
  scaler=
    scaler %>%
    select(-outcome) %>%
    lapply(function(x){
      median=quantile(x,0.5)
      q1=quantile(x,0.25)
      q3=quantile(x,0.75)
      minq=q1-1.5*(q3-q1)
      maxq=q3+1.5*(q3-q1)
      outliers=x<minq | x>maxq
      avg=
        x %>%
        .[!outliers] %>%
        mean()
      
      std=
        x %>%
        .[!outliers] %>%
        sd()
      
      list(avg=avg,std=std)
    })
  
  input %>%
    lapply(X=seq(ncol(.)),Y=.,Z=scaler,function(X,Y,Z){
      if(X==1){
        Y[[X]]
      }else{
        (Y[[X]]-Z[[colnames(Y)[X]]]$avg)/Z[[colnames(Y)[X]]]$std
      }
    }) %>%
    as.data.frame() %>%
    `dimnames<-`(dimnames(input))
}
```

```{r Create a function to re-evaluate the biomarker tree leaves, include=FALSE}
re_evalm=function(data
                  ,train_mod
                  ,metric_list
                  ,ref_list
                  ,nonevent_name='nonevent'
                  ,event_name='event'){
  eval=
    train_mod %>%
    predict(type='prob') %>%
    cbind(select(train_mod$trainingData,.outcome)) %>%
    rename(outcome=.outcome) %>%
    rename_at(event_name,function(x)'prob') %>%
    select(prob,outcome) %>%
    cbind(
      data.frame(
          th=paste0('th_',str_pad(seq(0,100),'3','left','0'))
          ,value=seq(0,1,0.01)
        ) %>%
        spread(th,value)
    ) %>%
    gather(key,th,-prob,-outcome) %>%
    select(-key) %>%
    mutate(
      pred=factor(ifelse(prob>=th,event_name,nonevent_name),levels(outcome))
    ) %>%
    group_by(th) %>%
    summarize(
      tp=sum(outcome==event_name & pred==event_name)
      ,fn=sum(outcome==event_name & pred==nonevent_name)
      ,fp=sum(outcome==nonevent_name & pred==event_name)
      ,tn=sum(outcome==nonevent_name & pred==nonevent_name)
    ) %>%
    mutate(
      sens=tp/(tp+fn)
      ,spec=tn/(fp+tn)
      ,ppv=tp/(tp+fp)
      ,npv=tn/(tn+fn)
      ,nb=(tp-fp*(th/(1-th)))/(tp+fn+fp+tn)
      ,mcc=(tp*tn-fp*fn)/sqrt((tp+fp)*(tp+fn)*(tn+fp)*(tn+fn))
    ) %>%
    gather(metric,value,-th)
  
  th_value=
    metric_list %>%
    lapply(X=seq(length(.)),Y=.,Z=ref_list,function(X,Y,Z){
      if(length(Y[[X]])>1){
        K=eval %>%
          group_by(th) %>%
          mutate(
            sd=
              ifelse(metric%in%Y[[X]],value,NA) %>%
              sd(na.rm=T)
          ) %>%
          ungroup() %>%
          mutate(
            th_value=
              ifelse(sd==min(sd,na.rm=T),th,NA) %>%
              median(na.rm=T) %>%
              round(2)
          ) %>%
          select(-sd)
      }else{
        K=eval %>%
          group_by(th) %>%
          mutate(
            metric_closest=
              abs(ifelse(metric%in%Y[[X]],value,NA)-Z[[X]]) %>%
              min(na.rm=T)
          ) %>%
          ungroup() %>%
          mutate(
            th_value=
              ifelse(metric_closest==min(metric_closest,na.rm=T),th,NA) %>%
              median(na.rm=T) %>%
              round(2)
          ) %>%
          select(-metric_closest)
      }
      
      K %>%
        select(th_value) %>%
        rename_at('th_value',function(x)paste0(x,X))
    }) %>%
    do.call(cbind,.)
  
  th_list=
    eval %>%
    cbind(th_value) %>%
    gather(th_type,th_value,-th,-metric,-value) %>%
    mutate(
      th_metrics=
        th_type %>%
        sapply(function(x){
          y=str_remove_all(x,'th_value') %>%
            as.numeric()
          paste0(
            paste0(metric_list[[y]],collapse=',')
            ,'|'
            ,paste0(ref_list[[y]],collapse=',')
          )
        })
      ,th_metrics=
        ifelse(
          str_detect(th_metrics,',')
          ,str_remove_all(th_metrics,'\\|')
          ,th_metrics
        )
    ) %>%
    mutate(th_diff=abs(th-th_value)) %>%
    group_by(th_type) %>%
    filter(th_diff==min(th_diff)) %>%
    ungroup() %>%
    select(-th_diff) %>%
    spread(metric,value) %>%
    select(th,th_metrics) %>%
    lapply(X=seq(nrow(.)),Y=.,function(X,Y)Y[X,,drop=T]) %>%
    lapply(as.character) %>%
    lapply(setNames,c('th','th_metrics'))
  
  th_list %>%
    lapply(function(x){
      y=train_mod %>%
        predict(newdata=data,type='prob') %>%
        cbind(select(data,outcome)) %>%
        rename_at(event_name,function(x)'prob') %>%
        select(prob,outcome) %>%
        mutate(
          pred=
            factor(
              ifelse(prob>=x['th'],event_name,nonevent_name)
              ,levels(outcome)
            )
        )
      
      suppressWarnings(set.seed(33,sample.kind=sample.kind))
      seq(30) %>%
        lapply(X=.,function(X){
          y %>%
            .[sample(seq(nrow(.)),nrow(.),T),] %>%
            summarize(
              b=X
              ,tp=sum(outcome==event_name & pred==event_name)
              ,fn=sum(outcome==event_name & pred==nonevent_name)
              ,fp=sum(outcome==nonevent_name & pred==event_name)
              ,tn=sum(outcome==nonevent_name & pred==nonevent_name)
            ) %>%
            mutate(
              sens=tp/(tp+fn)
              ,spec=tn/(fp+tn)
              ,ppv=tp/(tp+fp)
              ,npv=tn/(tn+fn)
              ,nb=
                (tp-fp*(as.numeric(x['th'])/(1-as.numeric(x['th']))))
                /(tp+fn+fp+tn)
              ,mcc=(tp*tn-fp*fn)/sqrt((tp+fp)*(tp+fn)*(tn+fp)*(tn+fn))
            )
        }) %>%
        do.call(rbind,.) %>%
        gather(metric,value,-b) %>%
        mutate(value=ifelse(is.nan(value),NA,value)) %>%
        group_by(metric) %>%
        summarize(
          Score=
            mean(value,na.rm=T)
          ,LB=
            mean(value,na.rm=T)
            -qnorm(0.975)*sd(value,na.rm=T)/sqrt(sum(!is.na(value)))
          ,UB=
            mean(value,na.rm=T)
            +qnorm(0.975)*sd(value,na.rm=T)/sqrt(sum(!is.na(value)))
        ) %>%
        mutate(
          metric=
            factor(
              str_to_upper(metric)
              ,str_to_upper(
                c('tp','fn','fp','tn','sens','spec','ppv','npv','nb','mcc')
              )
            )
        ) %>%
        arrange(metric) %>%
        mutate(
          th=x['th']
          ,th_metrics=x['th_metrics']
        ) %>%
        select(th,th_metrics,everything())
    }) %>%
    do.call(rbind,.)
}
```

```{r Determine metric and reference to choose thresholds}
th_selection=list(c('sens','spec'))
th_values=list(NULL)
```

```{r Get features to train biomarker tree and evaluate it, include=FALSE}
if(run_heavy_computation){
  # Preprocess the features per combination in the development dataset
  tmp3d=
    tmp3c %>%
    pull(rank) %>%
    .[!duplicated(.)] %>%
    sort() %>%
    pblapply(X=.
             ,Y=GSE108497_phenotype %>%
                select(outcome) %>%
                rownames_to_column(var='id') %>%
                left_join(
                  GSE108497 %>%
                    exprs() %>%
                    mb_predictor(
                      list(maternal_blood=GSE108497_maternal_blood)
                    ) %>%
                    rownames_to_column(var='id')
                  ,by='id'
                ) %>%
                column_to_rownames(var='id')
             ,function(X,Y){
      tmp3c %>%
        filter(rank==X) %>%
        pull(grank) %>%
        .[!duplicated(.)] %>%
        sort() %>%
        lapply(X=.,Y=X,Z=Y,function(X,Y,Z){
          Z %>%
            select_at(c(
              'outcome'
              ,paste0(
                'maternal_blood:'
                ,tmp3c$blood_t[tmp3c$grank==X & tmp3c$rank==Y]
              )
            ))
        }) %>%
        setNames(paste0(
          'grank'
          ,tmp3c %>%
            filter(rank==X) %>%
            pull(grank) %>%
            .[!duplicated(.)] %>%
            sort()
        ))
    }) %>%
    setNames(paste0(
      'rank'
      ,tmp3c %>%
        pull(rank) %>%
        .[!duplicated(.)] %>%
        sort()
    ))
  
  saveRDS(tmp3d,'data/tmp3d.rds')
  
  # Create multiple CPU clusters for parallel computing.
  cl=detectCores()-2
  cl=makePSOCKcluster(cl)
  registerDoParallel(cl)
  
  # Close the CPU clusters and clean up memory at exit.
  on.exit(stopCluster(cl))
  on.exit(registerDoSEQ())
  on.exit(rm(cl))
  on.exit(rm(the_method))
  on.exit(gc())
  
  # Bring these packages to the CPU clusters.
  clusterEvalQ(cl,{
    library('parallel')
    library('doParallel')
    library('tidyverse')
    library('pbapply')
    library('caret')
    library('rpart')
  })
  
  # Train the biomarker tree models
  tmp3e=
    tmp3d %>%
    pblapply(X=.
             ,training_parameters=training_parameters
             ,sample.kind=sample.kind
             ,std_no_outlier=std_no_outlier
             ,cl=cl
             ,function(X,training_parameters,sample.kind,std_no_outlier){
      X %>%
        lapply(function(x){
          suppressWarnings(set.seed(33,sample.kind=sample.kind))
          suppressWarnings(
            x %>%
              std_no_outlier(x) %>%
              `colnames<-`(
                str_remove_all(colnames(.),'maternal_blood:')
              ) %>%
              caret::train(
                outcome~.
                ,data=.
                ,method='rpart'
                ,weights=
                  x %>%
                  `colnames<-`(
                    str_remove_all(colnames(.),'maternal_blood:')
                  ) %>%
                  pull(outcome) %>%
                  table() %>%
                  as.data.frame() %>%
                  setNames(c('outcome','Freq')) %>%
                  mutate(total=sum(Freq)) %>%
                  mutate(weight=1/(Freq/total)*0.5) %>%
                  right_join(
                    x %>%
                      rownames_to_column(var='id')
                    ,by='outcome'
                  ) %>%
                  pull(weight) %>%
                  setNames(rownames(x))
                ,metric='ROC'
                ,trControl=training_parameters$final_trControl
                ,tuneLength=10
                ,control=
                  rpart.control(
                    maxdepth=
                      sum(str_detect(colnames(x),'maternal_blood'))
                  )
                ,parms=list(split='gini')
              )
          )
        })
    })
  
  tmp3e %>%
    lapply(X=names(.),Y=.,function(X,Y){
      saveRDS(Y[[X]],paste0('data/tmp3e/tmp3e_',X,'.rds'))
    })
  
  # Evaluate the biomarker tree models using development dataset
  tmp3f=
    tmp3e %>%
    pblapply(function(x){
      x %>%
        lapply(function(x){
          x$trainingData %>%
            rename(outcome=.outcome) %>%
            re_evalm(x,th_selection,th_values) %>%
            column_to_rownames(var='metric')
        })
    })
  
  saveRDS(tmp3f,'data/tmp3f.rds')
}else{
  cat(readRDS('data/log.rds')[['tmp3d']],'\n')
  cat(readRDS('data/log.rds')[['tmp3e']],'\n')
  cat(readRDS('data/log.rds')[['tmp3f']])
  tmp3d=readRDS('data/tmp3d.rds')
  tmp3e=
    list.files('data/tmp3e/',pattern='tmp3e_',full.names=T) %>%
    .[order(as.numeric(str_remove_all(.,'data/tmp3e/tmp3e_|rank|\\.rds')))] %>%
    lapply(readRDS) %>%
    setNames(
      list.files('data/tmp3e/',pattern='tmp3e_',full.names=T) %>%
        .[order(as.numeric(
          str_remove_all(.,'data/tmp3e/tmp3e_|rank|\\.rds')
        ))] %>%
        str_remove_all('data/tmp3e/tmp3e_|\\.rds')
    )
  tmp3f=readRDS('data/tmp3f.rds')
}
```

```{r Get features of the replication dataset 3 and evaluate it, include=FALSE}
if(run_heavy_computation){
  tmp3d2=
    tmp3c %>%
    pull(rank) %>%
    .[!duplicated(.)] %>%
    sort() %>%
    pblapply(X=.
           ,Y=GSE149437_phenotype %>%
              select(outcome) %>%
              rownames_to_column(var='id') %>%
              left_join(
                GSE149437 %>%
                  exprs() %>%
                  mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
                  rownames_to_column(var='id')
                ,by='id'
              ) %>%
              column_to_rownames(var='id')
           ,function(X,Y){
      tmp3c %>%
        filter(rank==X) %>%
        pull(grank) %>%
        .[!duplicated(.)] %>%
        sort() %>%
        lapply(X=.,Y=X,Z=Y,function(X,Y,Z){
          Z %>%
            select_at(c(
              'outcome'
              ,paste0(
                'maternal_blood:'
                ,tmp3c$blood_t[tmp3c$grank==X & tmp3c$rank==Y]
              )
            ))
        }) %>%
        setNames(paste0(
          'grank'
          ,tmp3c %>%
            filter(rank==X) %>%
            pull(grank) %>%
            .[!duplicated(.)] %>%
            sort()
        ))
    }) %>%
    setNames(paste0(
      'rank'
      ,tmp3c %>%
        pull(rank) %>%
        .[!duplicated(.)] %>%
        sort()
    ))
  
  saveRDS(tmp3d2,'data/tmp3d2.rds')
  
  tmp3f2=
    tmp3e %>%
    pblapply(X=names(.),Y=.,function(X,Y){
      Y[[X]] %>%
        lapply(X=names(.),Y=.,Z=X,function(X,Y,Z){
          tmp3d2[[Z]][[X]] %>%
            std_no_outlier(tmp3d[[Z]][[X]]) %>%
            `colnames<-`(
              str_remove_all(colnames(.),'maternal_blood:')
            ) %>%
            re_evalm(Y[[X]],th_selection,th_values) %>%
            column_to_rownames(var='metric')
        }) %>%
        setNames(names(Y[[X]]))
    }) %>%
    setNames(names(tmp3e))
  
  saveRDS(tmp3f2,'data/tmp3f2.rds')
}else{
  cat(readRDS('data/log.rds')[['tmp3d2']],'\n')
  cat(readRDS('data/log.rds')[['tmp3f2']])
  tmp3d2=readRDS('data/tmp3d2.rds')
  tmp3f2=readRDS('data/tmp3f2.rds')
}
```

```{r Check if the combinations are devDEGs, eval=FALSE, include=FALSE}
tmp3c %>%
  pull(rank) %>%
  .[!duplicated(.)] %>%
  sort() %>%
  sapply(function(x){
    tmp3c %>%
      filter(rank==x) %>%
      pull(grank) %>%
      .[!duplicated(.)] %>%
      sort() %>%
      sapply(X=.,Y=x,function(X,Y){
        Z=tmp3c$blood_t[tmp3c$grank==X & tmp3c$rank==Y] %>%
          .[!duplicated(.)]
        K=GSE108497_maternal_blood$result %>%
          filter(adj.P.Val<0.05) %>%
          rownames()
        mean(Z%in%K)
      }) %>%
      setNames(paste0(
        'grank'
        ,tmp3c %>%
          filter(rank==x) %>%
          pull(grank) %>%
          .[!duplicated(.)] %>%
          sort()
      ))
  }) %>%
  `colnames<-`(paste0(
    'rank'
    ,tmp3c %>%
      pull(rank) %>%
      .[!duplicated(.)] %>%
      sort()
  ))
```

```{r Check if the best-mod features are devrep3DEGs, eval=FALSE, include=FALSE}
tmp3$gene %>%
  .[!duplicated(.)] %>%
  intersect(
    GSE108497_maternal_blood$result %>%
      filter(adj.P.Val<0.05) %>%
      rownames()
  ) %>%
  intersect(
    GSE108497_maternal_blood$result %>%
      rownames_to_column(var='gene') %>%
      filter(adj.P.Val<0.05) %>%
      filter(
        gene%in%rownames(filter(
          GSE149437_maternal_blood$result
          ,adj.P.Val<0.05
        ))
      ) %>%
      column_to_rownames(var='gene') %>%
      rownames()
  )
```

```{r Get features to train logFC2 DEG and evaluate it, include=FALSE}
# Preprocess the features of logFC2 DEG in the development dataset
tmp3d3=
  GSE108497_phenotype %>%
  select(outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    GSE108497 %>%
      exprs() %>%
      mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  column_to_rownames(var='id') %>%
  select_at(c(
    'outcome'
    ,paste0(
      'maternal_blood:'
      ,GSE108497_maternal_blood$result %>%
        filter(adj.P.Val<0.05 & abs(logFC)>2) %>%
        rownames()
    )
  ))

# Preprocess the features of logFC2 DEG in the replication dataset 3
tmp3d4=
  GSE149437_phenotype %>%
  select(outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    GSE149437 %>%
      exprs() %>%
      mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  column_to_rownames(var='id') %>%
  select_at(c(
    'outcome'
    ,paste0(
      'maternal_blood:'
      ,GSE108497_maternal_blood$result %>%
        filter(adj.P.Val<0.05 & abs(logFC)>2) %>%
        rownames()
    )
  ))

# Train the logFC2-DEG tree model
tmp3e3=
  tmp3d3 %>%
  lapply(X=1,Y=.,function(X,Y){
    suppressWarnings(set.seed(33,sample.kind=sample.kind))
    suppressWarnings(
      Y %>%
        std_no_outlier(Y) %>%
        `colnames<-`(
          str_remove_all(colnames(.),'maternal_blood:')
        ) %>%
        caret::train(
          outcome~.
          ,data=.
          ,method='rpart'
          ,weights=
            Y %>%
            `colnames<-`(
              str_remove_all(colnames(.),'maternal_blood:')
            ) %>%
            pull(outcome) %>%
            table() %>%
            as.data.frame() %>%
            setNames(c('outcome','Freq')) %>%
            mutate(total=sum(Freq)) %>%
            mutate(weight=1/(Freq/total)*0.5) %>%
            right_join(
              Y %>%
                rownames_to_column(var='id')
              ,by='outcome'
            ) %>%
            pull(weight) %>%
            setNames(rownames(Y))
          ,metric='ROC'
          ,trControl=training_parameters$final_trControl
          ,tuneLength=10 
          ,control=
            rpart.control(
              maxdepth=
                sum(str_detect(colnames(Y),'maternal_blood'))
            )
          ,parms=list(split='gini')
        )
    )
  }) %>%
  .[[1]]

# Evaluate the logFC2-DEG tree model using development dataset
tmp3f3=
  tmp3d3 %>%
  std_no_outlier(tmp3d3) %>%
  `colnames<-`(
    str_remove_all(colnames(.),'maternal_blood:')
  ) %>%
  re_evalm(tmp3e3,th_selection,th_values) %>%
  column_to_rownames(var='metric')

# Evaluate the logFC2-DEG tree model using replication dataset 3
tmp3f4=
  tmp3d4 %>%
  std_no_outlier(tmp3d3) %>%
  `colnames<-`(
    str_remove_all(colnames(.),'maternal_blood:')
  ) %>%
  re_evalm(tmp3e3,th_selection,th_values) %>%
  column_to_rownames(var='metric')
```

```{r Get features to train devrep3DEG and evaluate it, include=FALSE}
# Get the possible combinations of devrep3DEG
tmp3c2=
  GSE108497_maternal_blood$result %>%
  filter(adj.P.Val<0.05) %>%
  rownames() %>%
  intersect(
    GSE149437_maternal_blood$result %>%
      filter(adj.P.Val<0.05) %>%
      rownames()
  ) %>%
  lapply(X=seq(2),Y=.,function(X,Y){
    if(X==1){
      as.list(Y)
    }else{
      rep(list(Y),X) %>%
        expand.grid() %>%
        lapply(X=seq(nrow(.)),Y=.,Z=X,function(X,Y,Z){
          Y=Y %>%
            mutate_all(as.character)
          K=c()
          for(i in seq(Z)){
            K[i]=Y[[i]][X]
          }
          K %>%
            sort() %>%
            data.frame(value=.) %>%
            mutate(key=paste0('Var',seq(Z))) %>%
            spread(key,value)
        }) %>%
        do.call(rbind,.) %>%
        filter(!duplicated(.)) %>%
        mutate(seq=seq(nrow(.))) %>%
        gather(key,value,-seq) %>%
        arrange(seq,key) %>%
        group_by(seq) %>%
        filter(!duplicated(value)) %>%
        filter(n()==X) %>%
        lapply(X=.$seq %>% .[!duplicated(.)],Y=.,function(X,Y){
          Y %>%
            filter(seq==X) %>%
            pull(value)
        })
    }
  }) %>%
  lapply(X=1,Y=.,function(X,Y){
    for(i in seq(length(Y))){
      if(i==1){
        Z=Y[[i]]
      }else{
        Z=Z %>%
          c(Y[[i]])
      }
    }
    Z
  }) %>%
  .[[1]]

if(run_heavy_computation){
  # Preprocess the features of devrep3DEG in the development dataset
  tmp3d5=
    GSE108497_phenotype %>%
    select(outcome) %>%
    rownames_to_column(var='id') %>%
    left_join(
      GSE108497 %>%
        exprs() %>%
        mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
        rownames_to_column(var='id')
      ,by='id'
    ) %>%
    column_to_rownames(var='id') %>%
    pblapply(X=seq(length(tmp3c2)),Y=tmp3c2,Z=.,function(X,Y,Z){
      Z %>%
        select_at(c(
          'outcome'
          ,paste0(
            'maternal_blood:'
            ,Y[[X]]
          )
        ))
    })
  
  saveRDS(tmp3d5,'data/tmp3d5.rds')
  
  # Preprocess the features of devrep3DEG in the replication dataset 3
  tmp3d6=
    GSE149437_phenotype %>%
    select(outcome) %>%
    rownames_to_column(var='id') %>%
    left_join(
      GSE149437 %>%
        exprs() %>%
        mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
        rownames_to_column(var='id')
      ,by='id'
    ) %>%
    column_to_rownames(var='id') %>%
    pblapply(X=seq(length(tmp3c2)),Y=tmp3c2,Z=.,function(X,Y,Z){
      Z %>%
        select_at(c(
          'outcome'
          ,paste0(
            'maternal_blood:'
            ,Y[[X]]
          )
        ))
    })
  
  saveRDS(tmp3d6,'data/tmp3d6.rds')
  
  # Create multiple CPU clusters for parallel computing.
  cl=detectCores()-2
  cl=makePSOCKcluster(cl)
  registerDoParallel(cl)
  
  # Close the CPU clusters and clean up memory at exit.
  on.exit(stopCluster(cl))
  on.exit(registerDoSEQ())
  on.exit(rm(cl))
  on.exit(rm(the_method))
  on.exit(gc())
  
  # Bring these packages to the CPU clusters.
  clusterEvalQ(cl,{
    library('parallel')
    library('doParallel')
    library('tidyverse')
    library('pbapply')
    library('caret')
    library('rpart')
  })
  
  # Train the devrep3DEG tree model
  tmp3e5=
    tmp3d5 %>%
    pblapply(X=.
             ,training_parameters=training_parameters
             ,sample.kind=sample.kind
             ,std_no_outlier=std_no_outlier
             ,cl=cl
             ,function(X,training_parameters,sample.kind,std_no_outlier){
      suppressWarnings(set.seed(33,sample.kind=sample.kind))
      suppressWarnings(
        X %>%
          std_no_outlier(X) %>%
          `colnames<-`(
            str_remove_all(colnames(.),'maternal_blood:')
          ) %>%
          caret::train(
            outcome~.
            ,data=.
            ,method='rpart'
            ,weights=
              X %>%
              `colnames<-`(
                str_remove_all(colnames(.),'maternal_blood:')
              ) %>%
              pull(outcome) %>%
              table() %>%
              as.data.frame() %>%
              setNames(c('outcome','Freq')) %>%
              mutate(total=sum(Freq)) %>%
              mutate(weight=1/(Freq/total)*0.5) %>%
              right_join(
                X %>%
                  rownames_to_column(var='id')
                ,by='outcome'
              ) %>%
              pull(weight) %>%
              setNames(rownames(X))
            ,metric='ROC'
            ,trControl=training_parameters$final_trControl
            ,tuneLength=10 
            ,control=
              rpart.control(
                maxdepth=
                  sum(str_detect(colnames(X),'maternal_blood'))
              )
            ,parms=list(split='gini')
          )
      )
    })
  
  tmp3e5 %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      saveRDS(Y[[X]],paste0('data/tmp3e5/tmp3e5_',X,'.rds'))
    })
  
  # Evaluate the devrep3DEG tree model using development dataset
  tmp3f5=
    tmp3e5 %>%
    pblapply(X=seq(length(.)),Y=.,function(X,Y){
      tmp3d5[[X]] %>%
        std_no_outlier(tmp3d5[[X]]) %>%
        `colnames<-`(
          str_remove_all(colnames(.),'maternal_blood:')
        ) %>%
        re_evalm(Y[[X]],th_selection,th_values) %>%
        column_to_rownames(var='metric')
    })
  
  saveRDS(tmp3f5,'data/tmp3f5.rds')
  
  # Evaluate the devrep3DEG tree model using replication dataset 3
  tmp3f6=
    tmp3e5 %>%
    pblapply(X=seq(length(.)),Y=.,function(X,Y){
      tmp3d6[[X]] %>%
        std_no_outlier(tmp3d5[[X]]) %>%
        `colnames<-`(
          str_remove_all(colnames(.),'maternal_blood:')
        ) %>%
        re_evalm(Y[[X]],th_selection,th_values) %>%
        column_to_rownames(var='metric')
    })
  
  saveRDS(tmp3f6,'data/tmp3f6.rds')
}else{
  cat(readRDS('data/log.rds')[['tmp3d5']],'\n')
  cat(readRDS('data/log.rds')[['tmp3d6']],'\n')
  cat(readRDS('data/log.rds')[['tmp3e5']],'\n')
  cat(readRDS('data/log.rds')[['tmp3f5']],'\n')
  cat(readRDS('data/log.rds')[['tmp3f6']])
  tmp3d5=readRDS('data/tmp3d5.rds')
  tmp3d6=readRDS('data/tmp3d6.rds')
  tmp3e5=
    list.files('data/tmp3e5/',pattern='tmp3e5_',full.names=T) %>%
    .[order(as.numeric(
      str_remove_all(.,'data/tmp3e5/tmp3e5_|rank|\\.rds')
    ))] %>%
    lapply(readRDS)
  tmp3f5=readRDS('data/tmp3f5.rds')
  tmp3f6=readRDS('data/tmp3f6.rds')
}
```

```{r Get features to train devnorep3DEG and evaluate it, include=FALSE}
# Get the possible combinations of devnorep3DEG
tmp3c3=
  GSE108497_maternal_blood$result %>%
  filter(adj.P.Val<0.05) %>%
  rownames() %>%
  setdiff(
    GSE149437_maternal_blood$result %>%
      filter(adj.P.Val<0.05) %>%
      rownames()
  ) %>%
  lapply(X=seq(1),Y=.,function(X,Y){
    if(X==1){
      as.list(Y)
    }else{
      Z=rep(list(Y),X) %>%
        expand.grid() %>%
        pblapply(X=seq(nrow(.)),Y=.,Z=X,function(X,Y,Z){
          Y=Y %>%
            mutate_all(as.character)
          K=c()
          for(i in seq(Z)){
            K[i]=Y[[i]][X]
          }
          K %>%
            sort() %>%
            data.frame(value=.) %>%
            mutate(key=paste0('Var',seq(Z))) %>%
            spread(key,value)
        }) %>%
        do.call(rbind,.) %>%
        filter(!duplicated(.)) %>%
        mutate(seq=seq(nrow(.))) %>%
        gather(key,value,-seq) %>%
        arrange(seq,key) %>%
        group_by(seq) %>%
        filter(!duplicated(value)) %>%
        filter(n()==X) %>%
        lapply(X=.$seq %>% .[!duplicated(.)],Y=.,function(X,Y){
          Y %>%
            filter(seq==X) %>%
            pull(value)
        })
    }
  }) %>%
  lapply(X=1,Y=.,function(X,Y){
    for(i in seq(length(Y))){
      if(i==1){
        Z=Y[[i]]
      }else{
        Z=Z %>%
          c(Y[[i]])
      }
    }
    Z
  }) %>%
  .[[1]]

if(run_heavy_computation){
  # Preprocess the features of devnorep3DEG in the development dataset
  tmp3d7=
    GSE108497_phenotype %>%
    select(outcome) %>%
    rownames_to_column(var='id') %>%
    left_join(
      GSE108497 %>%
        exprs() %>%
        mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
        # mutate_all(function(x)ifelse(x>=0,1,0)) %>%
        rownames_to_column(var='id')
      ,by='id'
    ) %>%
    column_to_rownames(var='id') %>%
    pblapply(X=seq(length(tmp3c3)),Y=tmp3c3,Z=.,function(X,Y,Z){
      Z %>%
        select_at(c(
          'outcome'
          ,paste0(
            'maternal_blood:'
            ,Y[[X]]
          )
        ))
    })
  
  saveRDS(tmp3d7,'data/tmp3d7.rds')
  
  # Preprocess the features of devnorep3DEG in the replication dataset 3
  tmp3d8=
    GSE149437_phenotype %>%
    select(outcome) %>%
    rownames_to_column(var='id') %>%
    left_join(
      GSE149437 %>%
        exprs() %>%
        mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
        # mutate_all(function(x)ifelse(x>=0,1,0)) %>%
        rownames_to_column(var='id')
      ,by='id'
    ) %>%
    column_to_rownames(var='id') %>%
    pblapply(X=seq(length(tmp3c3)),Y=tmp3c3,Z=.,function(X,Y,Z){
      Z %>%
        select_at(c(
          'outcome'
          ,paste0(
            'maternal_blood:'
            ,Y[[X]]
          )
        ))
    })
  
  saveRDS(tmp3d8,'data/tmp3d8.rds')
  
  # Create multiple CPU clusters for parallel computing.
  cl=detectCores()-2
  cl=makePSOCKcluster(cl)
  registerDoParallel(cl)
  
  # Close the CPU clusters and clean up memory at exit.
  on.exit(stopCluster(cl))
  on.exit(registerDoSEQ())
  on.exit(rm(cl))
  on.exit(rm(the_method))
  on.exit(gc())
  
  # Bring these packages to the CPU clusters.
  clusterEvalQ(cl,{
    library('parallel')
    library('doParallel')
    library('tidyverse')
    library('pbapply')
    library('caret')
    library('rpart')
  })
  
  # Train the devnorep3DEG tree model
  tmp3e7=
    tmp3d7 %>%
    pblapply(X=.
             ,training_parameters=training_parameters
             ,sample.kind=sample.kind
             ,std_no_outlier=std_no_outlier
             ,cl=cl
             ,function(X,training_parameters,sample.kind,std_no_outlier){
      suppressWarnings(set.seed(33,sample.kind=sample.kind))
      suppressWarnings(
        X %>%
          std_no_outlier(X) %>%
          `colnames<-`(
            str_remove_all(colnames(.),'maternal_blood:')
          ) %>%
          caret::train(
            outcome~.
            ,data=.
            ,method='rpart'
            ,weights=
              X %>%
              `colnames<-`(
                str_remove_all(colnames(.),'maternal_blood:')
              ) %>%
              pull(outcome) %>%
              table() %>%
              as.data.frame() %>%
              setNames(c('outcome','Freq')) %>%
              mutate(total=sum(Freq)) %>%
              mutate(weight=1/(Freq/total)*0.5) %>%
              right_join(
                X %>%
                  rownames_to_column(var='id')
                ,by='outcome'
              ) %>%
              pull(weight) %>%
              setNames(rownames(X))
            ,metric='ROC'
            ,trControl=training_parameters$final_trControl
            ,tuneLength=10 
            ,control=
              rpart.control(
                maxdepth=
                  sum(str_detect(colnames(X),'maternal_blood'))
              )
            ,parms=list(split='gini')
          )
      )
    })
  
  tmp3e7 %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      saveRDS(Y[[X]],paste0('data/tmp3e7/tmp3e7_',X,'.rds'))
    })
  
  # Evaluate the devnorep3DEG tree model using development dataset
  tmp3f7=
    tmp3e7 %>%
    pblapply(X=seq(length(.)),Y=.,function(X,Y){
      tmp3d7[[X]] %>%
        std_no_outlier(tmp3d7[[X]]) %>%
        `colnames<-`(
          str_remove_all(colnames(.),'maternal_blood:')
        ) %>%
        re_evalm(Y[[X]],th_selection,th_values) %>%
        column_to_rownames(var='metric')
    })
  
  saveRDS(tmp3f7,'data/tmp3f7.rds')
  
  # Evaluate the devnorep3DEG tree model using replication dataset 3
  tmp3f8=
    tmp3e7 %>%
    pblapply(X=seq(length(.)),Y=.,function(X,Y){
      tmp3d8[[X]] %>%
        std_no_outlier(tmp3d7[[X]]) %>%
        `colnames<-`(
          str_remove_all(colnames(.),'maternal_blood:')
        ) %>%
        re_evalm(Y[[X]],th_selection,th_values) %>%
        column_to_rownames(var='metric')
    })
  
  saveRDS(tmp3f8,'data/tmp3f8.rds')
}else{
  cat(readRDS('data/log.rds')[['tmp3d7']],'\n')
  cat(readRDS('data/log.rds')[['tmp3d8']],'\n')
  cat(readRDS('data/log.rds')[['tmp3e7']],'\n')
  cat(readRDS('data/log.rds')[['tmp3f7']],'\n')
  cat(readRDS('data/log.rds')[['tmp3f8']])
  tmp3d7=readRDS('data/tmp3d7.rds')
  tmp3d8=readRDS('data/tmp3d8.rds')
  tmp3e7=
    list.files('data/tmp3e7/',pattern='tmp3e7_',full.names=T) %>%
    .[order(as.numeric(
      str_remove_all(.,'data/tmp3e7/tmp3e7_|rank|\\.rds')
    ))] %>%
    lapply(readRDS)
  tmp3f7=readRDS('data/tmp3f7.rds')
  tmp3f8=readRDS('data/tmp3f8.rds')
}
```

```{r Get features to train recentDEG and evaluate it, include=FALSE}
# Get the possible combinations of recentDEG
tmp3c4=
  c('CAMK2G'
    ,'DERA'
    ,'FAM46A'
    ,'KIAA1109'
    ,'LRRC58'
    ,'MYLIP'
    ,'NDUFV3'
    ,'NMRK1'
    ,'PI4KA'
    ,'PRTFDC1'
    ,'PYGO2'
    ,'RNF149'
    ,'TFIP11'
    ,'TRIM21'
    ,'USB1'
    ,'YWHAQP5'
    ,'Y_RNA'
  ) %>%
  # "FAM46A"  "PI4KA"   "PRTFDC1" "USB1"    "YWHAQP5" "Y_RNA"
  intersect(rownames(GSE108497_maternal_blood$result)) %>%
  # "MYLIP"
  intersect(rownames(GSE177477)) %>%
  lapply(X=seq(2),Y=.,function(X,Y){
    if(X==1){
      as.list(Y)
    }else{
      rep(list(Y),X) %>%
        expand.grid() %>%
        lapply(X=seq(nrow(.)),Y=.,Z=X,function(X,Y,Z){
          Y=Y %>%
            mutate_all(as.character)
          K=c()
          for(i in seq(Z)){
            K[i]=Y[[i]][X]
          }
          K %>%
            sort() %>%
            data.frame(value=.) %>%
            mutate(key=paste0('Var',seq(Z))) %>%
            spread(key,value)
        }) %>%
        do.call(rbind,.) %>%
        filter(!duplicated(.)) %>%
        mutate(seq=seq(nrow(.))) %>%
        gather(key,value,-seq) %>%
        arrange(seq,key) %>%
        group_by(seq) %>%
        filter(!duplicated(value)) %>%
        filter(n()==X) %>%
        lapply(X=.$seq %>% .[!duplicated(.)],Y=.,function(X,Y){
          Y %>%
            filter(seq==X) %>%
            pull(value)
        })
    }
  }) %>%
  lapply(X=1,Y=.,function(X,Y){
    for(i in seq(length(Y))){
      if(i==1){
        Z=Y[[i]]
      }else{
        Z=Z %>%
          c(Y[[i]])
      }
    }
    Z
  }) %>%
  .[[1]]

if(run_heavy_computation){
  # Preprocess the features of recentDEG in the development dataset
  tmp3d9=
    GSE108497_phenotype %>%
    select(outcome) %>%
    rownames_to_column(var='id') %>%
    left_join(
      GSE108497 %>%
        exprs() %>%
        mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
        # mutate_all(function(x)ifelse(x>=0,1,0)) %>%
        rownames_to_column(var='id')
      ,by='id'
    ) %>%
    column_to_rownames(var='id') %>%
    pblapply(X=seq(length(tmp3c4)),Y=tmp3c4,Z=.,function(X,Y,Z){
      Z %>%
        select_at(c(
          'outcome'
          ,paste0(
            'maternal_blood:'
            ,Y[[X]]
          )
        ))
    })
  
  saveRDS(tmp3d9,'data/tmp3d9.rds')
  
  # Preprocess the features of recentDEG in the replication dataset 3
  tmp3d10=
    GSE149437_phenotype %>%
    select(outcome) %>%
    rownames_to_column(var='id') %>%
    left_join(
      GSE149437 %>%
        exprs() %>%
        mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
        # mutate_all(function(x)ifelse(x>=0,1,0)) %>%
        rownames_to_column(var='id')
      ,by='id'
    ) %>%
    column_to_rownames(var='id') %>%
    pblapply(X=seq(length(tmp3c4)),Y=tmp3c4,Z=.,function(X,Y,Z){
      Z %>%
        select_at(c(
          'outcome'
          ,paste0(
            'maternal_blood:'
            ,Y[[X]]
          )
        ))
    })
  
  saveRDS(tmp3d10,'data/tmp3d10.rds')
  
  # Create multiple CPU clusters for parallel computing.
  cl=detectCores()-2
  cl=makePSOCKcluster(cl)
  registerDoParallel(cl)
  
  # Close the CPU clusters and clean up memory at exit.
  on.exit(stopCluster(cl))
  on.exit(registerDoSEQ())
  on.exit(rm(cl))
  on.exit(rm(the_method))
  on.exit(gc())
  
  # Bring these packages to the CPU clusters.
  clusterEvalQ(cl,{
    library('parallel')
    library('doParallel')
    library('tidyverse')
    library('pbapply')
    library('caret')
    library('rpart')
  })
  
  # Train the recentDEG tree model
  tmp3e9=
    tmp3d9 %>%
    pblapply(X=.
             ,training_parameters=training_parameters
             ,sample.kind=sample.kind
             ,std_no_outlier=std_no_outlier
             ,cl=cl
             ,function(X,training_parameters,sample.kind,std_no_outlier){
      suppressWarnings(set.seed(33,sample.kind=sample.kind))
      suppressWarnings(
        X %>%
          std_no_outlier(X) %>%
          `colnames<-`(
            str_remove_all(colnames(.),'maternal_blood:')
          ) %>%
          caret::train(
            outcome~.
            ,data=.
            ,method='rpart'
            ,weights=
              X %>%
              `colnames<-`(
                str_remove_all(colnames(.),'maternal_blood:')
              ) %>%
              pull(outcome) %>%
              table() %>%
              as.data.frame() %>%
              setNames(c('outcome','Freq')) %>%
              mutate(total=sum(Freq)) %>%
              mutate(weight=1/(Freq/total)*0.5) %>%
              right_join(
                X %>%
                  rownames_to_column(var='id')
                ,by='outcome'
              ) %>%
              pull(weight) %>%
              setNames(rownames(X))
            ,metric='ROC'
            ,trControl=training_parameters$final_trControl
            ,tuneLength=10 
            ,control=
              rpart.control(
                maxdepth=
                  sum(str_detect(colnames(X),'maternal_blood'))
              )
            ,parms=list(split='gini')
          )
      )
    })
  
  tmp3e9 %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      saveRDS(Y[[X]],paste0('data/tmp3e9/tmp3e9_',X,'.rds'))
    })
  
  # Evaluate the recentDEG tree model using development dataset
  tmp3f9=
    tmp3e9 %>%
    pblapply(X=seq(length(.)),Y=.,function(X,Y){
      tmp3d9[[X]] %>%
        std_no_outlier(tmp3d9[[X]]) %>%
        `colnames<-`(
          str_remove_all(colnames(.),'maternal_blood:')
        ) %>%
        re_evalm(Y[[X]],th_selection,th_values) %>%
        column_to_rownames(var='metric')
    })
  
  saveRDS(tmp3f9,'data/tmp3f9.rds')
  
  # Evaluate the recentDEG tree model using replication dataset 3
  tmp3f10=
    tmp3e9 %>%
    pblapply(X=seq(length(.)),Y=.,function(X,Y){
      tmp3d10[[X]] %>%
        std_no_outlier(tmp3d9[[X]]) %>%
        `colnames<-`(
          str_remove_all(colnames(.),'maternal_blood:')
        ) %>%
        re_evalm(Y[[X]],th_selection,th_values) %>%
        column_to_rownames(var='metric')
    })
  
  saveRDS(tmp3f10,'data/tmp3f10.rds')
}else{
  cat(readRDS('data/log.rds')[['tmp3d9']],'\n')
  cat(readRDS('data/log.rds')[['tmp3d10']],'\n')
  cat(readRDS('data/log.rds')[['tmp3e9']],'\n')
  cat(readRDS('data/log.rds')[['tmp3f9']],'\n')
  cat(readRDS('data/log.rds')[['tmp3f10']])
  tmp3d9=readRDS('data/tmp3d9.rds')
  tmp3d10=readRDS('data/tmp3d10.rds')
  tmp3e9=
    list.files('data/tmp3e9/',pattern='tmp3e9_',full.names=T) %>%
    .[order(as.numeric(
      str_remove_all(.,'data/tmp3e9/tmp3e9_|rank|\\.rds')
    ))] %>%
    lapply(readRDS)
  tmp3f9=readRDS('data/tmp3f9.rds')
  tmp3f10=readRDS('data/tmp3f10.rds')
}
```

```{r Create a function to compute AUROCs given a list, include=FALSE}
get_auroc=function(data,train_mod){
  data %>%
    pblapply(X=names(.),Y=.,Z=train_mod,function(X,Y,Z){
      Y[[X]] %>%
        lapply(X=names(.),Y=.,Z=Z[[X]],function(X,Y,Z){
          Y[[X]]=
            Y[[X]] %>%
            `colnames<-`(str_remove_all(colnames(.),'maternal_blood:'))
          Z[[X]] %>%
            predict(newdata=Y[[X]],type='prob') %>%
            cbind(select(Y[[X]],outcome)) %>%
            select(nonevent,event,outcome) %>%
            evalm(silent=T,showplots=F) %>%
            .$optres %>%
            .$Group1 %>%
            .['AUC-ROC',] %>%
            `rownames<-`(NULL) %>%
            mutate(grank=X)
        }) %>%
        do.call(rbind,.) %>%
        mutate(rank=X)
    }) %>%
    do.call(rbind,.) %>%
    unite(id,rank,grank,sep='.') %>%
    separate(CI,c('LB','UB'),sep='-') %>%
    mutate_at(c('LB','UB'),as.numeric) %>%
    left_join(
      train_mod %>%
        pblapply(X=names(.),Y=.,function(X,Y){
          Y[[X]] %>%
            lapply(X=names(.),Y=.,Z=X,function(X,Y,Z){
              K=Y[[X]] %>%
                .$finalModel %>%
                rpart.rules() %>%
                as.data.frame() %>%
                setNames(paste0('col',seq(ncol(.)))) %>%
                gather() %>%
                filter(
                  value!=''
                  & str_detect(value,'[:alpha:]')
                  & !value%in%c('when','is','to','null model')
                )
              if(nrow(K)==0){
                data.frame(
                  predictor=''
                  ,cand_predictor=
                    Y[[X]] %>%
                    .$trainingData %>%
                    colnames() %>%
                    .[.!='.outcome'] %>%
                    sort() %>%
                    paste0(collapse=',')
                  ,rank=Z
                  ,grank=X
                )
              }else{
                K %>%
                  pull(value) %>%
                  .[!duplicated(.)] %>%
                  setNames(NULL) %>%
                  data.frame(
                    predictor=.
                    ,cand_predictor=
                      Y[[X]] %>%
                      .$trainingData %>%
                      colnames() %>%
                      .[.!='.outcome'] %>%
                      sort() %>%
                      paste0(collapse=',')
                    ,rank=Z
                    ,grank=X
                  )
              }
            }) %>%
            do.call(rbind,.)
        }) %>%
        do.call(rbind,.) %>%
        group_by(rank,grank,cand_predictor) %>%
        summarize(predictors=paste0(sort(predictor),collapse=',')) %>%
        unite(id,rank,grank,sep='.')
      ,by='id'
    )
}
```

```{r Compute AUROCs for all the methods of predictor discovery, include=FALSE}
if(run_heavy_computation){
  tmp3g=list()
  
  tmp3g$tmp3e$train=
    tmp3d %>%
    get_auroc(tmp3e)
  
  tmp3g$tmp3e$test=
    tmp3d2 %>%
    get_auroc(tmp3e)
  
  tmp3g$tmp3e3$train=
    list(ext=list(reme=tmp3d3)) %>%
    get_auroc(list(ext=list(reme=tmp3e3)))
  
  tmp3g$tmp3e3$test=
    list(ext=list(reme=tmp3d4)) %>%
    get_auroc(list(ext=list(reme=tmp3e3)))
  
  tmp3g$tmp3e5$train=
    tmp3d5 %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      list(both=Y[[X]])
    }) %>%
    setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0'))) %>%
    get_auroc(
      tmp3e5 %>%
        lapply(X=seq(length(.)),Y=.,function(X,Y){
          list(both=Y[[X]])
        }) %>%
        setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0')))
    )
  
  tmp3g$tmp3e5$test=
    tmp3d6 %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      list(both=Y[[X]])
    }) %>%
    setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0'))) %>%
    get_auroc(
      tmp3e5 %>%
        lapply(X=seq(length(.)),Y=.,function(X,Y){
          list(both=Y[[X]])
        }) %>%
        setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0')))
    )
  
  tmp3g$tmp3e7$train=
    tmp3d7 %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      list(not_in=Y[[X]])
    }) %>%
    setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0'))) %>%
    get_auroc(
      tmp3e7 %>%
        lapply(X=seq(length(.)),Y=.,function(X,Y){
          list(not_in=Y[[X]])
        }) %>%
        setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0')))
    )
  
  tmp3g$tmp3e7$test=
    tmp3d8 %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      list(not_in=Y[[X]])
    }) %>%
    setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0'))) %>%
    get_auroc(
      tmp3e7 %>%
        lapply(X=seq(length(.)),Y=.,function(X,Y){
          list(not_in=Y[[X]])
        }) %>%
        setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0')))
    )
  
  tmp3g$tmp3e9$train=
    tmp3d9 %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      list(recent=Y[[X]])
    }) %>%
    setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0'))) %>%
    get_auroc(
      tmp3e9 %>%
        lapply(X=seq(length(.)),Y=.,function(X,Y){
          list(recent=Y[[X]])
        }) %>%
        setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0')))
    )
  
  tmp3g$tmp3e9$test=
    tmp3d10 %>%
    lapply(X=seq(length(.)),Y=.,function(X,Y){
      list(recent=Y[[X]])
    }) %>%
    setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0'))) %>%
    get_auroc(
      tmp3e9 %>%
        lapply(X=seq(length(.)),Y=.,function(X,Y){
          list(recent=Y[[X]])
        }) %>%
        setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0')))
    )
  
  saveRDS(tmp3g,'data/tmp3g.rds')
}else{
  cat(readRDS('data/log.rds')[['tmp3g']])
  tmp3g=readRDS('data/tmp3g.rds')
}
```

```{r Summarize the AUROC comparison for the predictor discovery, include=FALSE}
if(run_heavy_computation){
  tmp3h=
    tmp3g %>%
    lapply(X=names(.),Y=.,function(X,Y){
      Y[[X]] %>%
        lapply(X=names(.),Y=.,function(X,Y){
          Y[[X]] %>%
            mutate(set=X)
        }) %>%
        do.call(rbind,.) %>%
        mutate(mod_list=X)
    }) %>%
    do.call(rbind,.) %>%
    pivot_wider(names_from=c('set'),values_from=c('Score','LB','UB')) %>%
    lapply(X=1,Y=.,function(X,Y){
      
      Z=GSE177477_phenotype %>%
        select(outcome) %>%
        rownames_to_column(var='id') %>%
        left_join(
          GSE177477 %>%
            exprs() %>%
            mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
            rownames_to_column(var='id')
          ,by='id'
        ) %>%
        left_join(
          GSE177477 %>%
            phenoData() %>%
            pData() %>%
            rownames_to_column(var='gsm') %>%
            select(c('gsm',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
            setNames(
              colnames(.) %>%
                str_remove_all(':ch1') %>%
                str_replace_all('\\s','_')
            ) %>%
            select(-id) %>%
            rename(id=gsm) %>%
            select(id,symptons,severity)
          ,by='id'
        ) %>%
        mutate(
          outcome=
            factor(ifelse(
              !symptons%in%c('Healthy controls')
              ,'event','nonevent'
            ),levels(outcome))
        ) %>%
        select(-symptons,-severity) %>%
        `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
        column_to_rownames(var='id')
      
      K=list(
          tmp3e=
            tmp3e
          ,tmp3e3=
            list(ext=list(reme=tmp3e3))
          ,tmp3e5=
            tmp3e5 %>%
            lapply(X=seq(length(.)),Y=.,function(X,Y){
              list(both=Y[[X]])
            }) %>%
            setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0')))
          ,tmp3e7=
            tmp3e7 %>%
            lapply(X=seq(length(.)),Y=.,function(X,Y){
              list(not_in=Y[[X]])
            }) %>%
            setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0')))
          ,tmp3e9=
            tmp3e9 %>%
            lapply(X=seq(length(.)),Y=.,function(X,Y){
              list(recent=Y[[X]])
            }) %>%
            setNames(paste0('mod',str_pad(seq(length(.)),3,'left','0')))
        )
      
      Y=Y %>%
        mutate(
          biom_avail=
            cand_predictor %>%
            sapply(function(x){
              y=x %>%
                str_split(',') %>%
                .[[1]]
              all(y%in%colnames(Z))
            })
        )
      
      pblapply(X=seq(nrow(Y)),Y=Y,Z=Z,K=K,function(X,Y,Z,K){
        L=Y[X,,drop=F] %>%
          separate(id,c('level1','level2'),sep='\\.')
        if(L$biom_avail){
          M=K %>%
            .[[L$mod_list]] %>%
            .[[L$level1]] %>%
            .[[L$level2]] %>%
            predict(newdata=Z,type='prob') %>%
            cbind(select(Z,outcome)) %>%
            select(nonevent,event,outcome) %>%
            evalm(silent=T,showplots=F) %>%
            .$optres %>%
            .$Group1 %>%
            .['AUC-ROC',] %>%
            `rownames<-`(NULL) %>%
            separate(CI,c('LB','UB'),sep='-') %>%
            mutate_at(c('LB','UB'),as.numeric) %>%
            rename_all(function(x)paste0(x,'_covid'))
        }else{
          M=data.frame(
            Score_covid=NA
            ,LB_covid=NA
            ,UB_covid=NA
          )
        }
        cbind(L,M)
      }) %>%
      do.call(rbind,.)
      
    }) %>%
    .[[1]]
  
  saveRDS(tmp3h,'data/tmp3h.rds')
}else{
  cat(readRDS('data/log.rds')[['tmp3h']])
  tmp3h=readRDS('data/tmp3h.rds')
}
```

```{r Show outcomes of the COVID-19 dataset, eval=FALSE, include=FALSE}
GSE177477_phenotype %>%
  select(outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    GSE177477 %>%
      exprs() %>%
      mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
      # mutate_all(function(x)ifelse(x>=0,1,0)) %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  left_join(
    GSE177477 %>%
      phenoData() %>%
      pData() %>%
      rownames_to_column(var='gsm') %>%
      select(c('gsm',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
      setNames(
        colnames(.) %>%
          str_remove_all(':ch1') %>%
          str_replace_all('\\s','_')
      ) %>%
      select(-id) %>%
      rename(id=gsm) %>%
      select(id,symptons,severity)
    ,by='id'
  ) %>%
  mutate(
    outcome=
      factor(ifelse(
        !symptons%in%c('Healthy controls')
        ,'event','nonevent'
      ),levels(outcome))
  ) %>%
  group_by(symptons,severity,outcome) %>%
  summarize(n=n())
```

```{r Assess the eligibility of the emulated biomarkers, include=FALSE}
tmp3i=
  tmp3h %>%
  mutate(
    pred_avail=predictors!=''
    ,replicable=Score_test>=LB_train & Score_test<=UB_train
    ,covidable=
      case_when(
        Score_covid>=UB_train ~ 'upper'
        ,Score_covid>=LB_train & Score_covid<=UB_train ~ 'on_train'
        ,Score_covid<=LB_train ~ 'lower'
        ,TRUE ~ ''
      )
    ,covid_signif=LB_covid>=0.5
    ,quality=
      ifelse(
        biom_avail
        & pred_avail
        & replicable
        & covidable=='lower'
        & !is.na(covid_signif)
        & !covid_signif
        ,'high','low'
      ) %>%
      factor(c('low','high'))
  )
```

```{r Conduct permutation tests to predictor discovery methods, include=FALSE}
tmp3i %>%
  mutate(
    mod_list=
      ifelse(
        mod_list%in%c('tmp3e3','tmp3e7')
        ,mod_list#'tmp3e37'
        ,mod_list
      )
  ) %>%
  lapply(X=1,Y=.,function(X,Y){
    Z=Y %>%
      group_by(mod_list,quality) %>%
      summarize(n=n()) %>%
      ungroup() %>%
      group_by(mod_list) %>%
      mutate(
        total=sum(n)
        ,p=n/total
      ) %>%
      ungroup()
    
    suppressWarnings(set.seed(33,sample.kind=sample.kind))
    K=seq(500) %>%
      lapply(function(x){
        Y %>%
          mutate(
            quality=
              quality %>%
              sample(length(.),F)
          ) %>%
          group_by(mod_list,quality) %>%
          summarize(n=n()) %>%
          ungroup() %>%
          group_by(mod_list) %>%
          mutate(
            total=sum(n)
            ,p_rand=n/total
          ) %>%
          ungroup() %>%
          mutate(b=x)
      }) %>%
      do.call(rbind,.) %>%
      select(b,mod_list,quality,p_rand)
    
    L=K %>%
      spread(quality,p_rand,fill=0) %>%
      mutate(
        d_rand=high-low
      ) %>%
      select(-low,-high) %>%
      left_join(
        Z %>%
          select(mod_list,quality,p)%>%
          spread(quality,p,fill=0) %>%
          mutate(
            d=high-low
          ) %>%
          select(-low,-high)
        ,by=c('mod_list')
      ) %>%
      group_by(mod_list) %>%
      summarize(p.value=mean(d<d_rand))
    
    Z %>%
      pivot_wider(
        names_from=c('quality')
        ,values_from=c('n','p')
        ,values_fill=0
      ) %>%
      left_join(L,by='mod_list')
  }) %>%
  .[[1]] %>%
  mutate(
    mod_list=
      case_when(
        mod_list=='tmp3e'~'blood-derived surrogate transcriptome'
        ,mod_list=='tmp3e3'~'blood transcriptome, p.val<.05 & abs(logFC)>2'
        ,mod_list=='tmp3e5'~'blood transcriptome, p.val<.05 both train & test'
        ,mod_list=='tmp3e7'~'blood transcriptome, p.val<.05 train but not test'
        ,mod_list=='tmp3e9'~'blood transcriptome, a recent study'
        ,mod_list=='tmp3e37'~'blood transcriptome, p.val<.05 train but not test'
        ,TRUE~''
      )
  )
```

```{r Get the best tree from the best method of discovery, include=FALSE}
tmp3j=
  tmp3i %>%
  filter(quality=='high' & mod_list=='tmp3e') %>%
  arrange(desc(LB_train),desc(Score_test),LB_covid,level1,level2) %>%
  slice(1) %>%
  unite(id,level1,level2,sep='.') %>%
  pull(id) %>%
  lapply(function(x){
    y=str_split(x,'\\.')[[1]]
    list(
      mod=tmp3e[[y[1]]][[y[2]]]
      ,eval=tmp3f[[y[1]]][[y[2]]]
      ,train=tmp3d[[y[1]]][[y[2]]]
      ,test=tmp3d2[[y[1]]][[y[2]]]
      ,covid=
        GSE177477_phenotype %>%
        select(outcome) %>%
        rownames_to_column(var='id') %>%
        left_join(
          GSE177477 %>%
            exprs() %>%
            mb_predictor(list(maternal_blood=GSE108497_maternal_blood)) %>%
            rownames_to_column(var='id')
          ,by='id'
        ) %>%
        left_join(
          GSE177477 %>%
            phenoData() %>%
            pData() %>%
            rownames_to_column(var='gsm') %>%
            select(c('gsm',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
            setNames(
              colnames(.) %>%
                str_remove_all(':ch1') %>%
                str_replace_all('\\s','_')
            ) %>%
            select(-id) %>%
            rename(id=gsm) %>%
            select(id,symptons,severity)
          ,by='id'
        ) %>%
        mutate(
          outcome=
            factor(ifelse(
              !symptons%in%c('Healthy controls')
              ,'event','nonevent'
            ),levels(outcome))
        ) %>%
        select(-symptons,-severity) %>%
        column_to_rownames(var='id') %>%
        select_at(colnames(tmp3d[[y[1]]][[y[2]]]))
    )
  }) %>%
  .[[1]]
```

```{r Candidate and selected predictors of biomarker trees, include=FALSE}
tmp3k=
  tmp3d %>%
  pblapply(function(x){
    x %>%
      lapply(function(x){
        x[rownames(x)%in%rownames(
          GSE108497 %>%
            phenoData() %>%
            pData() %>%
            select(c('title',colnames(.) %>% .[str_detect(.,':ch1')])) %>%
            setNames(colnames(.) %>% str_remove_all(':ch1')) %>%
            filter(
              apl==0
              & lac==0
              & fd==0
              & nnd==0
            ) %>%
            mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
            mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
            select(sle,early,pe,sga,time_point) %>%
            filter(
              (early==0 & pe==0 & sga==0)
              |
              (early==1 & pe==1 & sga==0)
              |
              (early==0 & pe==1 & sga==0)
              |
              (early==0 & pe==0 & sga==1)
            ) %>%
            filter(!(early==1 & pe==1 & sga==0))
        ),]
      })
  }) %>%
  get_auroc(tmp3e)
```

```{r Candidate and selected predictors of the best biomarker trees, echo=FALSE}
tmp3h %>%
  filter(mod_list=='tmp3e') %>%
  .[,
    colnames(.) %>% .[!.%in%paste0(c('Score','LB','UB'),'_test')]
  ] %>%
  left_join(
    tmp3k %>%
      separate(id,c('level1','level2'),sep='\\.') %>%
      rename_at(c('Score','LB','UB'),paste0,'_test')
    ,by=c('level1','level2','cand_predictor','predictors')
  ) %>%
  mutate(
    pred_avail=predictors!=''
    ,replicable=Score_test>=LB_train & Score_test<=UB_train
    ,covidable=
      case_when(
        Score_covid>=UB_train ~ 'upper'
        ,Score_covid>=LB_train & Score_covid<=UB_train ~ 'on_train'
        ,Score_covid<=LB_train ~ 'lower'
        ,TRUE ~ ''
      )
    ,covid_signif=LB_covid>=0.5
    ,quality=
      ifelse(
        biom_avail
        & pred_avail
        & replicable
        & covidable=='lower'
        & !is.na(covid_signif)
        & !covid_signif
        ,'high','low'
      ) %>%
      factor(c('low','high'))
  ) %>%
  filter(quality=='high')
```

```{r The best biomarkers, echo=FALSE, fig.height=3.46457, fig.width=3.46457}
tmp3e$rank2$grank3 %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )
```

```{r Show averages and SDs to scale the biomarkers, eval=FALSE, include=FALSE}
tmp3d$rank2$grank3 %>%
  select(-outcome) %>%
  lapply(X=colnames(.),Y=.,function(X,Y){
    median=quantile(Y[[X]],0.5)
    q1=quantile(Y[[X]],0.25)
    q3=quantile(Y[[X]],0.75)
    minq=q1-1.5*(q3-q1)
    maxq=q3+1.5*(q3-q1)
    outliers=Y[[X]]<minq | Y[[X]]>maxq
    avg=
      Y[[X]] %>%
      .[!outliers] %>%
      mean()
    
    std=
      Y[[X]] %>%
      .[!outliers] %>%
      sd()
    
    data.frame(predictor=X,avg=avg,std=std)
  }) %>%
  do.call(rbind,.)
```

```{r Create a function to evaluate leaves of a biomarker tree, include=FALSE}
tree_eval=function(data,train_mod,train_eval){
  suppressWarnings(set.seed(33,sample.kind=sample.kind))
  train_mod %>%
    predict(newdata=data,type='prob') %>%
    rownames_to_column(var='id') %>%
    left_join(
      train_mod$finalModel$where %>%
        data.frame(node=.) %>%
        rownames_to_column(var='id') %>%
        left_join(
          data.frame(
            node=unique(.$node)
            ,leaf=LETTERS[seq(length(unique(.$node)))]
          )
          ,by='node'
        ) %>%
        right_join(
          train_mod %>%
            predict(type='prob') %>%
            rownames_to_column(var='id')
          ,by='id'
        ) %>%
        select(event,nonevent,node,leaf) %>%
        filter(!duplicated(.))
      ,by=c('event','nonevent')
    ) %>%
    mutate(th=train_eval$th[1]) %>%
    mutate(pred=ifelse(event>=th,'event','nonevent')) %>%
    cbind(select(data,outcome)) %>%
    lapply(X=seq(30),Y=.,function(X,Y){
      Y %>%
        .[sample(seq(nrow(.)),nrow(.),T),] %>%
        group_by(th,leaf,pred,outcome) %>%
        summarize(n=n()) %>%
        ungroup() %>%
        group_by(th,leaf,pred) %>%
        summarize(
          true_pred=sum(ifelse(outcome==pred,n,NA),na.rm=T)
          ,false_pred=sum(ifelse(outcome==pred,NA,n),na.rm=T)
          ,leaf_total=sum(n)
          ,leaf_ppv_npv=true_pred/leaf_total
        ) %>%
        ungroup() %>%
        group_by(th,pred) %>%
        mutate(
          pred_total=sum(leaf_total)
          ,leaf_p=leaf_total/pred_total
          ,pred_ppv_npv=sum(true_pred)/pred_total
        ) %>%
        ungroup() %>%
        mutate(
          total=sum(leaf_total)
          ,pred_p=sum(ifelse(pred=='event',leaf_total,NA),na.rm=T)/total
          ,outcome_p=sum(ifelse(pred=='event',true_pred,false_pred))/total
        ) %>%
        gather(metric,value,-th,-leaf,-pred) %>%
        arrange(th,leaf,pred) %>%
        mutate(b=X) %>%
        select(b,everything())
    }) %>%
    do.call(rbind,.) %>%
    group_by(th,leaf,pred,metric) %>%
    summarize(
      Score=mean(value)
      ,LB=mean(value)-qnorm(0.975)*sd(value)/sqrt(n())
      ,UB=mean(value)+qnorm(0.975)*sd(value)/sqrt(n())
    ) %>%
    mutate(
      metric=
        metric %>%
        factor(c(
          'true_pred'
          ,'false_pred'
          ,'leaf_total'
          ,'leaf_ppv_npv'
          ,'pred_total'
          ,'leaf_p'
          ,'pred_ppv_npv'
          ,'total'
          ,'pred_p'
          ,'outcome_p'
        ))
    ) %>%
    arrange(th,leaf,pred,metric)
}
```

```{r The best tree (training), echo=FALSE, fig.height=3.5, fig.width=3.5}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$mod %>%
  .$trainingData %>%
  rename(outcome=.outcome) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r The best tree (COVID-19), echo=FALSE, fig.height=3.5, fig.width=3.5}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$covid %>%
  `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r The surrogate transcriptome in the best model (within ranks), echo=FALSE}
tmp3c %>%
  filter(
    blood_t
    %in%str_split(
      tmp3i %>%
        filter(
          cand_predictor
          %in%paste0(
            tmp3j$mod$trainingData %>%
              colnames() %>%
              .[.!='.outcome'] %>%
              sort()
            ,collapse=','
          )
        ) %>%
        slice(1) %>%
        pull(predictors)
      ,','
    )[[1]]
    & rank==as.numeric(
      tmp3i %>%
        filter(
          cand_predictor
          %in%paste0(
            tmp3j$mod$trainingData %>%
              colnames() %>%
              .[.!='.outcome'] %>%
              sort()
            ,collapse=','
          )
        ) %>%
        slice(1) %>%
        pull(level1) %>%
        str_remove_all('rank')
    )
    & grank==as.numeric(
      tmp3i %>%
        filter(
          cand_predictor
          %in%paste0(
            tmp3j$mod$trainingData %>%
              colnames() %>%
              .[.!='.outcome'] %>%
              sort()
            ,collapse=','
          )
        ) %>%
        slice(1) %>%
        pull(level2) %>%
        str_remove_all('grank')
    )
  ) %>%
  rbind(
    left_join(.,tmp3c,by='gene') %>%
      filter(
        tissue.x!=tissue.y
        & blood_t.x==blood_t.y
      ) %>%
      select(
        -tissue.x,-blood_t.x,-avg_accu_w.x
        ,-rank.x,-n_tissue.x,-global_rank.x,-grank.x
      ) %>%
      rename_all(str_remove_all,'\\.y') %>%
      select(tissue,everything())
  ) %>%
  select(-n_tissue,-grank) %>%
  select(gene,tissue,everything()) %>%
  filter(!duplicated(.)) %>%
  arrange(factor(gene,unique(gene)))
```

```{r The surrogate transcriptome in the best model (any ranks), include=FALSE}
tmp3 %>%
  left_join(tmp3b,c('tissue','blood_t')) %>%
  left_join(tmp3c,c('tissue','gene','blood_t','avg_accu_w','rank')) %>%
  select(blood_t,gene,tissue,avg_accu_w,rank,global_rank) %>%
  filter(!duplicated(.)) %>%
  lapply(X=c('FANCI'
             ,'SELENOV'
             ,'TSEN15'
             ,'TPX2'
             ,'WIPF3'
             ,'ARID2'
             ,'INSM1'
             ,'TMEM38B'
             ,'ALS2CL '
             )
         ,Y=.
         ,function(X,Y){
    Y %>%
      filter(gene==X & blood_t%in%c('ITGA5','P2RX7','IRF6'))
  }) %>%
  do.call(rbind,.)
```

```{r The best tree (training PE), echo=FALSE, fig.height=3, fig.width=3}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$mod %>%
  .$trainingData %>%
  rename(outcome=.outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    pdata$GSE108497$phenotype %>%
      filter(
        apl==0
        & lac==0
        & fd==0
        & nnd==0
      ) %>%
      mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
      mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
      select(sle,early,pe,sga,time_point) %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  filter(
    # (early==0 & pe==0 & sga==0)
    # |
    # (early==0 & pe==0 & sga==1)
    # |
    (early==1 & pe==1 & sga==0)
    |
    (early==0 & pe==1 & sga==0)
  ) %>%
  `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r The best tree (training EOPE), echo=FALSE, fig.height=3, fig.width=3}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$mod %>%
  .$trainingData %>%
  rename(outcome=.outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    pdata$GSE108497$phenotype %>%
      filter(
        apl==0
        & lac==0
        & fd==0
        & nnd==0
      ) %>%
      mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
      mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
      select(sle,early,pe,sga,time_point) %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  filter(
    # (early==0 & pe==0 & sga==0)
    # |
    # (early==0 & pe==0 & sga==1)
    # |
    (early==1 & pe==1 & sga==0)
  ) %>%
  `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r The best tree (training LOPE), echo=FALSE, fig.height=3, fig.width=3}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$mod %>%
  .$trainingData %>%
  rename(outcome=.outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    pdata$GSE108497$phenotype %>%
      filter(
        apl==0
        & lac==0
        & fd==0
        & nnd==0
      ) %>%
      mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
      mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
      select(sle,early,pe,sga,time_point) %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  filter(
    # (early==0 & pe==0 & sga==0)
    # |
    # (early==0 & pe==0 & sga==1)
    # |
    (early==0 & pe==1 & sga==0)
  ) %>%
  `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r The best tree (training COVID), echo=FALSE, fig.height=3, fig.width=3}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$covid %>%
  filter(outcome=='event') %>%
  `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r The best tree (training no PE), echo=FALSE, fig.height=3, fig.width=3}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$mod %>%
  .$trainingData %>%
  rename(outcome=.outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    pdata$GSE108497$phenotype %>%
      filter(
        apl==0
        & lac==0
        & fd==0
        & nnd==0
      ) %>%
      mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
      mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
      select(sle,early,pe,sga,time_point) %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  filter(
    (early==0 & pe==0 & sga==0)
    |
    (early==0 & pe==0 & sga==1)
    # |
    # (early==1 & pe==1 & sga==0)
    # |
    # (early==0 & pe==1 & sga==0)
  ) %>%
  `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r The best tree (training normal), echo=FALSE, fig.height=3, fig.width=3}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$mod %>%
  .$trainingData %>%
  rename(outcome=.outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    pdata$GSE108497$phenotype %>%
      filter(
        apl==0
        & lac==0
        & fd==0
        & nnd==0
      ) %>%
      mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
      mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
      select(sle,early,pe,sga,time_point) %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  filter(
    (early==0 & pe==0 & sga==0)
    # |
    # (early==1 & pe==1 & sga==0)
    # |
    # (early==0 & pe==1 & sga==0)
  ) %>%
  `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r The best tree (training FGR), echo=FALSE, fig.height=3, fig.width=3}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$mod %>%
  .$trainingData %>%
  rename(outcome=.outcome) %>%
  rownames_to_column(var='id') %>%
  left_join(
    pdata$GSE108497$phenotype %>%
      filter(
        apl==0
        & lac==0
        & fd==0
        & nnd==0
      ) %>%
      mutate(early=as.integer(as.numeric(ga_at_end_of_pregnancy)<34)) %>%
      mutate(time_point=ifelse(is.na(time_point),'non_pregnant',time_point)) %>%
      select(sle,early,pe,sga,time_point) %>%
      rownames_to_column(var='id')
    ,by='id'
  ) %>%
  filter(
    (early==0 & pe==0 & sga==1)
    # |
    # (early==1 & pe==1 & sga==0)
    # |
    # (early==0 & pe==1 & sga==0)
  ) %>%
  `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r The best tree (training no COVID), echo=FALSE, fig.height=3, fig.width=3}
tmp3j$mod %>%
  .$finalModel %>%
  rpart.plot(
    type=5
    ,cex=0.65
    ,col=1
    ,extra=0
    ,box.palette=c('white','white')
    ,node.fun=function(x,labs,digits,varlen){
      labs[seq(length(labs)) %in% unique(.$where)]=
        LETTERS[seq(length(unique(.$where)))]
      labs
    }
  )

tmp3j$covid %>%
  filter(outcome=='nonevent') %>%
  `colnames<-`(str_remove_all(colnames(.),'maternal_blood:')) %>%
  tree_eval(tmp3j$mod,tmp3j$eval)
```

```{r}
timestamper[13]=as.character(now())
```

```{r Load the DIANA-mited miRNA data for PE, include=FALSE}
tmp4=
  read_tsv('data/diana_mited_20220302.tsv') %>%
  select(
    -S.No
    ,-Project_ID
    ,-Collection
    ,-Tissue_subregion
    ,-Cell_Line
    ,-Organism
    ,-Gender
    ,-Tissue_definition
  ) %>%
  gather(
    miRNA
    ,log2RPM
    ,-Sample_ID
    ,-Tissue_or_organ_of_origin
    ,-Disease
    ,-Health_state
  ) %>%
  mutate(
    Health_state=
      ifelse(
        Health_state=='disease'
        ,'event'
        ,'nonevent'
      ) %>%
      factor(c('nonevent','event'))
  )
```

```{r Load the GeneCards miRNA ID targeting the predictor genes, include=FALSE}
tmp4b=
  read_csv('data/mirna.csv') %>%
  separate(mirna,c('miRNA','id'),sep=' \\(') %>%
  mutate(
    id=str_remove_all(id,'\\)')
    ,gene=factor(gene,unique(gene))
  ) %>%
  left_join(tmp4,by='miRNA')
```

```{r Summarize the number of miRNAs, eval=FALSE, include=FALSE}
tmp4b %>%
  select(gene,miRNA,Tissue_or_organ_of_origin,Disease,Health_state) %>%
  filter(!duplicated(.)) %>%
  group_by(gene,Tissue_or_organ_of_origin,Disease,Health_state) %>%
  summarize(miRNAs=n()) %>%
  spread(gene,miRNAs,fill=0)
```

```{r Filter PE as outcome of the miRNA of interests, include=FALSE}
tmp4c=
  tmp4b %>%
  filter(
    !(Health_state=='event'
      & Disease=='Preeclampsia, intrauterine growth restriction')
    | is.na(Health_state)
  ) %>%
  mutate(
    Disease=
      ifelse(
        Health_state=='event'
        ,Disease
        ,NA
      ) %>%
      .[!is.na(.)] %>%
      .[!duplicated(.)]
  )
```

```{r Conduct regression analyses of the miRNAs on PE, include=FALSE}
tmp4d=
  tmp4c %>%
  filter(!is.na(Sample_ID)) %>%
  group_by(gene,miRNA,Disease,Health_state) %>%
  summarize(n=n()) %>%
  ungroup() %>%
  spread(Health_state,n,fill=0) %>%
  mutate(
    total=nonevent+event
    ,p=event/total
  ) %>%
  filter(!(p==0|p==1)) %>%
  left_join(tmp4c,by=c('gene','miRNA','Disease')) %>%
  group_by(gene,miRNA,Disease) %>%
  do(broom::tidy(suppressWarnings(
    glm(Health_state~log2RPM,data=.,family=binomial(link='logit'))
  ))) %>%
  filter(term=='log2RPM') %>%
  mutate(
    LB=round(exp(estimate-std.error),2)
    ,UB=round(exp(estimate+std.error),2)
    ,OR=round(exp(estimate),2)
  ) %>%
  select(gene,miRNA,Disease,OR,LB,UB) %>%
  ungroup()
```

```{r Determine total effect of miRNAs on PE per predictor, include=FALSE}
tmp4e=
  tmp4c %>%
  mutate(
    availability=
      ifelse(
        miRNA
        %in%pull(
          tmp4c %>%
            filter(!is.na(Sample_ID)) %>%
            group_by(gene,miRNA,Disease,Health_state) %>%
            summarize(n=n()) %>%
            ungroup() %>%
            spread(Health_state,n,fill=0) %>%
            mutate(
              total=nonevent+event
              ,p=event/total
            ) %>%
            filter(!(p==0|p==1))
          ,miRNA
        )
        ,'available'
        ,'not_available'
      )
  ) %>%
  select(gene,miRNA,availability) %>%
  filter(!duplicated(.)) %>%
  group_by(gene,availability) %>%
  summarize(
    n=n()
    ,unavailable_miRNAs=
      paste0(
        ifelse(availability=='not_available',miRNA,NA) %>%
          .[!is.na(.)]
        ,collapse='|'
      )
  ) %>%
  ungroup() %>%
  group_by(gene) %>%
  mutate(
    unavailable_miRNAs=
      paste0(
        unavailable_miRNAs %>%
          .[.!='']
        ,collapse='|'
      )
  ) %>%
  spread(availability,n,fill=0) %>%
  mutate(
    total=available+not_available
    ,p=available/total
    ,availability=
      paste0(
        available
        ,'/'
        ,total
        ,' ('
        ,round(p*100,ifelse(total<10,0,ifelse(total<100,1,2)))
        ,'%)'
      )
  ) %>%
  select(gene,availability,unavailable_miRNAs) %>%
  left_join(
    tmp4d %>%
      mutate(
        significance=
          ifelse(!is.na(OR) & (LB>1 | UB<1),'significant','not_significant')
      ) %>%
      group_by(gene,significance) %>%
      summarize(n=n()) %>%
      spread(significance,n,fill=0) %>%
      mutate(
        total=significant+not_significant
        ,p=significant/total
        ,significance=
          paste0(
            significant
            ,'/'
            ,total
            ,' ('
            ,round(p*100,ifelse(total<10,0,ifelse(total<100,1,2)))
            ,'%)'
          )
      ) %>%
      select(gene,significance)
    ,by='gene'
  ) %>%
  left_join(
    tmp4d %>%
      filter(!is.na(OR) & (LB>1 | UB<1)) %>%
      mutate(
        regulation=ifelse(UB<1,'downregulated','upregulated')
      ) %>%
      group_by(gene,regulation) %>%
      summarize(n=n()) %>%
      spread(regulation,n,fill=0) %>%
      mutate(
        total=downregulated+upregulated
        ,p=downregulated/total
        ,downregulation=
          paste0(
            downregulated
            ,'/'
            ,total
            ,' ('
            ,round(p*100,ifelse(total<10,0,ifelse(total<100,1,2)))
            ,'%)'
          )
      ) %>%
      select(gene,downregulation)
    ,by='gene'
  )

suppressWarnings(set.seed(33,sample.kind=sample.kind))
tmp4e=
  tmp4e %>%
  left_join(
    tmp4d %>%
      filter(!is.na(OR) & (LB>1 | UB<1)) %>%
      mutate(
        regulation=ifelse(UB<1,'downregulated','upregulated')
      ) %>%
      lapply(X=seq(100),Y=.,function(X,Y){
        Y %>%
          mutate(
            regulation=
              regulation %>%
              sample(length(.),F)
          ) %>%
          group_by(gene,regulation) %>%
          summarize(n=n()) %>%
          ungroup() %>%
          group_by(gene) %>%
          mutate(
            total=sum(n)
            ,p_rand=n/total
          ) %>%
          mutate(b=X)
      }) %>%
      do.call(rbind,.) %>%
      select(b,gene,regulation,p_rand) %>%
      spread(regulation,p_rand,fill=0) %>%
      mutate(
        d_rand=downregulated-upregulated
      ) %>%
      select(-downregulated,-upregulated) %>%
      left_join(
        tmp4d %>%
          filter(!is.na(OR) & (LB>1 | UB<1)) %>%
          mutate(
            regulation=ifelse(UB<1,'downregulated','upregulated')
          ) %>%
          group_by(gene,regulation) %>%
          summarize(n=n()) %>%
          ungroup() %>%
          group_by(gene) %>%
          mutate(
            total=sum(n)
            ,p=n/total
          ) %>%
          select(gene,regulation,p) %>%
          spread(regulation,p,fill=0) %>%
          mutate(
            d=downregulated-upregulated
          ) %>%
          select(-downregulated,-upregulated)
        ,by=c('gene')
      ) %>%
      group_by(gene) %>%
      summarize(p.value=mean(abs(d)>abs(d_rand)))
    ,by='gene'
  ) %>%
  left_join(
    tmp4d %>%
      filter(!is.na(OR) & (LB>1 | UB<1)) %>%
      mutate_at(c('OR','LB','UB'),format,digits=2,scientific=T)
    ,by='gene'
  )
```

```{r Filter PE-FGR as outcome of the miRNA of interests, include=FALSE}
tmp4c2=
  tmp4b %>%
  filter(
    !(Health_state=='event'
      & Disease=='Preeclampsia')
    | is.na(Health_state)
  ) %>%
  mutate(
    Disease=
      ifelse(
        Health_state=='event'
        ,Disease
        ,NA
      ) %>%
      .[!is.na(.)] %>%
      .[!duplicated(.)]
  )
```

```{r Conduct regression analyses of the miRNAs on PE-FGR, include=FALSE}
tmp4d2=
  tmp4c2 %>%
  filter(!is.na(Sample_ID)) %>%
  group_by(gene,miRNA,Disease,Health_state) %>%
  summarize(n=n()) %>%
  ungroup() %>%
  spread(Health_state,n,fill=0) %>%
  mutate(
    total=nonevent+event
    ,p=event/total
  ) %>%
  filter(!(p==0|p==1)) %>%
  left_join(tmp4c2,by=c('gene','miRNA','Disease')) %>%
  group_by(gene,miRNA,Disease) %>%
  do(broom::tidy(suppressWarnings(
    glm(Health_state~log2RPM,data=.,family=binomial(link='logit'))
  ))) %>%
  filter(term=='log2RPM') %>%
  mutate(
    LB=round(exp(estimate-std.error),2)
    ,UB=round(exp(estimate+std.error),2)
    ,OR=round(exp(estimate),2)
  ) %>%
  select(gene,miRNA,Disease,OR,LB,UB) %>%
  ungroup()
```

```{r Determine total effect of miRNAs on PE-FGR per predictor, include=FALSE}
tmp4e2=
  tmp4c2 %>%
  mutate(
    availability=
      ifelse(
        miRNA
        %in%pull(
          tmp4c2 %>%
            filter(!is.na(Sample_ID)) %>%
            group_by(gene,miRNA,Disease,Health_state) %>%
            summarize(n=n()) %>%
            ungroup() %>%
            spread(Health_state,n,fill=0) %>%
            mutate(
              total=nonevent+event
              ,p=event/total
            ) %>%
            filter(!(p==0|p==1))
          ,miRNA
        )
        ,'available'
        ,'not_available'
      )
  ) %>%
  select(gene,miRNA,availability) %>%
  filter(!duplicated(.)) %>%
  group_by(gene,availability) %>%
  summarize(
    n=n()
    ,unavailable_miRNAs=
      paste0(
        ifelse(availability=='not_available',miRNA,NA) %>%
          .[!is.na(.)]
        ,collapse='|'
      )
  ) %>%
  ungroup() %>%
  group_by(gene) %>%
  mutate(
    unavailable_miRNAs=
      paste0(
        unavailable_miRNAs %>%
          .[.!='']
        ,collapse='|'
      )
  ) %>%
  spread(availability,n,fill=0) %>%
  mutate(
    total=available+not_available
    ,p=available/total
    ,availability=
      paste0(
        available
        ,'/'
        ,total
        ,' ('
        ,round(p*100,ifelse(total<10,0,ifelse(total<100,1,2)))
        ,'%)'
      )
  ) %>%
  select(gene,availability,unavailable_miRNAs) %>%
  left_join(
    tmp4d2 %>%
      mutate(
        significance=
          ifelse(!is.na(OR) & (LB>1 | UB<1),'significant','not_significant')
      ) %>%
      group_by(gene,significance) %>%
      summarize(n=n()) %>%
      spread(significance,n,fill=0) %>%
      mutate(
        total=significant+not_significant
        ,p=significant/total
        ,significance=
          paste0(
            significant
            ,'/'
            ,total
            ,' ('
            ,round(p*100,ifelse(total<10,0,ifelse(total<100,1,2)))
            ,'%)'
          )
      ) %>%
      select(gene,significance)
    ,by='gene'
  ) %>%
  left_join(
    tmp4d2 %>%
      filter(!is.na(OR) & (LB>1 | UB<1)) %>%
      mutate(
        regulation=ifelse(UB<1,'downregulated','upregulated')
      ) %>%
      group_by(gene,regulation) %>%
      summarize(n=n()) %>%
      spread(regulation,n,fill=0) %>%
      mutate(
        total=downregulated+upregulated
        ,p=downregulated/total
        ,downregulation=
          paste0(
            downregulated
            ,'/'
            ,total
            ,' ('
            ,round(p*100,ifelse(total<10,0,ifelse(total<100,1,2)))
            ,'%)'
          )
      ) %>%
      select(gene,downregulation)
    ,by='gene'
  )

suppressWarnings(set.seed(33,sample.kind=sample.kind))
tmp4e2=
  tmp4e2 %>%
  left_join(
    tmp4d2 %>%
      filter(!is.na(OR) & (LB>1 | UB<1)) %>%
      mutate(
        regulation=ifelse(UB<1,'downregulated','upregulated')
      ) %>%
      lapply(X=seq(100),Y=.,function(X,Y){
        suppressWarnings(set.seed(33+X,sample.kind=sample.kind))
        Y %>%
          mutate(
            regulation=
              regulation %>%
              sample(length(.),F)
          ) %>%
          group_by(gene,regulation) %>%
          summarize(n=n()) %>%
          ungroup() %>%
          group_by(gene) %>%
          mutate(
            total=sum(n)
            ,p_rand=n/total
          ) %>%
          mutate(b=X)
      }) %>%
      do.call(rbind,.) %>%
      select(b,gene,regulation,p_rand) %>%
      spread(regulation,p_rand,fill=0) %>%
      mutate(
        d_rand=downregulated-upregulated
      ) %>%
      select(-downregulated,-upregulated) %>%
      left_join(
        tmp4d2 %>%
          filter(!is.na(OR) & (LB>1 | UB<1)) %>%
          mutate(
            regulation=ifelse(UB<1,'downregulated','upregulated')
          ) %>%
          group_by(gene,regulation) %>%
          summarize(n=n()) %>%
          ungroup() %>%
          group_by(gene) %>%
          mutate(
            total=sum(n)
            ,p=n/total
          ) %>%
          select(gene,regulation,p) %>%
          spread(regulation,p,fill=0) %>%
          mutate(
            d=downregulated-upregulated
          ) %>%
          select(-downregulated,-upregulated)
        ,by=c('gene')
      ) %>%
      group_by(gene) %>%
      summarize(p.value=mean(abs(d)>abs(d_rand)))
    ,by='gene'
  ) %>%
  left_join(
    tmp4d2 %>%
      filter(!is.na(OR) & (LB>1 | UB<1)) %>%
      mutate_at(c('OR','LB','UB'),format,digits=2,scientific=T)
    ,by='gene'
  )
```

```{r Total effect of miRNAs on PE per predictor, echo=FALSE}
tmp4e
```

```{r Total effect of miRNAs on PE-FGR per predictor, echo=FALSE}
tmp4e2
```

```{r}
timestamper[14]=as.character(now())
```

```{r Create a function to show the shortest path, include=FALSE}
show_shortest_path=function(interaction_data,title=NULL){
  combination=
    interaction_data$inputs %>%
    expand.grid(.,.,stringsAsFactors=F) %>%
    filter(Var1!=Var2) %>%
    mutate(
      Var3=
        seq(nrow(.)) %>%
        sapply(X=.,Y=Var1,Z=Var2,function(X,Y,Z){
          c(Y[X],Z[X]) %>%
            sort() %>%
            paste(collapse=',')
        })
    ) %>%
    filter(!duplicated(Var3))
  
  shortest_path=
    combination %>%
    lapply(X=seq(nrow(.)),Y=.,function(X,Y){
      Z=suppressWarnings(
          interaction_data$nodes %>%
            rename_all(str_remove_all,'#') %>%
            select(-identifier) %>%
            rename(name=node,x=x_position,y=y_position) %>%
            right_join(
              interaction_data$nodes %>%
                rename_all(str_remove_all,'#') %>%
                select(node,x_position,y_position) %>%
                rename(node2=node,xend=x_position,yend=y_position) %>%
                right_join(
                  interaction_data$edges %>%
                    select(-node1_string_id,-node2_string_id) %>%
                    `colnames<-`(str_remove_all(colnames(.),'#')) %>%
                    gather(relation,score,-node1,-node2,-combined_score) %>%
                    filter(score!=0) %>%
                    right_join(
                      graph_from_data_frame(.) %>%
                        induced_subgraph(
                          shortest_paths(.,Y$Var1[X],Y$Var2[X]) %>%
                            .$vpath %>%
                            unlist() %>%
                            names()
                        ) %>%
                        get.edgelist() %>%
                        as.data.frame() %>%
                        setNames(paste0('node',1:2))
                      ,by=paste0('node',1:2)
                    ) %>%
                    filter(!duplicated(.)) %>%
                    rename(name=node1)
                  ,by='node2'
                )
              ,by='name'
            ) %>%
            select(-node2) %>%
            select(x,y,name,xend,yend,everything())
        )
      
      K=Z %>%
          ggplot(aes(x=x,y=y,xend=xend,yend=yend)) +
          geom_edges(
            aes(color=relation)
            ,size=1
            ,show.legend=F
          ) +
          geom_nodes(
            aes(color=color)
            ,size=5
            ,show.legend=F
          ) +
          geom_nodelabel_repel(
            aes(label=name,fill=name%in%interaction_data$inputs)
            ,label.padding=unit(0.15,'line')
            ,alpha=0.75
            ,size=3
            ,show.legend=F
          ) +
          facet_wrap(~relation) +
          scale_y_reverse() +
          scale_fill_manual(values=c('white','yellow')) +
          theme_blank() +
          ggtitle(title)
      
      print(K)
      Z
    })
  
  list(combination=combination,shortest_path=shortest_path)
}
```

```{r Load the STRING PPI data of the biomarkers and surrogates, include=FALSE}
tmp5=
  list(
    blood_20220303=c('ITGA5','P2RX7','IRF6')
    ,blcord_20220303=c('ITGA5','FANCI','TSEN15')
    ,bldeci_20220303=c('ITGA5','TPX2','WIPF3','ARID2')
    ,blfund_20220303=c('ITGA5','ARID2')
    ,blamni_20220303=c('P2RX7','INSM1')
    ,blamni2_20220303=c('IRF6','ALS2CL')
    ,cordam_20220304=c('SELV','INSM1','ALS2CL','TMEM38B')
    ,bldeciplac_20220314=
      c('ITGA5','TPX2','WIPF3','ARID2'
        ,'IRF6','TMEM38B','PLOD2'
        )
    ,shopath_all_20220304=
      c('ITGA5'
        ,'P2RX7'
        ,'IRF6'
        ,'FANCI'
        ,'TSEN15'
        ,'TPX2'
        ,'WIPF3'
        ,'ARID2'
        ,'INSM1'
        ,'ALS2CL'
        ,'SELV'
        ,'TMEM38B'
        )
    ,shopath_blood_20220304=
      c('ITGA5'
        ,'P2RX7'
        ,'IRF6'
        )
    ,shopath_bldecifund_20220304=
      c('ITGA5'
        ,'TPX2'
        ,'WIPF3'
        ,'ARID2'
        )
    ,shopath_blcordamplac_20220304=
      c('ITGA5'
        ,'P2RX7'
        ,'IRF6'
        ,'INSM1'
        ,'ALS2CL'
        ,'SELV'
        ,'TMEM38B'
        )
    ,shopath_bldeci_20220304=
      c('ITGA5'
        ,'TPX2'
        ,'WIPF3'
        ,'ARID2'
        )
    ,shopath_blfund_20220304=
      c('ITGA5'
        ,'ARID2'
        )
    ,shopath_blcord_20220310=
      c('ITGA5'
        ,'FANCI'
        ,'TSEN15'
        )
    ,shopath_blamni_20220310=
      c('P2RX7'
        ,'IRF6'
        ,'INSM1'
        ,'ALS2CL'
        )
    ,shopath_amniplac_20220310=
      c('INSM1'
        ,'ALS2CL'
        ,'TMEM38B'
        )
    ,shopath_amnicord_20220310=
      c('INSM1'
        ,'ALS2CL'
        ,'SELV'
        )
    ,shopath_cordplac_20220310=
      c('SELV'
        ,'TMEM38B'
        )
    ,shopath_bldeciplac_20220314=
      c('ITGA5','TPX2','WIPF3','ARID2'
        ,'IRF6','TMEM38B','PLOD2'
        )
  )

tmp5=
  tmp5 %>%
  lapply(X=names(.),Y=.,function(X,Y){
    list(
      inputs=Y[[X]]
      ,nodes=read_tsv(paste0('data/',X,'_string_network_coordinates.tsv'))
      ,edges=read_tsv(paste0('data/',X,'_string_interactions.tsv'))
      ,pathways=read_tsv(paste0('data/',X,'_enrichment.all.tsv'))
      ,pubmeds=read_tsv(paste0('data/',X,'_enrichment.PMID.tsv'))
    )
  }) %>%
  setNames(names(tmp5))
```

```{r Identify and show the shortest paths, include=FALSE}
tmp5=
  tmp5 %>%
  pblapply(X=names(.),Y=.,function(X,Y){
    Y[[X]] %>%
      .[!names(.)%in%c('combination','shortest_path')] %>%
      c(show_shortest_path(Y[[X]],X))
  }) %>%
  setNames(names(tmp5))
```

```{r Select the pathways overrepresented by the shortest paths, include=FALSE}
if(run_heavy_computation){
  tmp6=
    tmp5 %>%
    lapply(X=names(.),Y=.,function(X,Y){
      Z=Y[[X]]$pubmeds %>%
        mutate(`#category`='PubMed') %>%
        rename(`term ID`=`#term ID`) %>%
        select(`#category`,everything()) %>%
        list() %>%
        c(list(Y[[X]]$pathways)) %>%
        .[2:1] %>%
        do.call(rbind,.) %>%
        `colnames<-`(
          colnames(.) %>%
            str_remove_all('[:punct:]') %>%
            str_to_lower() %>%
            str_replace_all('\\s+','_')
        ) %>%
        mutate(id=X) %>%
        select(id,everything())
      
      K=Y[[X]]$combination %>%
        lapply(X=seq(nrow(.)),Y=.,Z=Y[[X]]$shortest_path,function(X,Y,Z){
          K=Z[[X]] %>%
            select(name,x,y) %>%
            rename(node1=name) %>%
            filter(!duplicated(.)) %>%
            left_join(
              Z[[X]] %>%
                select(name,xend,yend) %>%
                rename(node2=name,x=xend,y=yend) %>%
                filter(!duplicated(.))
              ,by=c('x','y')
            ) %>%
            select(-x,-y) %>%
            graph_from_data_frame() %>%
            shortest_paths(Y$Var1[X],Y$Var2[X]) %>%
            .$vpath %>%
            unlist() %>%
            names()
          
          if(length(K)<=2){
            paste0(K,collapse=',')
          }else{
            seq(2,length(K)) %>%
              lapply(function(x){
                seq(1,length(K)-x+1,1) %>%
                  lapply(function(y){
                    K[seq(y,y+x-1)] %>%
                      paste0(collapse=',')
                  })
              }) %>%
              unlist()
          }
        }) %>%
        unlist() %>%
        sapply(function(x){
          paste0(sort(str_split(x,',')[[1]]),collapse=',')
        }) %>%
        .[!duplicated(.)]
      
      L=K %>%
        pblapply(function(x){
          Z %>%
            mutate(
              matched=
                matching_proteins_in_your_network_labels %>%
                sapply(function(x){
                  paste0(sort(str_split(x,',')[[1]]),collapse=',')
                })
              ,matched=matched==x
            ) %>%
            select(matched) %>%
            setNames(x)
        }) %>%
        do.call(cbind,.)
      
      cbind(Z,L) %>%
        pivot_longer(
          seq(ncol(Z)+1,ncol(Z)+ncol(L))
          ,names_to='gene'
          ,values_to='matched'
        ) %>%
        group_by_at(seq(1,ncol(Z))) %>%
        summarize(
          genes=
            ifelse(matched,gene,NA) %>%
            .[!is.na(.)] %>%
            sort() %>%
            paste0(collapse='|')
          ,matched=sum(matched)
        ) %>%
        ungroup() %>%
        arrange(factor(term_id,Z$term_id)) %>%
        mutate(
          background_gene_class=
            ifelse(
              matched==1
              & background_gene_count>=15 & background_gene_count<=200
              ,1
              ,ifelse(
                matched==1
                & background_gene_count>=10 & background_gene_count<=500
                ,2
                ,3
              )
            )
        ) %>%
        filter(matched==1) %>%
        group_by(genes) %>%
        filter(background_gene_class==min(background_gene_class)) %>%
        ungroup() %>%
        mutate(
          genes=factor(genes,unique(genes))
          ,category=factor(category,unique(category))
        ) %>%
        group_by(genes,category) %>%
        arrange(desc(strength),desc(observed_gene_count)) %>%
        filter(
          strength==max(strength)
          & observed_gene_count==max(observed_gene_count)
        ) %>%
        ungroup() %>%
        mutate_at(c('genes','category'),as.character) %>%
        select(
          id,genes,observed_gene_count,term_id,category
          ,background_gene_class
          ,everything()
        )
    }) %>%
    setNames(names(tmp5))
  
  saveRDS(tmp6,'data/tmp6.rds')
}else{
  cat(readRDS('data/log.rds')[['tmp6']])
  tmp6=readRDS('data/tmp6.rds')
}
```

```{r Show the selected pathways, eval=FALSE, include=FALSE}
tmp6 %>%
  lapply(
    arrange
    ,desc(observed_gene_count),genes,background_gene_class,desc(strength)
  ) %>%
  lapply(kable) %>%
  lapply(kable_classic)
```

```{r Select and show the PMID pathways, eval=FALSE, include=FALSE}
tmp5 %>%
  lapply(X=names(.),Y=.,function(X,Y){
    Y[[X]]$pathways %>%
      rbind(
        Y[[X]]$pubmeds %>%
          rename(`term ID`=`#term ID`) %>%
          mutate(`#category`='PubMed')
      ) %>%
      mutate(id=X) %>%
      select(id,everything())
  }) %>%
  do.call(rbind,.) %>%
  filter(str_detect(str_to_lower(`term description`),'vitamin d')) %>%
  filter(str_detect(id,'shopath')) %>%
  filter(!str_detect(id,'_all|_bldecifund|_blcordamplac|_bldeciplac')) %>%
  mutate(
    `background gene class`=
      ifelse(
        `background gene count`>=15 & `background gene count`<=200
        ,1
        ,ifelse(
          `background gene count`>=10 & `background gene count`<=500
          ,2
          ,3
        )
      )
  ) %>%
  select(
    `observed gene count`
    ,`matching proteins in your network (labels)`
    ,id
    ,`#category`
    ,`background gene class`
    ,strength
    ,everything()
  ) %>%
  arrange(
    desc(`observed gene count`)
    ,`matching proteins in your network (labels)`
    ,factor(id,unique(id))
    ,factor(`#category`,unique(`#category`))
    ,`background gene class`
    ,desc(strength)
  ) %>%
  group_by(`matching proteins in your network (labels)`,id,`#category`) %>%
  slice(1) %>%
  ungroup() %>%
  kable() %>%
  kable_classic()
```

```{r Identify available types of the interaction, eval=FALSE, include=FALSE}
tmp5 %>%
  lapply(X=names(.),Y=.,function(X,Y){
    Y[[X]]$edges %>%
      mutate(id=X)
  }) %>%
  do.call(rbind,.) %>%
  select(id,everything()) %>%
  gather(
    interaction,score
    ,-id,-`#node1`,-node2,-node1_string_id,-node2_string_id,-combined_score
  ) %>%
  filter(score!=0) %>%
  group_by(id,interaction) %>%
  summarize(n=n()) %>%
  filter(str_detect(id,'shopath') & !str_detect(id,'_all|_blcordamplac')) %>%
  group_by(interaction) %>%
  summarize(total=n())
```

```{r}
timestamper[15]=as.character(now())
```

```{r Load PTM and protein data, include=FALSE}
tmp7=read_csv('data/ptm.csv')
tmp7b=
  read_csv('data/ptm_effect.csv') %>%
  # https://en.wikipedia.org/wiki/ADP-ribosylation
  rbind(
    data.frame(
      ptm='ADP-ribosylation'
      ,effect=
        c('cell signaling'
          ,'DNA repair'
          ,'transcription' # gene regulation
          ,'apoptosis')
      ,covalent_attachment='small chemical group'
    )
  )
tmp8=read_csv('data/protein.csv')
```

```{r Identify location of protein overexpression and the PTM effect, echo=TRUE}
rbind(
    tmp7 %>%
      rename(entity=gene) %>%
      mutate(type='protein') %>%
      mutate(ptm=ifelse(ptm=='ADP-ribosylation','ribosylation',ptm)) %>%
      left_join(
        tmp7b %>%
          mutate(
            ptm=
              str_sub(
                ptm
                ,sapply(ptm,function(x){
                  sum(c(str_locate_all(x,'-')[[1]][1],1),na.rm=T)
                })
                ,str_length(ptm)
              )
          ) %>%
          select(ptm,covalent_attachment) %>%
          filter(!duplicated(.))
        ,by='ptm'
      )
    ,tmp7b %>%
      mutate(
        ptm=
          str_sub(
            ptm
            ,sapply(ptm,function(x){
              sum(c(str_locate_all(x,'-')[[1]][1],1),na.rm=T)
            })
            ,str_length(ptm)
          )
      ) %>%
      rename(entity=effect) %>%
      mutate(type='effect') %>%
      mutate(entity=str_to_sentence(entity) %>% str_replace_all('Dna ','DNA '))
  ) %>%
  mutate(value=1) %>%
  mutate(
    ptm=
      ptm %>%
      str_to_sentence() %>%
      factor(c(
        'phosphorylation'
        ,'glycosylation'
        ,'nitrosylation'
        ,'methylation'
        ,'acetylation'
        ,'ribosylation'
        ,'palmitoylation'
        ,'myristoylation'
        ,'prenylation'
        ,'sumoylation'
        ,'ubiquitination'
      ) %>% str_to_sentence())
    ,entity=
      entity %>%
      factor(c(
        c('apoptosis'
          ,'protein stability'
          ,'protein-protein interaction'
          ,'protein membrane'
          ,'protein trafficking'
          ,'protein thermodynamic and cinetic'
          ,'protein activity'
          ,'extracellular protein export'
          ,'cell signaling'
          ,'transcription'
          ,'DNA repair'
          ,'cell cycle division'
          ,'immune response'
          ,'chromosome maintenance'
          ,'chromosome assembly'
        ) %>% str_to_sentence() %>% str_replace_all('Dna ','DNA ')
        ,c('ITGA5'
           ,'P2RX7'
           ,'IRF6'
           ,'FANCI'
           ,'SELENOV'
           ,'TSEN15'
           ,'TPX2'
           ,'WIPF3'
           ,'ARID2'
           ,'INSM1'
           ,'TMEM38B'
           ,'ALS2CL'
        )
      ) %>% rev())
    ,covalent_attachment=
      covalent_attachment %>%
      str_to_sentence() %>%
      factor(c(
        'small chemical group'
        ,'acyl chain'
        ,'small protein'
      ) %>% str_to_sentence())
    ,type=
      type %>%
      str_to_sentence() %>%
      factor(c(
        'protein'
        ,'effect'
      ) %>% str_to_sentence())
  ) %>%
  mutate(
    value=
      entity
      %in%pull(
        tmp8 %>%
          rbind(
            rbind(tmp4e,tmp4e2) %>%
              filter(
                sapply(downregulation,function(x){
                  y=str_split(x,'\\s+')[[1]][2] %>%
                    str_sub(2,str_count(.)-2)
                  as.numeric(y)>50
                })
                & p.value<=0.05
                & as.numeric(UB)<1
              ) %>%
              select(gene) %>%
              filter(!duplicated(gene)) %>%
              mutate(location='placenta',strongest=NA)
          ) %>%
          filter(
            (gene%in%c('ITGA5','P2RX7','IRF6')
             & location%in%c(
               'blood'
               ,'plasma'
               ,'leukocytes','monocyte','CD4 T cells','B lymphocyte','NK cells'
             )
            )
            |
            (gene%in%c('ITGA5','FANCI','SELENOV','TSEN15')
             & location%in%c(
               'blood'
               ,'plasma'
               ,'leukocytes','monocyte','CD4 T cells','B lymphocyte','NK cells'
               ,'placenta'
             )
            )
            |
            (gene%in%c('ITGA5','TPX2','WIPF3','ARID2')
             & location%in%c(
               # 'leukocytes','CD4 T cells','B lymphocyte','NK cells'
               'uterus','placenta'
             )
            )
            |
            (gene%in%c('P2RX7','INSM1')
             & location%in%c(
               # 'leukocytes','CD4 T cells','B lymphocyte','NK cells'
               'placenta'
             )
            )
            |
            (gene%in%c('IRF6','ALS2CL','TMEM38B')
             & location%in%c(
               # 'leukocytes','CD4 T cells','B lymphocyte','NK cells'
               'placenta'
             )
            )
          )
        ,gene
      )
  ) %>%
  group_by(ptm) %>%
  mutate(ptm_on=sum(value,na.rm=T)>0) %>%
  ungroup() %>%
  mutate(
    value=
      ifelse(
        type=='Protein'
        ,ifelse(value,'Overexpressed','Not overexpressed')
        ,ifelse(
          ptm_on
          ,'Applied'
          ,'Not applied'
        )
      ) %>%
      factor(c(
        'Not overexpressed','Overexpressed'
        ,'Not applied','Applied'
      ))
  ) %>%
  ggplot(aes(ptm,entity,fill=value)) +
  geom_tile(color='white') +
  facet_grid(type~covalent_attachment,scales='free',space='free',switch='y') +
  xlab('Post-translational modification (PTM)') +
  ylab('') +
  scale_fill_manual('',values=c('grey','red','darkgrey','black')) +
  theme(
    panel.grid=element_blank()
    ,axis.text.x=
      element_text(size=unit(8,'pt'),family='sans',angle=22.5,hjust=1)
    ,axis.text.y=
      element_text(size=unit(8,'pt'),family='sans',vjust=0.5)
    ,strip.placement='outside'
    ,legend.position='top'
  )
```

```{r}
timestamper[16]=as.character(now())
```

```{r}
timestamper %>%
  data.frame(seq=seq(length(.)),timestamp=.) %>%
  mutate(timestamp=as_datetime(timestamp)) %>%
  mutate(timestamp=timestamp-min(timestamp)) %>%
  mutate(timestamp=as.duration(timestamp)/dminutes(1)) %>%
  left_join(
    mutate(.,seq=seq+1) %>%
      rename(timestamp0=timestamp)
    ,by='seq'
  ) %>%
  slice(-1) %>%
  mutate(duration=timestamp-timestamp0) %>%
  ggplot(aes(seq,duration)) +
  geom_col()
```






































